---
output: github_document
---

```{r, echo = FALSE, message = FALSE}
knitr::opts_chunk$set(
  fig.align = 'center',
  collapse = TRUE,
  comment = "#>",
  fig.path = "img/README-")
```

# valr <img src="man/figures/logo.png" align="right" />

[![Build Status](https://travis-ci.org/rnabioco/valr.svg?branch=master)](https://travis-ci.org/rnabioco/valr)
[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/rnabioco/valr?branch=master&svg=true)](https://ci.appveyor.com/project/jayhesselberth/valr)
[![Coverage Status](https://img.shields.io/codecov/c/github/rnabioco/valr/master.svg)](https://codecov.io/github/rnabioco/valr?branch=master)
[![](http://www.r-pkg.org/badges/version/valr)](https://CRAN.R-project.org/package=valr)
[![](http://cranlogs.r-pkg.org/badges/valr?color=FFD700)](http://www.r-pkg.org/pkg/valr)
[![DOI](https://zenodo.org/badge/49370633.svg)](https://zenodo.org/badge/latestdoi/49370633)

**`valr` provides tools to read and manipulate genome intervals and signals**, similar to the [`BEDtools`][1] suite. `valr` enables analysis in the R/RStudio environment, leveraging modern R tools in the [`tidyverse`][14] for a terse, expressive syntax. Compute-intensive algorithms are implemented in [`Rcpp`][3]/C++, and many methods take advantage of the speed and grouping capability provided by [`dplyr`][2].

## Installation

The latest stable version can be installed from CRAN:

``` r
install.packages('valr')
```

The latest development version can be installed from github:

``` r
# install.packages("devtools")
devtools::install_github('rnabioco/valr')
```

## Why `valr`?

**Why another tool set for interval manipulations?** Based on our experience teaching genome analysis, we were motivated to develop interval arithmetic software that faciliates genome analysis in a single environment (RStudio), eliminating the need to master both command-line and exploratory analysis tools.

**Note:** `valr` can currently be used for analysis of pre-processed data in BED and related formats. We plan to support BAM and VCF files soon via tabix indexes.

### Familiar tools, natively in R

The functions in `valr` have similar names to their `BEDtools` counterparts, and so will be familiar to users coming from the `BEDtools` suite. Unlike other tools that wrap `BEDtools` and write temporary files to disk, `valr` tools run natively in memory. Similar to [`pybedtools`](https://daler.github.io/pybedtools/#why-pybedtools), `valr` has a terse syntax:

```{r syntax_demo, message = FALSE}
library(valr)
library(dplyr)

snps <- read_bed(valr_example('hg19.snps147.chr22.bed.gz'), n_fields = 6)
genes <- read_bed(valr_example('genes.hg19.chr22.bed.gz'), n_fields = 6)

# find snps in intergenic regions
intergenic <- bed_subtract(snps, genes)
# find distance from intergenic snps to nearest gene
nearby <- bed_closest(intergenic, genes)

nearby %>%
  select(starts_with('name'), .overlap, .dist) %>%
  filter(abs(.dist) < 5000)
```

### Visual documentation

`valr` includes helpful glyphs to illustrate the results of specific operations, similar to those found in the `BEDtools` documentation. For example, `bed_glyph()` illustrates the result of intersecting `x` and `y` intervals with `bed_intersect()`:

```{r intersect_glyph}
library(valr)

x <- trbl_interval(
  ~chrom, ~start, ~end,
  'chr1', 25,     50,
  'chr1', 100,    125
)

y <- trbl_interval(
  ~chrom, ~start, ~end,
  'chr1', 30,     75
)

bed_glyph(bed_intersect(x, y))
```

### Reproducible reports 

`valr` can be used in RMarkdown documents to generate reproducible work-flows for data processing. Because computations in `valr` are fast, it can be for exploratory analysis with `RMarkdown`, and for interactive analysis using `shiny`.

### Remote databases

Remote databases can be accessed with `db_ucsc()` (to access the UCSC Browser) and `db_ensembl()` (to access Ensembl databases).

```{r db, eval = FALSE}
# access the `refGene` tbl on the `hg38` assembly
ucsc <- db_ucsc('hg38')
tbl(ucsc, 'refGene')
```

## API

Function names are similar to their their [BEDtools][1] counterparts, with some additions.

### Data types

* Create new interval sets with `tbl_interval()` and `tbl_genome()`. Coerce existing `GenomicRanges::GRanges` objects with `as.tbl_interval()`.

### Reading data

* Read BED and related files with `read_bed()`, `read_bed12()`, `read_bedgraph()`, `read_narrowpeak()` and `read_broadpeak()`.
  
* Read genome files containing chromosome name and size information with `read_genome()`.
  
* Load VCF files with `read_vcf()`.

* Access remote databases with `db_ucsc()` and `db_ensembl()`.

### Transforming single interval sets

* Adjust interval coordinates with `bed_slop()` and `bed_shift()`, and create new flanking intervals with `bed_flank()`.

* Combine nearby intervals with `bed_merge()` and identify nearby intervals with `bed_cluster()`.  

* Generate intervals not covered by a query with `bed_complement()`.

* Order intervals with `bed_sort()`.

### Comparing multiple interval sets

* Find overlaps between sets of intervals with `bed_intersect()`.

* Apply functions to overlapping sets of intervals with `bed_map()`.

* Remove intervals based on overlaps with `bed_subtract()`.

* Find overlapping intervals within a window with `bed_window()`.

* Find closest intervals independent of overlaps with `bed_closest()`.

### Randomizing intervals

* Generate random intervals with `bed_random()`.

* Shuffle the coordinates of intervals with `bed_shuffle()`.

* Sample input intervals with `dplyr::sample_n()` and `dplyr::sample_frac()`.

### Interval statistics

* Calculate significance of overlaps between sets of intervals with `bed_fisher()` and  `bed_projection()`.

* Quantify relative and absolute distances between sets of intervals with `bed_reldist()` and `bed_absdist()`.

* Quantify extent of overlap between sets of intervals with `bed_jaccard()`.

### Utilities

* Create features from BED12 files with `create_introns()`, `create_tss()`, `create_utrs5()`, and `create_utrs3()`.

* Visualize the actions of valr functions with `bed_glyph()`.

* Constrain intervals to a genome reference with `bound_intervals()`.

* Subdivide intervals with `bed_makewindows()`.

* Convert BED12 to BED6 format with `bed12_to_exons()`.

* Calculate spacing between intervals with `interval_spacing()`.

## Related work

* Command-line tools [BEDtools][1] and [bedops][5].

* The Python library [pybedtools][4] wraps BEDtools.

* The R packages [GenomicRanges][6], [bedr][7], [IRanges][8] and [GenometriCorr][9] provide similar capability with a different philosophy.

[1]: http://bedtools.readthedocs.org/en/latest/
[2]: https://github.com/hadley/dplyr
[3]: http://www.rcpp.org/
[4]: https://daler.github.io/pybedtools/
[5]: http://bedops.readthedocs.org/en/latest/index.html
[6]: https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html
[7]: https://CRAN.R-project.org/package=bedr 
[8]: https://bioconductor.org/packages/release/bioc/html/IRanges.html
[9]: http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529
[10]: http://rmarkdown.rstudio.com/
[11]: https://www.r-bloggers.com/why-i-dont-like-jupyter-fka-ipython-notebook/
[12]: https://bitbucket.org/snakemake/snakemake/wiki/Home
[13]: http://shiny.rstudio.com/
[14]: http://tidyverse.org
