---
title: "`valr`: Genome interval arithmetic in R"
author: "Jay Hesselberth <jay.hesselberth@gmail.com>"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: lumen
    highlight: tango
    keep_md: true
    # code_folding: show
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: true
---

```{r knitr_opts, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

```{r init, echo = FALSE, message = FALSE}
library(valr)
library(dplyr)
```

# Overview 

`valr` provides tools to read and manipulate genome intervals and signals, similar to the standalone [`BEDtools`][1] suite. `valr` natively enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in [`Rcpp`][5]/C++, and many methods take advantage of the speed and grouping capability provided by [`dplyr`][3]. 

We provide several introductions to `valr`:

  - A comprehensive [`vignette`][] covering the core methods.
  - A tutorial that demonstrates how to use `valr` in "real-world" applicatins.
  - A `shiny` application that demonstrates interactive analysis of genome-scale data sets.

## Installation

`valr` can be installed from github, and will be eventually deposited in CRAN.

``` r
> devtools::install_github('jayhesselberth/valr')
```
## Comparison to other tools

**Why another tool set for interval manipulations?** We already have [BEDtools][1], [bedops][8], [pybedtools][7], [GenomeRanges][9], [bedr][11] and [IRanges][12]. 

We were motivated to develop a toolset that:

  - Combines analysis and visualization in RStudio.
  - Can be used to generate reports with Rmarkdown.
  - Is highly extensible. New tools are quickly implemented on the R side.
  - Leverages the "modern R" syntax, using `dplyr` and `%>%` for pipes.
  - Maximizes speed by implementing compute-intensive algorithms in `Rcpp`.
  - Facilitates interactive visulaizations with [`shiny`][10].

We anticipate `valr` will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as `tidyr` makes it easy to convert these to BED12.

At this point you might be expecting [plots][13] of speed / memory usage versus interval number. Certain algorithms in `valr` were implemented in `Rcpp` to be fast (including [intersect, merge, subtract, closest][14]), enabling fluid interactive analysis. See the [benchmarks](#benchmarks) section for demonstrations.

## Non-standard evaluation

Several of the methods in `valr` use [NSE][4] for an expressive syntax. Columns are referred to by name and can be used in multiple name/value expressions for summaries.

```{r NSE, eval = FALSE}
bed_map(x, y, mean = mean(value), var = var(value))
bed_merge(x, concat = concat(value), max = max(value))
```


## Getting started

Here is an example using `valr` that creates 

```{r tour}
library(valr)
```

## Reading data {.tabset}

`valr` has several methods to read interval data. These methods:

  - Take local files or URLs as input.
  - Return a `data.frame` in `tibble::tbl_df` format.
  - Assign consistent `chrom`, `start` and `end` column names.
  - Use [`readr`][6] for speed.
  - Coerce column types.

The methods include:

  - `read_bed()`: read a BED3+ file
  - `read_bed12()`: read a BED12 file
  - `read_bedgraph()`: read a bedGraph file
  - `read_genome()`: read a UCSC "chrom size" file
  - `read_vcf()`: `XXX` not implemented yet

### Reading BED files 

```{r read_bed}
read_bed(system.file('extdata', '3fields.bed.gz', package = 'valr'))
read_bed(n_fields = 6, system.file('extdata', '6fields.bed.gz', package = 'valr'))
read_bed12(system.file('extdata', '12fields.bed.gz', package = 'valr'))
```

### Reading Genome files

```{r read_genome}
genome <- read_genome(system.file('extdata', 'hg19.chrom.sizes.gz', package = 'valr'))
```

### Reading bedGraph files

```{r read_bedgraph}
read_bedgraph(system.file('extdata', 'test.bg.gz', package = 'valr'))
```

### Reading VCF files

```{r read_vcf, eval = FALSE}
read_vcf(system.file('extdata', 'sample.vcf.gz', package = 'valr'))
```
  
## Interval manipulations

These methods compares sets of intervals in a variety of ways. Many have the same name as the corresponding `BEDtool`. Some familiar BEDtools are implemented as `dplyr` pipes (e.g., see the [`coverage`](#coverage) section).

  - `bed_intersect()`: find overlapping intervals
  - `bed_merge()`: combine overlapping and nearby intervals into a single interval.
  - `bed_cluster()`: Cluster (but don't merge) overlapping/nearby intervals.
  - `bed_random()`: Generate random intervals in a genome.
  - `bed_closest()`: find the closest intervals independent of overlaps
  - `bed_map()`: apply a function to selected columns for overlapping intervals
  - `bed_complement()`: extract intervals *not* represented by an interval file.
  - `bed_sort()`: order intervals 
  - `bed_flank()`: Generate new flanking intervals
  - `bed_slop()`: Expand the size of input intervals
  - `bed_window()`: Find overlapping intervals within a window
  - `bed_subtract()`: Remove intervals based on overlaps b/w two files (not implemented yet)
  - `bed_shuffle()`:  Randomly redistrubute intervals in a genome (not implemented yet)

### Calculating coverage

Instead of implementing `coverage` and `genomecov` directly, coverage calculations are done with the `summarize` tool in `dplyr`.

```{r coverage}
```

### Clustering intervals

`bed_cluster` identifies clustered intervals and assigns them a unique `.id`.

```{r cluster}
x <- bed_random(genome)
y <- bed_cluster(x)
y
```

### Mirror intervals

`bed_complement` identifies intervals in a genome that are not covered by an input.

```{r complement}
x <- bed_random(genome)

bed_complement(x, genome)
```

### Grouping related intervals

The `group_by` operation from `dplyr` serves a similar purpose to BEDtools `groupby`, and can be combined with `dplyr::summarize` to calculate summary statistics from groups.

```{r group_by}
x <- bed_random(genome)
y <- bed_random(genome)

bed_intersect(x, y) %>%
  group_by(chrom) %>%
  summarize(overlap.total = sum(.overlap))
```

### Intersecting intervals

`bed_intersect` is implement using an [interval tree][16] in `Rcpp`. Column names in the result have `.x` and `.y` suffixes, and an `.overlap` column contains the size of the intersection (values of `0` indicate book-ended, or touching intervals). See the [benchmarks](#benchmarks) section for timing. Though `bed_intersect` is pretty fast already, we intend to further improve upon this by parallization with [`RcppParallel`][15].

```{r intersect}
# intersect two sets of 1e6 intervals from hg19
x <- bed_random(genome)
y <- bed_random(genome)

bed_intersect(x, y)
```

### Creating sub-intervals

`bed_makewindows` creates sub-intervals from input intervals. The returned `data_frame` has a `win_id` column that can be used for grouping. Instead of naming intervals directly, new names for the intervals can be created with `dplyr`.

```{r makewindows}
# tally number of created intervals
x <- bed_random(genome, n = 1)
wins <- bed_makewindows(x, genome, win_size = 10)

group_by(wins, win_id) %>% summarize(num.win = n())

# create new ID based on `name` and `win_id`
bed_makewindows(x, genome, win_size = 10) %>%
  mutate(id = stringr::str_c('newid', '-', win_id))
```

### Mapping signals to intervals

`bed_map` maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs. 

```{r map}
bedfile <- system.file('extdata', 'genes.hg19.chr22.bed.gz', package = 'valr')
bgfile  <- system.file('extdata', 'hela.h3k4.chip.bg.gz', package = 'valr')

x <- read_bed(bedfile, n_fields = 6)
y <- read_bedgraph(bgfile)

bed_map(x, y, means = mean(value.y), sds = sd(value.y))
```

### Merging intervals

`bed_merge` identifies overlapping intervals and reports new merged ones. `is_merged` asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.

```{r merge}
# 1e6 random intervals
n <- 1e6
x <- bed_random(genome, n = n)
is_merged(x)
x

# add some signal
x <- x %>% mutate(signal = runif(n))

y <- bed_merge(x)
is_merged(y)
y

bed_merge(x, maxs = max(signal))
```

### Randomzing intervals {.tabset}

`valr` provides methods for creating new random intervals or permutations of existing intervals:

  - `bed_random` generates random intervals from an input `genome`.
  - `bed_shuffle` shuffles coordinates given a set of input intervals.
  - Random sampling of input intervals is done with `dplyr`.

#### Random

`bed_random` generates random intervals from an input genome. The numbers of intervals from each `chrom` are proporitional to each chrom `size`.

```{r random}
x <- bed_random(genome, n = 1e6, length = 1e3)
x

# numbers of sampled intervals are proportional to chrom size
group_by(x, chrom) %>% summarize(n = n()) %>% arrange(desc(n))
```

#### Sample

Sampling can be done using `dplyr`:

```{r sample}
x <- bed_random(genome)

# sample by number
sample_n(x, 1e3, replace = FALSE)

# or fraction
sample_frac(x, 0.1, replace = FALSE)

# or sample intervals within groups
group_by(x, chrom) %>% sample_n(1)
```

#### Shuffle

`bed_shuffle` shuffles input intervals. Interval sizes are equal in the input and output.

```{r shuffle, eval = FALSE}
y <- bed_shuffle(x)

any(x$start == y$start)

all(x$end - x$start == y$end - y$start)
```

### Sorting intervals

`bed_sort` orders intervals based on a specification. `is_sorted` asks whether a tbl is already sorted.

```{r sort}
x <- bed_random(genome)
is_sorted(x)

y <- bed_sort(x)
y

is_sorted(y)
```

# Combining tools

# Benchmarks

```{r benchmarks}
# 1e6 random 1 kb intervals from hg19
x <- bed_random(genome)
y <- bed_random(genome)

library(microbenchmark)
microbenchmark(
  bed_random(genome),
  bed_intersect(x, y),
  bed_merge(x),
  bed_subtract(x, y),
  bed_complement(x, genome),
  times = 1,
  unit = 's'
)
```

<!-- # On naming things -->

<!-- > In a false quarrel there is no true valor. -->
<!-- >   *- William Shakespeare* -->

<!-- ![](http://i.imgur.com/afBYE36.jpg) -->


[1]: http://bedtools.readthedocs.org/en/latest/
[2]: https://github.com/arq5x/chrom_sweep
[3]: https://github.com/hadley/dplyr
[4]: https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html
[5]: http://www.rcpp.org/
[6]: https://github.com/hadley/readr
[7]: https://pythonhosted.org/pybedtools/
[8]: http://bedops.readthedocs.org/en/latest/index.html
[9]: https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html
[10]: http://shiny.rstudio.com/
[11]: https://cran.r-project.org/web/packages/bedr/index.html
[12]: https://bioconductor.org/packages/release/bioc/html/IRanges.html
[13]: http://bedtools.readthedocs.io/en/latest/#performance
[14]: https://github.com/jayhesselberth/valr/tree/master/src
[15]: https://rcppcore.github.io/RcppParallel/
[16]: https://github.com/ekg/intervaltree/

