---
title: '`valr`: Genome interval arithmetic in R'
author: "Jay Hesselberth <jay.hesselberth@gmail.com>"
date: '`r Sys.Date()`'
output:
  html_document:
    highlight: pygments
    keep_md: yes
    mathjax: default
    self_contained: yes
    theme: lumen
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: no
      smooth_scroll: yes
  pdf_document:
    toc: yes
    toc_depth: '4'
  word_document:
    toc: yes
    toc_depth: '4'
---

```{r knitr_opts, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

```{r init, echo = FALSE, message = FALSE}
library(valr)
library(dplyr)
```

# Overview 

`valr` provides tools to read and manipulate genome intervals and signals, similar to the standalone [`BEDtools`][1] suite. `valr` enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in [`Rcpp`][5]/C++, and many methods take advantage of the speed and grouping capability provided by [`dplyr`][3]. 

We provide several introductions to `valr`:

  - This comprehensive `vignette` covering the core methods.
  - A tutorial that demonstrates how to use `valr` in "real-world" applications.
  - A `shiny` application that demonstrates interactive analysis of genome-scale data sets.

## Installation

`valr` can be installed from github, and will be eventually deposited in CRAN.

```r
devtools::install_github('jayhesselberth/valr')
```

## Comparison to other tools

**Why another tool set for interval manipulations?** We already have [BEDtools][1], [bedops][8], [pybedtools][7], [GenomicRanges][9], [bedr][11] and [IRanges][12]. 

We were motivated to develop a toolset that:

  - Combines analysis and visualization in RStudio.
  - Can be used to generate reports with Rmarkdown.
  - Is highly extensible. New tools are quickly implemented on the R side.
  - Leverages the "modern R" syntax, using `dplyr` and the pipe operator from `magrittr` (`%>%`).
  - Maximizes speed by implementing compute-intensive algorithms in `Rcpp`.
  - Facilitates interactive visulaizations with [`shiny`][10].

We anticipate `valr` will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as `tidyr` makes it easy to convert these to BED12.

Certain algorithms in `valr` were implemented in `Rcpp` (including [intersect, merge, subtract, closest][14]) to enable fluid interactive analysis. See the [benchmarks](#benchmarks) section for details.

## Non-standard evaluation

Several of the methods in `valr` use [NSE][4] for an expressive syntax. Columns are referred to by name and can be used in multiple name/value expressions for summaries.

```{r NSE, eval = FALSE}
bed_map(x, y, mean = mean(value), var = var(value))
bed_merge(x, concat = concat(value), max = max(value))
```

## Getting started

Here is an example using `valr` that creates 

```{r tour}
library(valr)
```

# Reading data {.tabset}

`valr` has several methods to read interval data. These methods:

  - Take local files or URLs as input.
  - Return a `data.frame` in `tibble::tbl_df` format.
  - Assign consistent `chrom`, `start` and `end` column names.
  - Use [`readr`][6] for speed.
  - Coerce column types.

The methods include:

  - `read_bed()`: read a BED3+ file
  - `read_bed12()`: read a BED12 file
  - `read_bedgraph()`: read a bedGraph file
  - `read_genome()`: read a UCSC "chrom size" file
  - `read_vcf()`: read the Variant Call Format
  - `read_narrowpeak()`: read narrowPeak files
  - `read_broadpeak()`: read broadPeak files

## BED files 

```{r read_bed}
read_bed(system.file('extdata', '3fields.bed.gz', package = 'valr'))
read_bed(n_fields = 6, system.file('extdata', '6fields.bed.gz', package = 'valr'))
read_bed12(system.file('extdata', '12fields.bed.gz', package = 'valr'))
```

## Genome files

```{r read_genome}
genome <- read_genome(system.file('extdata', 'hg19.chrom.sizes.gz', package = 'valr'))
```

## bedGraph files

```{r read_bedgraph}
read_bedgraph(system.file('extdata', 'test.bg.gz', package = 'valr'))
```

## VCF files

`read_vcf()` reads VCF files and assigns `chrom`, `start` and `end` columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the `REF` field in the original file.

```{r read_vcf}
read_vcf(system.file('extdata', 'test.vcf.gz', package = 'valr'))
```
  
# Interval manipulations

`valr` implements several methods for manipulating sets of intervals. Some methods  operate on a single set of intervals, while others compare two sets of intervals.

Many methods the same name as the corresponding `BEDtool`, and some commonly used `BEDtools` are implemented as `dplyr` pipes (e.g., see the [`group_by`](#groupby) section).

All methods accept one or more sets of `x` and `y` intervals, which must either be created using the [`read` methods](#reading data), or have `chrom`, `start` and `end` columns.

## Single set operations {.tabset}

These methods operate on a single set of intervals:

  - `bed_sort()`: order intervals 
  - `bed_cluster()`: Cluster (but don't merge) overlapping/nearby intervals.
  - `bed_complement()`: extract intervals *not* represented by an interval file.
  - `bed_merge()`: combine overlapping and nearby intervals into a single interval.
  - `bed_flank()`: Generate new flanking intervals
  - `bed_slop()`: Expand the size of input intervals
  - `bed_shift()`: Shift the coordinates of an input set, bounded by a genome
  
### Sort 

`bed_sort` orders intervals based on a specification. `is_sorted` asks whether a tbl is already sorted.

```{r sort}
x <- bed_random(genome)
is_sorted(x)

y <- bed_sort(x)
y

is_sorted(y)
```

### Cluster 

`bed_cluster` identifies clustered intervals based on a distance specification and assigns them a unique `.id`.

```{r cluster}
x <- bed_random(genome)
y <- bed_cluster(x, max_dist = 1000)
y
```

### Complement 

`bed_complement` identifies intervals in a genome that are not covered by an input.

```{r complement}
x <- bed_random(genome)

bed_complement(x, genome)
```

### Merge 

`bed_merge` identifies overlapping intervals and reports new merged ones. `is_merged` asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.

```{r merge}
n <- 1e6
x <- bed_random(genome, n = n)
is_merged(x)

# add some signal
x <- x %>% mutate(signal = runif(n))

bed_merge(x, maxs = max(signal))
```

### Flank

`bed_flank` creates new intervals that flank -- but do not contain -- the input intervals.

```{r flank, eval = FALSE}
bed_flank(x, genome, both = 100)
```

### Slop

`bed_slop` pads input intervals based on a specification

```{r slop, eval = FALSE}
bed_slop(x, genome, both = 100)
```

### Shift

`bed_shift` adjusts coordinates toward `start` or `end` by a defined `size`. Intervals created out of bounds are removed, or trimmed.

```{r shift}
bed_shift(x, genome, size = 100)
```

## Multiple set operations {.tabset}

These methods compare two sets of intervals:

  - `bed_intersect()`: find overlapping intervals
  - `bed_map()`: apply a function to selected columns for overlapping intervals
  - `bed_subtract()`: Remove intervals based on overlaps between two files
  - `bed_window()`: Find overlapping intervals within a window
  - `bed_closest()`: find the closest intervals independent of overlaps
  
### Intersection

`bed_intersect` is implemented using an [interval tree][16] in `Rcpp`. Column names in the result have `.x` and `.y` suffixes, and an `.overlap` column contains the size of the intersection (values of `0` indicate book-ended, or touching intervals). See the [benchmarks](#benchmarks) section for timing. Though `bed_intersect` is pretty fast already, we intend to further improve upon this by parallization with [`RcppParallel`][15].

```{r intersect}
# intersect two sets of 1e6 intervals from hg19
x <- bed_random(genome)
y <- bed_random(genome)

bed_intersect(x, y)

# A records with no overlaps (i.e., `-v`)
bed_intersect(x, y, invert = TRUE)
```

One can achieve behaviour similar to BEDtools by combining `bed_intersect` with `dplyr` tools.

```{r intersect_opts}
# `x` records with overlaps (i.e., `-wa`)
bed_intersect(x, y) %>% select(chrom, start = start.x, end = end.x)

# `y` records with overlaps (i.e., `-wb`)
bed_intersect(x, y) %>% select(chrom, start = start.y, end = end.y)

# Unique records in `x` (i.e., `-u`)
bed_intersect(x, y) %>% select(chrom, start = start.x, end = end.x) %>% unique()

# Count `y` intervals that intersect each `x` interval
bed_intersect(x, y) %>% group_by(chrom, start.x, end.x) %>% summarize(count = n())
```

### Map

`bed_map` maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs. 

```{r map}
bedfile <- system.file('extdata', 'genes.hg19.chr22.bed.gz', package = 'valr')
bgfile  <- system.file('extdata', 'hela.h3k4.chip.bg.gz', package = 'valr')

x <- read_bed(bedfile, n_fields = 6)
y <- read_bedgraph(bgfile)

bed_map(x, y, means = mean(value.y), sds = sd(value.y))
```

### Subtract

`bed_substract()` removes `x` intervals that intersect with `y`.

```{r subtract}
x <- bed_random(genome)
y <- bed_random(genome)

bed_subtract(x, y)
```

### Window

`bed_window()` identifies `y` intervals that intersect an expanded window of `x` intervals.

```{r window, warning = FALSE}
x <- bed_random(genome, n = 100)
y <- bed_random(genome, n = 100)

# a few intersections
bed_intersect(x, y)

# can be expanded by casting a wider net
bed_window(x, y, genome, both = 1e6)
```

### Closest

`bed_closest()` identifies `y` intervals that are closest to `x`.

```{r closest}
x <- bed_random(genome, n = 100)
y <- bed_random(genome, n = 100)

bed_closest(x, y)
```

## Randomzing intervals {.tabset}

`valr` provides methods for creating new random intervals or permutations of existing intervals:

  - `bed_random` generates random intervals from an input `genome`.
  - `bed_shuffle` shuffles coordinates given a set of input intervals.
  - Random sampling of input intervals is done with `dplyr`.

### Random

`bed_random` generates random intervals from an input genome. The numbers of intervals from each `chrom` are proporitional to each chrom `size`.

```{r random}
x <- bed_random(genome, n = 1e6, length = 1e3)
x

# numbers of sampled intervals are proportional to chrom size
group_by(x, chrom) %>% summarize(n = n()) %>% arrange(desc(n))
```

### Sample

Sampling can be done using `dplyr`:

```{r sample}
x <- bed_random(genome)

# sample by number
sample_n(x, 1e3, replace = FALSE)

# or fraction
sample_frac(x, 0.1, replace = FALSE)

# or sample intervals within groups
group_by(x, chrom) %>% sample_n(1)
```

### Shuffle

`bed_shuffle` shuffles input intervals. Interval sizes are equal in the input and output.

```{r shuffle, eval = FALSE}
y <- bed_shuffle(x)

any(x$start == y$start)

all(x$end - x$start == y$end - y$start)
```

## Other

### Grouping related intervals

The `group_by` operation from `dplyr` serves a similar purpose to BEDtools `groupby`, and can be combined with `dplyr::summarize` to calculate summary statistics from groups.

```{r group_by}
x <- bed_random(genome)
y <- bed_random(genome)

bed_intersect(x, y) %>%
  group_by(chrom) %>%
  summarize(overlap.total = sum(.overlap))
```

### Interval spacing

Interval spacing is easily computed using `dplyr`. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.

```{r spacing}
x <- bed_random(genome)
x <- bed_merge(x)

x %>%
  group_by(chrom) %>%
  mutate(.spacing = start - lag(end))
```

# Interval Statistics {.tabset}

`valr` provides several methods to assess statistical properties of interval sets including:

  - `bed_fisher()`: measure overlap significance of two sets of intervals.
  - `bed_absdist()`: XXX
  - `bed_reldist()`: quantify relative distances between query intervals and closest reference intervals
  - `bed_jaccard()`: quantify extent of overlap between two sets of intervals
  - `bed_projection()`: XXX

Several of these methods were described in the [Genometricorr][17] software package.

## Fisher's test

The Fisher's test assesses whether two sets of intervals are drawn from the same background genome.

```{r fisher}
x <- bed_random(genome, n = 100)
y <- bed_random(genome, n = 100)

bed_fisher(x, y, genome)
```

## Absolute distance

`bed_absdist()` - XXX

```{r absdist, eval = FALSE}

```

## Relative distance

`bed_reldist()` computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between `[0,0.5]`. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space ([17]). By default a summary with the frequency of each relative distance is reported.  

```{r reldist, eval = TRUE}
x <- bed_random(genome)
y <- bed_random(genome)

bed_reldist(x, y)
```

## Jaccard similarity

`bed_jaccard()` quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of `[0,1]` and is measured as:

$$
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
$$


```{r jaccard}
x <- bed_random(genome)
y <- bed_random(genome)

bed_jaccard(x, y)
```

## Projection

`bed_projection()` TBD

```{r projection, eval = FALSE}
```

# Benchmarks

```{r benchmarks}
# two sets of 1e6 random 1 kb intervals from hg19
x <- bed_random(genome)
y <- bed_random(genome)

library(microbenchmark)
microbenchmark(
  bed_random(genome),
  bed_closest(x, y),
  bed_intersect(x, y),
  bed_merge(x),
  bed_subtract(x, y),
  bed_complement(x, genome),
  times = 1,
  unit = 's'
)
```

[1]: http://bedtools.readthedocs.org/en/latest/
[2]: https://github.com/arq5x/chrom_sweep
[3]: https://github.com/hadley/dplyr
[4]: https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html
[5]: http://www.rcpp.org/
[6]: https://github.com/hadley/readr
[7]: https://pythonhosted.org/pybedtools/
[8]: http://bedops.readthedocs.org/en/latest/index.html
[9]: https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html
[10]: http://shiny.rstudio.com/
[11]: https://cran.r-project.org/web/packages/bedr/index.html
[12]: https://bioconductor.org/packages/release/bioc/html/IRanges.html
[13]: http://bedtools.readthedocs.io/en/latest/#performance
[14]: https://github.com/jayhesselberth/valr/tree/master/src
[15]: https://rcppcore.github.io/RcppParallel/
[16]: https://github.com/ekg/intervaltree/
[17]: http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529
