---
title: 'valr examples'
author: "Jay Hesselberth <jay.hesselberth@gmail.com>"
date: '`r Sys.Date()`'
output:
  rmarkdown::html_vignette:
  vignette: >
    %\VignetteIndexEntry{valr-examples}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---
  
```{r knitr_opts, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "img/valr-examples-",
  fig.height = 3,
  fig.align = "center",
  fig.width = 4
)
```

# Demonstrations 

## Creating intron intervals {#demo-intron}

Intron intervals can be identified by calculating:

1. Intervals that are covered by any gene annotation
1. Then, intervals that are not covered by exons.

```{r demo-intron}
# genome <- system.file('extdata', 'hg19.chrom.sizes.gz', package = 'valr')
# 
# exons <-
#   
# introns <- 
# introns
```

## Histone density and transcription start sites

This demonstration illustrates how to use `valr` tools to perform a "meta-analysis" of signals relative to genomic features. Here we to analyze the distribution of histone marks surrounding transcription start sites.

First we load libraries and relevant data.

```{r demo-tss, warning = FALSE, message = FALSE}
library(valr)
library(dplyr)

# `valr_example()` identifies the path of example files
bedfile <- valr_example('genes.hg19.chr22.bed.gz')
genomefile <- valr_example('hg19.chrom.sizes.gz')
bgfile  <- valr_example('hela.h3k4.chip.bg.gz')

genes <- read_bed(bedfile, n_fields = 6)
genome <- read_genome(genomefile)
y <- read_bedgraph(bgfile)
```

Then we generate 1 bp intervals to represent transcription start sites (TSSs). We focus on `+` strand genes, but `-` genes are easily accomodated by filtering them and using `bed_makewindows()` with `reverse`d window numbers.

```{r tss}
# generate 1 bp TSS intervals, `+` strand only
tss <- genes %>%
  filter(strand == '+') %>%
  mutate(end = start + 1)

# 1000 bp up and downstream
region_size <- 1000
# 50 bp windows
win_size <- 50

# add slop to the TSS, break into windows and add a group
x <- tss %>%
  bed_slop(genome, both = region_size) %>%
  bed_makewindows(genome, win_size) %>%
  group_by(win_id)

x
```

Note that the data are now grouped by `win_id`. This grouping is used below with `bed_map()` to caluclate summary statistics identified by mapping `y` signals onto the intervals in `x`. Note that `x` intervals with no `y` overlaps have values of `NA`, so `na.rm = TRUE` must be used during computation.

```{r map}
# map signals to TSS regions and calculate summary statistics.
res <- bed_map(x, y, win_sum = sum(value, na.rm = TRUE)) %>%
  group_by(win_id) %>%
  summarize(win_mean = mean(win_sum, na.rm = TRUE),
            win_sd = sd(win_sum, na.rm = TRUE))
```

Finally, these summary statistics are used to construct a plot that illustrates histone density surrounding TSSs.

```{r plot, warning = FALSE, message = FALSE, fig.align='center', fig.width=6}
library(ggplot2)

x_labels <- seq(-region_size, region_size, by = win_size * 5)
x_breaks <- seq(1, 41, by = 5)

sd_limits <- aes(ymax = win_mean + win_sd, ymin = win_mean - win_sd)

ggplot(res, aes(x = win_id, y = win_mean)) +
  geom_point() + geom_pointrange(sd_limits) + 
  scale_x_continuous(labels = x_labels, breaks = x_breaks) + 
  xlab('Position (bp from TSS)') + ylab('Signal') + 
  ggtitle('Human H3K4me3 signal near transcription starts sites') +
  theme_bw()
```
