<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interval statistics • valr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Interval statistics">
<meta property="og:description" content="valr">
<meta property="og:image" content="https://rnabioco.github.io/valr/logo.png">
<meta name="robots" content="noindex">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-54095998-3"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-54095998-3');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">valr</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="In-development version">0.6.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/valr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/benchmarks.html">valr Benchmarks</a>
    </li>
    <li>
      <a href="../articles/interval-stats.html">Interval statistics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rnabioco/valr/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="interval-stats_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="interval-stats_files/htmlwidgets-1.5.2/htmlwidgets.js"></script><script src="interval-stats_files/jquery-1.12.4/jquery.min.js"></script><link href="interval-stats_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="interval-stats_files/datatables-binding-0.15/datatables.js"></script><link href="interval-stats_files/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="interval-stats_files/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="interval-stats_files/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><link href="interval-stats_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="interval-stats_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Interval statistics</h1>
            
            <h4 class="date">October 07 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/valr/blob/master/vignettes/interval-stats.Rmd"><code>vignettes/interval-stats.Rmd</code></a></small>
      <div class="hidden name"><code>interval-stats.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p><code>valr</code> includes several functions for exploring statistical relationships between sets of intervals.</p>
<ul>
<li><p>Calculate significance of overlaps between sets of intervals with <code><a href="../reference/bed_fisher.html">bed_fisher()</a></code> and <code><a href="../reference/bed_projection.html">bed_projection()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between sets of intervals with <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>.</p></li>
</ul>
<p>In this vignette we explore the relationship between transcription start sites and repetitive elements in the human genome.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rnabioco/valr">valr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>

<span class="co"># load repeats and genes. Data in the valr package is restricted to chr22; the entire</span>
<span class="co"># files can be downloaded from UCSC.</span>
<span class="va">rpts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'hg19.rmsk.chr22.bed.gz'</span><span class="op">)</span>, n_fields <span class="op">=</span> <span class="fl">6</span><span class="op">)</span> 
<span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'hg19.refGene.chr22.bed.gz'</span><span class="op">)</span>, n_fields <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>

<span class="co"># load chrom sizes</span>
<span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_genome.html">read_genome</a></span><span class="op">(</span><span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'hg19.chrom.sizes.gz'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># create 1 bp intervals representing transcription start sites</span>
<span class="va">tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_tss.html">create_tss</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span>

<span class="va">tss</span>
<span class="co">#&gt; # A tibble: 1,267 x 6</span>
<span class="co">#&gt;    chrom    start      end name         score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;        &lt;chr&gt; &lt;chr&gt; </span>
<span class="co">#&gt;  1 chr22 16193008 16193009 NR_122113    0     -     </span>
<span class="co">#&gt;  2 chr22 16157078 16157079 NR_133911    0     +     </span>
<span class="co">#&gt;  3 chr22 16162065 16162066 NR_073459    0     +     </span>
<span class="co">#&gt;  4 chr22 16162065 16162066 NR_073460    0     +     </span>
<span class="co">#&gt;  5 chr22 16231288 16231289 NR_132385    0     -     </span>
<span class="co">#&gt;  6 chr22 16287936 16287937 NM_001136213 0     -     </span>
<span class="co">#&gt;  7 chr22 16274608 16274609 NR_046571    0     +     </span>
<span class="co">#&gt;  8 chr22 16449803 16449804 NM_001005239 0     -     </span>
<span class="co">#&gt;  9 chr22 17073699 17073700 NM_014406    0     -     </span>
<span class="co">#&gt; 10 chr22 17082800 17082801 NR_001591    0     +     </span>
<span class="co">#&gt; # … with 1,257 more rows</span></pre></div>
</div>
<div id="distance-metrics" class="section level2">
<h2 class="hasAnchor">
<a href="#distance-metrics" class="anchor"></a>Distance metrics</h2>
<p>First we define a function that takes <code>x</code> and <code>y</code> intervals and computes distance statistics (using <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>) for specified groups. The value of each statistic is assigned to a <code>.value</code> column.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">distance_stats</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">genome</span>, <span class="va">group_var</span>, <span class="va">type</span> <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">x</span>, <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html">sym</a></span><span class="op">(</span><span class="va">group_var</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span><span class="op">(</span>
       reldist <span class="op">=</span> <span class="fu"><a href="../reference/bed_reldist.html">bed_reldist</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">y</span>, detail <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>
         <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span>.value <span class="op">=</span> <span class="va">.reldist</span><span class="op">)</span>,
       absdist <span class="op">=</span> <span class="fu"><a href="../reference/bed_absdist.html">bed_absdist</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">y</span>, <span class="va">genome</span><span class="op">)</span> <span class="op">%&gt;%</span>
         <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span>.value <span class="op">=</span> <span class="va">.absdist</span><span class="op">)</span>
       <span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="va">name</span>, 
                        names_to <span class="op">=</span> <span class="st">'stat'</span>,
                        values_to <span class="op">=</span> <span class="st">'value'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="va">type</span><span class="op">)</span>
<span class="op">}</span> </pre></div>
<p>We use the <code>distance_stats()</code> function to apply the <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code> function to each group of data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">obs_stats</span> <span class="op">&lt;-</span> <span class="fu">distance_stats</span><span class="op">(</span><span class="va">rpts</span>, <span class="va">tss</span>, <span class="va">genome</span>, <span class="st">'name'</span>, <span class="st">'obs'</span><span class="op">)</span>
<span class="va">obs_stats</span>
<span class="co">#&gt; # A tibble: 2,106 x 4</span>
<span class="co">#&gt;    name       stat    value             type </span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;   &lt;list&gt;            &lt;chr&gt;</span>
<span class="co">#&gt;  1 (A)n       reldist &lt;tibble [27 × 1]&gt; obs  </span>
<span class="co">#&gt;  2 (A)n       absdist &lt;tibble [28 × 1]&gt; obs  </span>
<span class="co">#&gt;  3 (AAAAACA)n reldist &lt;tibble [1 × 1]&gt;  obs  </span>
<span class="co">#&gt;  4 (AAAAACA)n absdist &lt;tibble [1 × 1]&gt;  obs  </span>
<span class="co">#&gt;  5 (AAAAC)n   reldist &lt;tibble [6 × 1]&gt;  obs  </span>
<span class="co">#&gt;  6 (AAAAC)n   absdist &lt;tibble [7 × 1]&gt;  obs  </span>
<span class="co">#&gt;  7 (AAAAG)n   reldist &lt;tibble [2 × 1]&gt;  obs  </span>
<span class="co">#&gt;  8 (AAAAG)n   absdist &lt;tibble [2 × 1]&gt;  obs  </span>
<span class="co">#&gt;  9 (AAAAT)n   reldist &lt;tibble [3 × 1]&gt;  obs  </span>
<span class="co">#&gt; 10 (AAAAT)n   absdist &lt;tibble [4 × 1]&gt;  obs  </span>
<span class="co">#&gt; # … with 2,096 more rows</span></pre></div>
<p>And the same is done for a set of shuffled group of data. <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code> is used to shuffle coordinates of the repeats within each chromosome (i.e., the coordinates change, but the chromosome stays the same.)</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">shfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span><span class="op">(</span><span class="va">rpts</span>, <span class="va">genome</span>, within <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">shf_stats</span> <span class="op">&lt;-</span> <span class="fu">distance_stats</span><span class="op">(</span><span class="va">shfs</span>, <span class="va">tss</span>, <span class="va">genome</span>, <span class="st">'name'</span>, <span class="st">'shuf'</span><span class="op">)</span></pre></div>
<p>Now we can bind the observed and shuffled data together, and do some tidying to put the data into a format appropriate for a statistical test. This involves:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest()</a></code>ing the data frames</li>
<li>creating groups for each repeat (<code>name</code>), stat (<code>reldist</code> or <code>absdist</code>) and type (<code>obs</code> or <code>shf</code>)</li>
<li>adding unique surrogate row numbers for each group</li>
<li>using <code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">tidyr::pivot_wider()</a></code> to create two new <code>obs</code> and <code>shuf</code> columns</li>
<li>removing rows with <code>NA</code> values.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="va">obs_stats</span>, <span class="va">shf_stats</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">stat</span>, <span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>.id <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/ranking.html">row_number</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"type"</span>,
                     values_from <span class="op">=</span> <span class="st">".value"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html">na.omit</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">res</span>
<span class="co">#&gt; # A tibble: 16,810 x 5</span>
<span class="co">#&gt; # Groups:   name, stat [1,925]</span>
<span class="co">#&gt;    name  stat      .id   obs   shuf</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;chr&gt;   &lt;int&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 (A)n  reldist     1 0.363 0.235 </span>
<span class="co">#&gt;  2 (A)n  reldist     2 0.429 0.400 </span>
<span class="co">#&gt;  3 (A)n  reldist     3 0.246 0.282 </span>
<span class="co">#&gt;  4 (A)n  reldist     4 0.478 0.490 </span>
<span class="co">#&gt;  5 (A)n  reldist     5 0.260 0.250 </span>
<span class="co">#&gt;  6 (A)n  reldist     6 0.286 0.491 </span>
<span class="co">#&gt;  7 (A)n  reldist     7 0.498 0.145 </span>
<span class="co">#&gt;  8 (A)n  reldist     8 0.237 0.0576</span>
<span class="co">#&gt;  9 (A)n  reldist     9 0.314 0.0849</span>
<span class="co">#&gt; 10 (A)n  reldist    10 0.149 0.259 </span>
<span class="co">#&gt; # … with 16,800 more rows</span></pre></div>
<p>Now that the data are formatted, we can use the non-parametric <code><a href="https://rdrr.io/r/stats/ks.test.html">ks.test()</a></code> to determine whether there are significant differences between the observed and shuffled data for each group. <code><a href="https://rdrr.io/pkg/generics/man/tidy.html">broom::tidy()</a></code> is used to reformat the results of each test into a <code>tibble</code>, and the results of each test are <code>pivot</code>ed to into a <code>type</code> column for each test type.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://broom.tidymodels.org/">broom</a></span><span class="op">)</span>

<span class="va">pvals</span> <span class="op">&lt;-</span> <span class="va">res</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span><span class="op">(</span>twosided <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/generics/man/tidy.html">tidy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">obs</span>, <span class="va">.</span><span class="op">$</span><span class="va">shuf</span><span class="op">)</span><span class="op">)</span>,
                    less <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/generics/man/tidy.html">tidy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">obs</span>, <span class="va">.</span><span class="op">$</span><span class="va">shuf</span>, alternative <span class="op">=</span> <span class="st">'less'</span><span class="op">)</span><span class="op">)</span>,
                    greater <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/generics/man/tidy.html">tidy</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/ks.test.html">ks.test</a></span><span class="op">(</span><span class="va">.</span><span class="op">$</span><span class="va">obs</span>, <span class="va">.</span><span class="op">$</span><span class="va">shuf</span>, alternative <span class="op">=</span> <span class="st">'greater'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">stat</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"alt"</span>, values_to <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">name</span><span class="op">:</span><span class="va">p.value</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span></pre></div>
<p>Histgrams of the different stats help visulaize the distribution of p.values.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">pvals</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">stat</span> <span class="op">~</span> <span class="va">alt</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span></pre></div>
<p><img src="interval-stats_files/figure-html/pvalue_viz-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can also assess false discovery rates (q.values) using <code><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">pvals</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">pvals</span>, <span class="va">stat</span>, <span class="va">alt</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>q.value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">q.value</span><span class="op">)</span></pre></div>
<p>Finally we can visualize these results using <code><a href="https://ggplot2.tidyverse.org/reference/stat_ecdf.html">stat_ecdf()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">res_gather</span> <span class="op">&lt;-</span> <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="va">res</span>, 
                                  cols <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">name</span>, <span class="va">stat</span>, <span class="va">.id</span><span class="op">)</span>,
                                  names_to <span class="op">=</span> <span class="st">"type"</span>,
                                  values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span>

<span class="va">signif</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pvals</span>, <span class="fl">5</span><span class="op">)</span>

<span class="va">res_signif</span> <span class="op">&lt;-</span>
  <span class="va">signif</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">res_gather</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'name'</span>, <span class="st">'stat'</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">res_signif</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ecdf.html">stat_ecdf</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">stat</span> <span class="op">~</span> <span class="va">name</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html">theme_cowplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">'Set1'</span><span class="op">)</span></pre></div>
<p><img src="interval-stats_files/figure-html/ecfs-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="projection-test" class="section level2">
<h2 class="hasAnchor">
<a href="#projection-test" class="anchor"></a>Projection test</h2>
<p><code><a href="../reference/bed_projection.html">bed_projection()</a></code> is a statistical approach to assess the relationship between two intervals based on the binomial distribution. Here, we examine the distribution of repetitive elements within the promoters of coding or non-coding genes.</p>
<p>First we’ll extract 5 kb regions upstream of the transcription start sites to represent the promoter regions for coding and non-coding genes.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co"># create intervals 5kb upstream of tss representing promoters</span>
<span class="va">promoters</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="../reference/bed_flank.html">bed_flank</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">genome</span>, left <span class="op">=</span> <span class="fl">5000</span>, strand <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'NR_'</span>, <span class="va">name</span><span class="op">)</span>, <span class="st">'non-coding'</span>, <span class="st">'coding'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">chrom</span><span class="op">:</span><span class="va">strand</span><span class="op">)</span>
  
<span class="co"># select coding and non-coding promoters</span>
<span class="va">promoters_coding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">promoters</span>, <span class="va">name</span> <span class="op">==</span> <span class="st">'coding'</span><span class="op">)</span>
<span class="va">promoters_ncoding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">promoters</span>, <span class="va">name</span> <span class="op">==</span> <span class="st">'non-coding'</span><span class="op">)</span>

<span class="va">promoters_coding</span>
<span class="co">#&gt; # A tibble: 973 x 6</span>
<span class="co">#&gt;    chrom    start      end name   score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; </span>
<span class="co">#&gt;  1 chr22 16287937 16292937 coding 0     -     </span>
<span class="co">#&gt;  2 chr22 16449804 16454804 coding 0     -     </span>
<span class="co">#&gt;  3 chr22 17073700 17078700 coding 0     -     </span>
<span class="co">#&gt;  4 chr22 17302589 17307589 coding 0     -     </span>
<span class="co">#&gt;  5 chr22 17302589 17307589 coding 0     -     </span>
<span class="co">#&gt;  6 chr22 17489112 17494112 coding 0     -     </span>
<span class="co">#&gt;  7 chr22 17560848 17565848 coding 0     +     </span>
<span class="co">#&gt;  8 chr22 17560848 17565848 coding 0     +     </span>
<span class="co">#&gt;  9 chr22 17602213 17607213 coding 0     -     </span>
<span class="co">#&gt; 10 chr22 17602257 17607257 coding 0     -     </span>
<span class="co">#&gt; # … with 963 more rows</span>

<span class="va">promoters_ncoding</span>
<span class="co">#&gt; # A tibble: 294 x 6</span>
<span class="co">#&gt;    chrom    start      end name       score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;      &lt;chr&gt; &lt;chr&gt; </span>
<span class="co">#&gt;  1 chr22 16152078 16157078 non-coding 0     +     </span>
<span class="co">#&gt;  2 chr22 16157065 16162065 non-coding 0     +     </span>
<span class="co">#&gt;  3 chr22 16157065 16162065 non-coding 0     +     </span>
<span class="co">#&gt;  4 chr22 16193009 16198009 non-coding 0     -     </span>
<span class="co">#&gt;  5 chr22 16231289 16236289 non-coding 0     -     </span>
<span class="co">#&gt;  6 chr22 16269608 16274608 non-coding 0     +     </span>
<span class="co">#&gt;  7 chr22 17077800 17082800 non-coding 0     +     </span>
<span class="co">#&gt;  8 chr22 17156430 17161430 non-coding 0     -     </span>
<span class="co">#&gt;  9 chr22 17229328 17234328 non-coding 0     -     </span>
<span class="co">#&gt; 10 chr22 17303363 17308363 non-coding 0     +     </span>
<span class="co">#&gt; # … with 284 more rows</span></pre></div>
<p>Next we’ll apply the <code><a href="../reference/bed_projection.html">bed_projection()</a></code> test for each repeat class for both coding and non-coding regions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">

<span class="co"># function to apply bed_projection to groups</span>
<span class="va">projection_stats</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">genome</span>, <span class="va">group_var</span>, <span class="va">type</span> <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">x</span>, <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html">sym</a></span><span class="op">(</span><span class="va">group_var</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/do.html">do</a></span><span class="op">(</span>n_repeats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span>,
       projection <span class="op">=</span> <span class="fu"><a href="../reference/bed_projection.html">bed_projection</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">y</span>, <span class="va">genome</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="va">type</span><span class="op">)</span>
<span class="op">}</span> 

<span class="va">pvals_coding</span> <span class="op">&lt;-</span> <span class="fu">projection_stats</span><span class="op">(</span><span class="va">rpts</span>, <span class="va">promoters_coding</span>, <span class="va">genome</span>, <span class="st">'name'</span>, <span class="st">'coding'</span><span class="op">)</span>
<span class="va">pvals_ncoding</span> <span class="op">&lt;-</span> <span class="fu">projection_stats</span><span class="op">(</span><span class="va">rpts</span>, <span class="va">promoters_ncoding</span>, <span class="va">genome</span>, <span class="st">'name'</span>, <span class="st">'non_coding'</span><span class="op">)</span>

<span class="va">pvals</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="va">pvals_ncoding</span>, <span class="va">pvals_coding</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html">unnest</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">n_repeats</span>, <span class="va">projection</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">chrom</span><span class="op">)</span>

<span class="co"># filter for repeat classes with at least 10 intervals</span>
<span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">pvals</span>,
                <span class="va">n_repeats</span> <span class="op">&gt;</span> <span class="fl">10</span>,
                <span class="va">obs_exp_ratio</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span>

<span class="co"># adjust pvalues </span>
<span class="va">pvals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">pvals</span>, q.value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span>

<span class="va">pvals</span>
<span class="co">#&gt; # A tibble: 179 x 7</span>
<span class="co">#&gt;    name   n_repeats p.value obs_exp_ratio lower_tail type       q.value</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;int&gt;   &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt;  1 (A)n          28 0.00353         4.72  FALSE      non_coding   0.558</span>
<span class="co">#&gt;  2 (AT)n         48 0.298           0.917 FALSE      non_coding   1    </span>
<span class="co">#&gt;  3 (CA)n         31 0.156           1.42  FALSE      non_coding   1    </span>
<span class="co">#&gt;  4 (GT)n         42 0.247           1.05  FALSE      non_coding   1    </span>
<span class="co">#&gt;  5 (T)n          61 0.405           0.721 FALSE      non_coding   1    </span>
<span class="co">#&gt;  6 (TG)n         40 0.0622          2.20  FALSE      non_coding   1    </span>
<span class="co">#&gt;  7 A-rich        54 0.348           0.815 FALSE      non_coding   1    </span>
<span class="co">#&gt;  8 Alu           15 0.0446          2.93  FALSE      non_coding   1    </span>
<span class="co">#&gt;  9 AluJb        271 0.0225          1.79  FALSE      non_coding   1    </span>
<span class="co">#&gt; 10 AluJo        208 0.0216          1.90  FALSE      non_coding   1    </span>
<span class="co">#&gt; # … with 169 more rows</span></pre></div>
<p>The projection test is a two-tailed statistical test. A significant p-value indicates either enrichment or depletion of query intervals compared to the reference interval sets. A value of <code>lower_tail = TRUE</code> column indicates that the query intervals are depleted, whereas <code>lower_tail = FALSE</code> indicates that the query intervals are enriched.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT">DT</a></span><span class="op">)</span>

<span class="co"># find and show top 5 most significant repeats</span>
<span class="va">signif_tests</span> <span class="op">&lt;-</span>
  <span class="va">pvals</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">q.value</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="va">q.value</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span>

<span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span><span class="op">(</span><span class="va">signif_tests</span><span class="op">)</span></pre></div>
<div id="htmlwidget-f4c67885b2b7aa4d0a55" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-f4c67885b2b7aa4d0a55">{"x":{"filter":"none","data":[["1","2","3","4","5","6","7","8","9","10"],["AluSx1","AluSx3","AluJb","Charlie4z","AluSz","MER3","L1MC4a","L1ME1","G-rich","L1MC5"],[364,95,271,11,259,27,23,38,54,29],[1.13727959125676e-05,0.000142248310274207,0.00022198786170069,0.00122898000927452,0.00131822309756457,2.70117007751258e-05,0.000144755028319676,0.000203406678286999,0.000217198716728206,0.000456128135381095],[2.10614752188776,2.91411990455348,2.04310989618141,5.80786134823596,1.89110543256589,8.14944717831681,7.65339387181057,5.79039667933037,4.88966830699009,6.06993307074632],["FALSE","FALSE","FALSE","FALSE","FALSE","FALSE","FALSE","FALSE","FALSE","FALSE"],["coding","coding","coding","coding","coding","non_coding","non_coding","non_coding","non_coding","non_coding"],[0.00203573046834959,0.0251779509185346,0.0384039000742193,0.205239661548844,0.218825034195719,0.0048080827379724,0.025476884984263,0.0355961687002249,0.0377925767107079,0.0784540392855484]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>name<\/th>\n      <th>n_repeats<\/th>\n      <th>p.value<\/th>\n      <th>obs_exp_ratio<\/th>\n      <th>lower_tail<\/th>\n      <th>type<\/th>\n      <th>q.value<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Kent Riemondy, <a href="https://rockyrna.org">RNA Bioscience Initiative</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
