<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ valr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><link href="../deps/_Source%20Sans%20Pro-0.4.0/font.css" rel="stylesheet"><link href="../deps/_Source%20Code%20Pro-0.4.0/font.css" rel="stylesheet"><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://rnabioco.github.io/valr/logo.png"><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">valr</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.6.4.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/valr.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/interval-stats.html">Interval statistics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/valr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="data-types">Data types<a class="anchor" aria-label="anchor" href="#data-types"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      
    </div><div class="section level2">
      <h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="read_bed.html">read_bed()</a></code> <code><a href="read_bed.html">read_bed12()</a></code> <code><a href="read_bed.html">read_bedgraph()</a></code> <code><a href="read_bed.html">read_narrowpeak()</a></code> <code><a href="read_bed.html">read_broadpeak()</a></code> 
        </dt>
        <dd>Read BED and related files.</dd>
      </dl><dl><dt>
          
          <code><a href="read_bigwig.html">read_bigwig()</a></code> 
        </dt>
        <dd>Import and convert a bigwig file into a valr compatible tbl</dd>
      </dl><dl><dt>
          
          <code><a href="read_genome.html">read_genome()</a></code> 
        </dt>
        <dd>Read genome files.</dd>
      </dl><dl><dt>
          
          <code><a href="read_gtf.html">read_gtf()</a></code> 
        </dt>
        <dd>Import and convert a GTF/GFF file into a valr compatible bed tbl format</dd>
      </dl><dl><dt>
          
          <code><a href="read_vcf.html">read_vcf()</a></code> 
        </dt>
        <dd>Read a VCF file.</dd>
      </dl><dl><dt>
          
          <code><a href="db.html">db_ucsc()</a></code> <code><a href="db.html">db_ensembl()</a></code> 
        </dt>
        <dd>Fetch data from remote databases.</dd>
      </dl></div><div class="section level2">
      <h2 id="single-set-operations">Single set operations<a class="anchor" aria-label="anchor" href="#single-set-operations"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="bed_cluster.html">bed_cluster()</a></code> 
        </dt>
        <dd>Cluster neighboring intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_complement.html">bed_complement()</a></code> 
        </dt>
        <dd>Identify intervals in a genome not covered by a query.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_flank.html">bed_flank()</a></code> 
        </dt>
        <dd>Create flanking intervals from input intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_merge.html">bed_merge()</a></code> 
        </dt>
        <dd>Merge overlapping intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_partition.html">bed_partition()</a></code> 
        </dt>
        <dd>Partition intervals into elemental intervals</dd>
      </dl><dl><dt>
          
          <code><a href="bed_slop.html">bed_slop()</a></code> 
        </dt>
        <dd>Increase the size of input intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_shift.html">bed_shift()</a></code> 
        </dt>
        <dd>Adjust intervals by a fixed size.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_sort.html">bed_sort()</a></code> 
        </dt>
        <dd>Sort a set of intervals.</dd>
      </dl></div><div class="section level2">
      <h2 id="multiple-set-operations">Multiple set operations<a class="anchor" aria-label="anchor" href="#multiple-set-operations"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="bed_closest.html">bed_closest()</a></code> 
        </dt>
        <dd>Identify closest intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_coverage.html">bed_coverage()</a></code> 
        </dt>
        <dd>Compute coverage of intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_intersect.html">bed_intersect()</a></code> 
        </dt>
        <dd>Identify intersecting intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_map.html">bed_map()</a></code> <code><a href="bed_map.html">concat()</a></code> <code><a href="bed_map.html">values_unique()</a></code> <code><a href="bed_map.html">values()</a></code> 
        </dt>
        <dd>Calculate summaries from overlapping intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_subtract.html">bed_subtract()</a></code> 
        </dt>
        <dd>Subtract two sets of intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_window.html">bed_window()</a></code> 
        </dt>
        <dd>Identify intervals within a specified distance.</dd>
      </dl></div><div class="section level2">
      <h2 id="randomizing-intervals">Randomizing intervals<a class="anchor" aria-label="anchor" href="#randomizing-intervals"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="bed_random.html">bed_random()</a></code> 
        </dt>
        <dd>Generate randomly placed intervals on a genome.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_shuffle.html">bed_shuffle()</a></code> 
        </dt>
        <dd>Shuffle input intervals.</dd>
      </dl></div><div class="section level2">
      <h2 id="interval-statistics">Interval statistics<a class="anchor" aria-label="anchor" href="#interval-statistics"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="bed_absdist.html">bed_absdist()</a></code> 
        </dt>
        <dd>Compute absolute distances between intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_reldist.html">bed_reldist()</a></code> 
        </dt>
        <dd>Compute relative distances between intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_fisher.html">bed_fisher()</a></code> 
        </dt>
        <dd>Fisher's test to measure overlap between two sets of intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_jaccard.html">bed_jaccard()</a></code> 
        </dt>
        <dd>Calculate the Jaccard statistic for two sets of intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_projection.html">bed_projection()</a></code> 
        </dt>
        <dd>Projection test for query interval overlap.</dd>
      </dl></div><div class="section level2">
      <h2 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="create_introns.html">create_introns()</a></code> 
        </dt>
        <dd>Create intron features.</dd>
      </dl><dl><dt>
          
          <code><a href="create_tss.html">create_tss()</a></code> 
        </dt>
        <dd>Create transcription start site features.</dd>
      </dl><dl><dt>
          
          <code><a href="create_utrs3.html">create_utrs3()</a></code> 
        </dt>
        <dd>Create 3' UTR features.</dd>
      </dl><dl><dt>
          
          <code><a href="create_utrs5.html">create_utrs5()</a></code> 
        </dt>
        <dd>Create 5' UTR features.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_makewindows.html">bed_makewindows()</a></code> 
        </dt>
        <dd>Divide intervals into new sub-intervals ("windows").</dd>
      </dl><dl><dt>
          
          <code><a href="bed12_to_exons.html">bed12_to_exons()</a></code> 
        </dt>
        <dd>Convert BED12 to individual exons in BED6.</dd>
      </dl><dl><dt>
          
          <code><a href="bed_glyph.html">bed_glyph()</a></code> 
        </dt>
        <dd>Create example glyphs for valr functions.</dd>
      </dl><dl><dt>
          
          <code><a href="bound_intervals.html">bound_intervals()</a></code> 
        </dt>
        <dd>Select intervals bounded by a genome.</dd>
      </dl><dl><dt>
          
          <code><a href="flip_strands.html">flip_strands()</a></code> 
        </dt>
        <dd>Flip strands in intervals.</dd>
      </dl><dl><dt>
          
          <code><a href="gr_to_bed.html">gr_to_bed()</a></code> 
        </dt>
        <dd>Convert Granges to bed tibble</dd>
      </dl><dl><dt>
          
          <code><a href="interval_spacing.html">interval_spacing()</a></code> 
        </dt>
        <dd>Calculate interval spacing.</dd>
      </dl><dl><dt>
          
          <code><a href="ivl_df.html">check_interval()</a></code> <code><a href="ivl_df.html">check_genome()</a></code> 
        </dt>
        <dd>Bed-like data.frame requirements for valr functions</dd>
      </dl><dl><dt>
          
          <code><a href="valr.html">valr</a></code> 
        </dt>
        <dd>valr: genome interval arithmetic in R</dd>
      </dl><dl><dt>
          
          <code><a href="valr_example.html">valr_example()</a></code> 
        </dt>
        <dd>Provide working directory for valr example files.</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

