<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Genome Interval Arithmetic in R • valr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/valr.html">Overview</a>
    </li>
    <li>
      <a href="articles/interval-stats.html">Interval Statistics</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/rnabioco/valr/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    

    
    
<div class="contents">
<div id="valr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#valr" class="anchor"></a>valr <img src="reference/figures/logo.png" align="right">
</h1></div>

<p><strong><code>valr</code> provides tools to read and manipulate genome intervals and signals</strong>, similar to the <a href="http://bedtools.readthedocs.org/en/latest/"><code>BEDtools</code></a> suite. <code>valr</code> enables analysis in the R/RStudio environment, leveraging modern R tools in the <a href="http://tidyverse.org"><code>tidyverse</code></a> for a terse, expressive syntax. Compute-intensive algorithms are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a>/C++, and many methods take advantage of the speed and grouping capability provided by <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a>.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>The latest stable version can be installed from CRAN:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">'valr'</span>)</code></pre></div>
<p>The latest development version can be installed from github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages("devtools")</span>
devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">'rnabioco/valr'</span>)</code></pre></div>
</div>
<div id="why-valr" class="section level2">
<h2 class="hasAnchor">
<a href="#why-valr" class="anchor"></a>Why <code>valr</code>?</h2>
<p><strong>Why another tool set for interval manipulations?</strong> Based on our experience teaching genome analysis, we were motivated to develop interval arithmetic software that faciliates genome analysis in a single environment (RStudio), eliminating the need to master both command-line and exploratory analysis tools.</p>
<p><strong>Note:</strong> <code>valr</code> can currently be used for analysis of pre-processed data in BED and related formats. We plan to support BAM and VCF files soon via tabix indexes.</p>
<div id="familiar-tools-all-within-r" class="section level3">
<h3 class="hasAnchor">
<a href="#familiar-tools-all-within-r" class="anchor"></a>Familiar tools, all within R</h3>
<p>The functions in <code>valr</code> have similar names to their <code>BEDtools</code> counterparts, and so will be familiar to users coming from the <code>BEDtools</code> suite. Similar to <a href="https://daler.github.io/pybedtools/#why-pybedtools"><code>pybedtools</code></a>, <code>valr</code> has a terse syntax:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(dplyr)

snps &lt;-<span class="st"> </span><span class="kw"><a href="reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.snps147.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>)

<span class="co"># find snps in intergenic regions</span>
intergenic &lt;-<span class="st"> </span><span class="kw"><a href="reference/bed_subtract.html">bed_subtract</a></span>(snps, genes)
<span class="co"># find distance from intergenic snps to nearest gene</span>
nearby &lt;-<span class="st"> </span><span class="kw"><a href="reference/bed_closest.html">bed_closest</a></span>(intergenic, genes)

nearby <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="kw">starts_with</span>(<span class="st">'name'</span>), .overlap, .dist) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">abs</span>(.dist) <span class="op">&lt;</span><span class="st"> </span><span class="dv">5000</span>)</code></pre></div>
</div>
<div id="visual-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#visual-documentation" class="anchor"></a>Visual documentation</h3>
<p><code>valr</code> includes helpful glyphs to illustrate the results of specific operations, similar to those found in the <code>BEDtools</code> documentation. For example, <code><a href="reference/bed_glyph.html">bed_glyph()</a></code> illustrates the result of intersecting <code>x</code> and <code>y</code> intervals with <code><a href="reference/bed_intersect.html">bed_intersect()</a></code>:</p>
<p><img src="img/README-intersect_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
<div id="reproducible-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#reproducible-reports" class="anchor"></a>Reproducible reports</h3>
<p><code>valr</code> can be used in RMarkdown documents to generate reproducible work-flows for data processing. Because computations in <code>valr</code> are fast, it can be for exploratory analysis with <code>RMarkdown</code>, and for interactive analysis using <code>shiny</code>.</p>
</div>
<div id="remote-databases" class="section level3">
<h3 class="hasAnchor">
<a href="#remote-databases" class="anchor"></a>Remote databases</h3>
<p>Remote databases can be accessed with <code><a href="reference/db.html">db_ucsc()</a></code> (to access the UCSC Browser) and <code><a href="reference/db.html">db_ensembl()</a></code> (to access Ensembl databases).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># access the `refGene` tbl on the `hg38` assembly</span>
ucsc &lt;-<span class="st"> </span><span class="kw"><a href="reference/db.html">db_ucsc</a></span>(<span class="st">'hg38'</span>)
<span class="kw">tbl</span>(ucsc, <span class="st">'refGene'</span>)</code></pre></div>
</div>
</div>
<div id="api" class="section level2">
<h2 class="hasAnchor">
<a href="#api" class="anchor"></a>API</h2>
<p>Function names are similar to their their <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> counterparts, with some additions.</p>
<div id="data-types" class="section level3">
<h3 class="hasAnchor">
<a href="#data-types" class="anchor"></a>Data types</h3>
<ul>
<li>
<code><a href="reference/tbl_interval.html">tbl_interval()</a></code> and <code><a href="reference/tbl_genome.html">tbl_genome()</a></code> wrap tibbles and enforce strict column naming. <code><a href="reference/trbl_interval.html">trbl_interval()</a></code> and <code><a href="reference/trbl_genome.html">trbl_genome()</a></code> are constructors that take <code>tibble::tribble()</code> formatting.</li>
</ul>
</div>
<div id="reading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-data" class="anchor"></a>Reading data</h3>
<ul>
<li><p>BED and related files are read with <code><a href="reference/read_bed.html">read_bed()</a></code>, <code><a href="reference/read_bed.html">read_bed12()</a></code>, <code><a href="reference/read_bed.html">read_bedgraph()</a></code>, <code><a href="reference/read_bed.html">read_narrowpeak()</a></code> and <code><a href="reference/read_bed.html">read_broadpeak()</a></code>.</p></li>
<li><p>Genome files containing chromosome name and size information are loaded with <code><a href="reference/read_genome.html">read_genome()</a></code>.</p></li>
<li><p>VCF files are loaded with <code><a href="reference/read_vcf.html">read_vcf()</a></code>.</p></li>
<li><p>Remote databases can be accessed with <code><a href="reference/db.html">db_ucsc()</a></code> and <code><a href="reference/db.html">db_ensembl()</a></code>.</p></li>
</ul>
</div>
<div id="transforming-single-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#transforming-single-interval-sets" class="anchor"></a>Transforming single interval sets</h3>
<ul>
<li><p>Interval coordinates are adjusted with <code><a href="reference/bed_slop.html">bed_slop()</a></code> and <code><a href="reference/bed_shift.html">bed_shift()</a></code>, and new flanking intervals are created with <code><a href="reference/bed_flank.html">bed_flank()</a></code>.</p></li>
<li><p>Nearby intervals are combined with <code><a href="reference/bed_merge.html">bed_merge()</a></code> and identified (but not merged) with <code><a href="reference/bed_cluster.html">bed_cluster()</a></code>.</p></li>
<li><p>Intervals not covered by a query are created with <code><a href="reference/bed_complement.html">bed_complement()</a></code>.</p></li>
<li><p>Intervals can be ordered with <code>dplyr::arrange()</code>.</p></li>
</ul>
</div>
<div id="comparing-multiple-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-multiple-interval-sets" class="anchor"></a>Comparing multiple interval sets</h3>
<ul>
<li><p>Find overlaps between two sets of intervals with <code><a href="reference/bed_intersect.html">bed_intersect()</a></code>.</p></li>
<li><p>Apply functions to selected columns for overlapping intervals with <code><a href="reference/bed_map.html">bed_map()</a></code>.</p></li>
<li><p>Remove intervals based on overlaps between two files with <code><a href="reference/bed_subtract.html">bed_subtract()</a></code>.</p></li>
<li><p>Find overlapping intervals within a window with <code><a href="reference/bed_window.html">bed_window()</a></code>.</p></li>
<li><p>Find the closest intervals independent of overlaps with <code><a href="reference/bed_closest.html">bed_closest()</a></code>.</p></li>
</ul>
</div>
<div id="randomizing-intervals" class="section level3">
<h3 class="hasAnchor">
<a href="#randomizing-intervals" class="anchor"></a>Randomizing intervals</h3>
<ul>
<li><p>Generate random intervals from an input genome with <code><a href="reference/bed_random.html">bed_random()</a></code>.</p></li>
<li><p>Shuffle the coordinates of input intervals with <code><a href="reference/bed_shuffle.html">bed_shuffle()</a></code>.</p></li>
<li><p>Random sampling of input intervals is done with the <code>sample_</code> function family in <code>dplyr</code>.</p></li>
</ul>
</div>
<div id="interval-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-statistics" class="anchor"></a>Interval statistics</h3>
<ul>
<li><p>Calculate significance of overlaps between two sets of intervals with <code><a href="reference/bed_fisher.html">bed_fisher()</a></code> and <code><a href="reference/bed_projection.html">bed_projection()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between two sets of intervals with <code><a href="reference/bed_jaccard.html">bed_jaccard()</a></code>.</p></li>
</ul>
</div>
<div id="utilities" class="section level3">
<h3 class="hasAnchor">
<a href="#utilities" class="anchor"></a>Utilities</h3>
<ul>
<li><p>Create features from BED12 files with <code><a href="reference/create_introns.html">create_introns()</a></code>, <code><a href="reference/create_utrs5.html">create_utrs5()</a></code>, and <code><a href="reference/create_utrs3.html">create_utrs3()</a></code>.</p></li>
<li><p>Visualize the actions of valr functions with <code><a href="reference/bed_glyph.html">bed_glyph()</a></code>.</p></li>
<li><p>Constrain intervals to a genome reference with <code><a href="reference/bound_intervals.html">bound_intervals()</a></code>.</p></li>
<li><p>Subdivide intervals with <code><a href="reference/bed_makewindows.html">bed_makewindows()</a></code>.</p></li>
<li><p>Convert BED12 to BED6 format with <code><a href="reference/bed12_to_exons.html">bed12_to_exons()</a></code>.</p></li>
<li><p>Calculate spacing between intervals with <code><a href="reference/interval_spacing.html">interval_spacing()</a></code>.</p></li>
<li><p>Access remote databases with <code><a href="reference/db.html">db_ucsc()</a></code> and <code><a href="reference/db.html">db_ensembl()</a></code>.</p></li>
</ul>
</div>
</div>
<div id="related-work" class="section level2">
<h2 class="hasAnchor">
<a href="#related-work" class="anchor"></a>Related work</h2>
<ul>
<li><p>Command-line tools <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> and <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://CRAN.R-project.org/package=bedr">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a> and <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cran.r-project.org/package=valr">https://​cran.r-project.org/​package=valr</a>
</li>
<li>Browse source code at <br><a href="http://github.com/rnabioco/valr/">http://​github.com/​rnabioco/​valr/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/rnabioco/valr/issues">https://​github.com/​rnabioco/​valr/​issues</a>
</li>
<li>External data sets at <br><a href="http://github.com/rnabioco/valrdata">http://​github.com/​rnabioco/​valrdata</a>
</li>
</ul>
<h2>License</h2>
<p><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE">LICENSE</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Jay Hesselberth <br><small class="roles"> Author, maintainer </small> </li>
<li>Kent Riemondy <br><small class="roles"> Author </small> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/rnabioco/valr"><img src="https://travis-ci.org/rnabioco/valr.svg?branch=master" alt="Build Status"></a></li>
<li><a href="https://ci.appveyor.com/project/jayhesselberth/valr"><img src="https://ci.appveyor.com/api/projects/status/github/rnabioco/valr?branch=master&amp;svg=true" alt="AppVeyor Build Status"></a></li>
<li><a href="https://codecov.io/github/rnabioco/valr?branch=master"><img src="https://img.shields.io/codecov/c/github/rnabioco/valr/master.svg" alt="Coverage Status"></a></li>
<li><a href="http://www.r-pkg.org/pkg/valr"><img src="http://www.r-pkg.org/badges/version/valr"></a></li>
<li><a href="https://CRAN.R-project.org/package=valr"><img src="http://cranlogs.r-pkg.org/badges/valr?color=FFD700"></a></li>
</ul>
</div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jay Hesselberth, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
