<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Home. valr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>

<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container">
      <header>
      
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="articles/valr-overview.html">Overview</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header>

      <div class="row">
  <div class="col-md-9">
    
<h1 id="valr">valr</h1>
<p><a href="https://travis-ci.org/jayhesselberth/valr"><img src="https://travis-ci.org/jayhesselberth/valr.svg?branch=master" alt="Build Status"></a> <a href="https://codecov.io/github/jayhesselberth/valr?branch=master"><img src="https://img.shields.io/codecov/c/github/jayhesselberth/valr/master.svg" alt="Coverage Status"></a> <a href="http://www.repostatus.org/#wip"><img src="http://www.repostatus.org/badges/0.1.0/wip.svg" alt="Project Status: Wip - Initial development is in progress, but there has not yet been a stable, usable release suitable for the public."></a> <a href="http://www.r-pkg.org/pkg/valr"><img src="http://www.r-pkg.org/badges/version/valr"></a></p>
<p><code>valr</code> provides tools for fast, exploratory analysis of large biological data sets <strong>within R/RStudio</strong>. <code>valr</code> was developed from a desire to teach the powerful concepts of genome interval arithmetic, but without the cumbersome back-and-forth between command-line and exploratory analysis tools.</p>
<p>Key parts of <code>valr</code> are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a> for speed. Moreover, <code>valr</code> integrates with <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a> and the <code>magrittr</code> pipe operator (<code>%&gt;%</code>) for an expressive syntax and can be used in reproducible reports written in <a href="http://rmarkdown.rstudio.com/"><code>RMarkdown</code></a>.</p>
<p>See the <a href="http://jayhesselberth.github.io/valr"><code>valr</code> documentation</a> for a complete API reference and examples.</p>
<h2 id="installation">Installation</h2>
<p><code>valr</code> can be installed from github:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">'jayhesselberth/valr'</span>)</code></pre></div>
<p><strong>Note</strong> that <code>valr</code> requires a full-featured C++11 compiler (<code>gcc&gt;=6.0</code> or <code>clang++</code>)</p>
<h2 id="api">API</h2>
<p>Function names are similar to their their <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> counterparts, with some additions.</p>
<h3 id="reading-data">Reading data</h3>
<ul>
<li><p>BED and related files are read with <code><a href="reference/read_bed.html">read_bed()</a></code>, <code><a href="reference/read_bed.html">read_bed12()</a></code>, <code><a href="reference/read_bed.html">read_bedgraph()</a></code>, <code><a href="reference/read_bed.html">read_narrowpeak()</a></code> and <code><a href="reference/read_bed.html">read_broadpeak()</a></code>.</p></li>
<li><p>Genome files containing chromosome name and size information are loaded with <code><a href="reference/read_genome.html">read_genome()</a></code>.</p></li>
<li><p>VCF files are loaded with <code><a href="reference/read_vcf.html">read_vcf()</a></code>.</p></li>
</ul>
<h3 id="transforming-single-interval-sets">Transforming single interval sets</h3>
<ul>
<li><p>Intervals are ordered with <code><a href="reference/bed_sort.html">bed_sort()</a></code>.</p></li>
<li><p>Interval coordinates are adjusted with <code><a href="reference/bed_slop.html">bed_slop()</a></code> and <code><a href="reference/bed_shift.html">bed_shift()</a></code>, and new flanking intervals are created with <code><a href="reference/bed_flank.html">bed_flank()</a></code>.</p></li>
<li><p>Nearby intervals are combined with <code><a href="reference/bed_merge.html">bed_merge()</a></code> and identified (but not merged) with <code><a href="reference/bed_cluster.html">bed_cluster()</a></code>.</p></li>
<li><p>Intervals not covered by a query are created with <code><a href="reference/bed_complement.html">bed_complement()</a></code>.</p></li>
</ul>
<h3 id="comparing-multiple-interval-sets">Comparing multiple interval sets</h3>
<ul>
<li><p>Find overlaps between two sets of intervals with <code><a href="reference/bed_intersect.html">bed_intersect()</a></code>.</p></li>
<li><p>Apply functions to selected columns for overlapping intervals with <code><a href="reference/bed_map.html">bed_map()</a></code>.</p></li>
<li><p>Remove intervals based on overlaps between two files with <code><a href="reference/bed_subtract.html">bed_subtract()</a></code>.</p></li>
<li><p>Find overlapping intervals within a window with <code><a href="reference/bed_window.html">bed_window()</a></code>.</p></li>
<li><p>Find the closest intervals independent of overlaps with <code><a href="reference/bed_closest.html">bed_closest()</a></code>.</p></li>
</ul>
<h3 id="randomizing-intervals">Randomizing intervals</h3>
<ul>
<li><p>Generate random intervals from an input genome with <code><a href="reference/bed_random.html">bed_random()</a></code>.</p></li>
<li><p>Shuffle the coordinates of input intervals with <code><a href="reference/bed_shuffle.html">bed_shuffle()</a></code>.</p></li>
<li><p>Random sampling of input intervals is done with the <code>sample_</code> function family in <code>dplyr</code>.</p></li>
</ul>
<h3 id="interval-statistics">Interval statistics</h3>
<ul>
<li><p>Quantify overlaps between two sets of intervals with <code><a href="reference/bed_fisher.html">bed_fisher()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between two sets of intervals with <code><a href="reference/bed_jaccard.html">bed_jaccard()</a></code>.</p></li>
</ul>
<h2 id="related-work">Related work</h2>
<ul>
<li><p>Command-line tools <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> and <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://cran.r-project.org/web/packages/bedr/index.html">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a> and <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>

  </div>
</div>

      <footer>
      <p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer>
   </div>

  </body>
</html>
