<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • valr</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Atkinson_Hyperlegible-0.4.9/font.css" rel="stylesheet"><link href="../deps/Fira_Code-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="https://rnabioco.github.io/valr/logo.png"><meta name="robots" content="noindex"><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">valr</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.8.2.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/valr.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/interval-stats.html">Interval statistics</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/rnabioco/valr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/rnabioco/valr/blob/main/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="development version" id="valr-development-version">valr (development version)<a class="anchor" aria-label="anchor" href="#valr-development-version"></a></h2>
<ul><li><p><code><a href="../reference/read_bigwig.html">read_bigwig()</a></code> now uses cpp11bigwig on CRAN. The <code>set_strand</code> param was removed to be more consistent with expected bigWig contents.</p></li>
<li><p><code><a href="../reference/read_gtf.html">read_gtf()</a></code> was deprecated. The rtracklayer package used for this functionality is no longer a dependency of valr due to errors from CRAN AddressSantizer checks of the UCSC c-library code vendored in rtracklayer.</p></li>
<li><p>valr now depends on R &gt;= 4.0.0.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.2" id="valr-082">valr 0.8.2<a class="anchor" aria-label="anchor" href="#valr-082"></a></h2><p class="text-muted">CRAN release: 2024-08-30</p>
<ul><li><p>Address NOTE on CRAN about Rd link targets.</p></li>
<li><p>Change maintainer email address.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.1" id="valr-081">valr 0.8.1<a class="anchor" aria-label="anchor" href="#valr-081"></a></h2><p class="text-muted">CRAN release: 2024-04-22</p>
<ul><li>Make vdiffr dependency optional during package testing.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.0" id="valr-080">valr 0.8.0<a class="anchor" aria-label="anchor" href="#valr-080"></a></h2><p class="text-muted">CRAN release: 2024-04-04</p>
<ul><li>Added <code><a href="../reference/bed_genomecov.html">bed_genomecov()</a></code> to compute interval coverage across a genome.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.0" id="valr-070">valr 0.7.0<a class="anchor" aria-label="anchor" href="#valr-070"></a></h2><p class="text-muted">CRAN release: 2023-09-18</p>
<ul><li>
<code>read_bed</code> and related functions now automatically calculate the fields. Use of <code>n_fields</code> was deprecated.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.8" id="valr-068">valr 0.6.8<a class="anchor" aria-label="anchor" href="#valr-068"></a></h2><p class="text-muted">CRAN release: 2023-05-16</p>
<ul><li><p><code><a href="../reference/bed_closest.html">bed_closest()</a></code> now reports all x intervals, even when there are no closest y intervals (e.g. when there is no matching chromosome in y intervals). These intervals are returned populated with <code>NA</code> for <code>.overlap</code>, <code>.dist</code> and y interval locations.</p></li>
<li><p>Reimplemented <code><a href="../reference/bed_closest.html">bed_closest()</a></code> to use binary search rather than an interval tree search. The closest y interval can be missed with the previous search strategy in high depth interval trees.</p></li>
<li><p>Fix off by one error when using <code>max_dist</code> argument in <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code> (<a href="https://github.com/rnabioco/valr/issues/401" class="external-link">#401</a>).</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.7" id="valr-067">valr 0.6.7<a class="anchor" aria-label="anchor" href="#valr-067"></a></h2><p class="text-muted">CRAN release: 2023-02-18</p>
<ul><li><p>Removed <code>SystemRequirements</code> from DESCRIPTION to eliminate a NOTE on CRAN.</p></li>
<li><p><code><a href="../reference/bed_coverage.html">bed_coverage()</a></code> now reports intervals from <code>x</code> with no matching group in <code>y</code> (<a href="https://github.com/rnabioco/valr/issues/395" class="external-link">#395</a>).</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.6" id="valr-066">valr 0.6.6<a class="anchor" aria-label="anchor" href="#valr-066"></a></h2><p class="text-muted">CRAN release: 2022-10-11</p>
<ul><li><p>Updated <a href="https://github.com/ekg/intervaltree" class="external-link">intervalTree</a> header to commit f0c4046</p></li>
<li><p>valr now uses <a href="https://cli.r-lib.org/index.html" class="external-link">cli</a> for more consistent errors and messages during interactive use.</p></li>
<li><p>deprecated <code>genome</code> argument to <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> was removed.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.5" id="valr-065">valr 0.6.5<a class="anchor" aria-label="anchor" href="#valr-065"></a></h2><p class="text-muted">CRAN release: 2022-08-19</p>
<ul><li>Handle <code>max_dist</code> for first intervals in <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code> (<a href="https://github.com/rnabioco/valr/issues/388" class="external-link">#388</a>)</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.4" id="valr-064">valr 0.6.4<a class="anchor" aria-label="anchor" href="#valr-064"></a></h2><p class="text-muted">CRAN release: 2021-12-08</p>
<ul><li><p>Fixed intron score numbering error in <code>create_introns</code> (<a href="https://github.com/rnabioco/valr/issues/377" class="external-link">#377</a> <a href="https://github.com/sheridar" class="external-link">@sheridar</a>)</p></li>
<li><p>Fixed bug in handling of list inputs for <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>(<a href="https://github.com/rnabioco/valr/issues/380" class="external-link">#380</a> <a href="https://github.com/sheridar" class="external-link">@sheridar</a>)</p></li>
<li><p>Added <code>read_bigwig</code> and <code>read_gtf</code> functions to import data into valr compatible tibbles (<a href="https://github.com/rnabioco/valr/issues/379" class="external-link">#379</a>)</p></li>
<li><p>Kent Riemondy is now maintainer.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.3" id="valr-063">valr 0.6.3<a class="anchor" aria-label="anchor" href="#valr-063"></a></h2><p class="text-muted">CRAN release: 2021-05-15</p>
<ul><li>Update to prepare for readr 2.0.0</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.2" id="valr-062">valr 0.6.2<a class="anchor" aria-label="anchor" href="#valr-062"></a></h2><p class="text-muted">CRAN release: 2020-10-07</p>
<div class="section level3">
<h3 id="minor-changes-0-6-2">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-6-2"></a></h3>
<ul><li><p><code>RMariaDB</code> has replaced the deprecated <code>RMySQL</code> package as the database backend.</p></li>
<li><p>valr now imports Rcpp, which should have always been the case, but was masked by its Import by readr, which recently dropped use of Rcpp.</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.1" id="valr-061">valr 0.6.1<a class="anchor" aria-label="anchor" href="#valr-061"></a></h2><p class="text-muted">CRAN release: 2020-05-08</p>
<div class="section level3">
<h3 id="bug-fixes-0-6-1">Bug Fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-6-1"></a></h3>
<ul><li>Fixed rchk unprotect error (<a href="https://github.com/rnabioco/valr/issues/365" class="external-link">#365</a>)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.0" id="valr-060">valr 0.6.0<a class="anchor" aria-label="anchor" href="#valr-060"></a></h2><p class="text-muted">CRAN release: 2020-05-04</p>
<div class="section level3">
<h3 id="major-changes-0-6-0">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-6-0"></a></h3>
<ul><li><p><code>trbl_interval()</code> and <code>trbl_genome()</code> custom <code>tibble</code> subclasses have been deemed unnecessary and have been removed from the package.</p></li>
<li><p>coercing <code>GRanges</code> to a <code>valr</code> compatible data.frame now uses the <code><a href="../reference/gr_to_bed.html">gr_to_bed()</a></code> function rather than <code>as.trbl_interal()</code> methods.</p></li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-6-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-6-0"></a></h3>
<ul><li><p>dplyr version &lt; 0.8.0 is no longer supported due to unnecessary code bloat and challenges with handling multiple grouping structures (<a href="https://github.com/rnabioco/valr/issues/359" class="external-link">#359</a>).</p></li>
<li><p>The <code>sort_by</code> argument of <code><a href="../reference/bed_random.html">bed_random()</a></code> has been changed to <code>sorted</code>, and will now by default use <code><a href="../reference/bed_sort.html">bed_sort()</a></code> to sort the output, rather than rely on naming the sorting columns. Sorting can be suppressed by using <code>sorted = FALSE</code>.</p></li>
<li><p><code><a href="../reference/bed_sort.html">bed_sort()</a></code> now uses base R sorting with the <code>radix</code> method for increased speed. (<a href="https://github.com/rnabioco/valr/issues/353" class="external-link">#353</a>)</p></li>
<li><p><code>tbls</code> processed by <code><a href="../reference/bed_merge.html">bed_merge()</a></code>or <code><a href="../reference/bed_sort.html">bed_sort()</a></code> no longer store either <code>merged</code> or <code>sorted</code> as attributes, due to these attributes being rarely checked in the codebase and potential sources of unexpected behavior.</p></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-6-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-6-0"></a></h3>
<ul><li><p>Fixed <code><a href="../reference/bed_closest.html">bed_closest()</a></code> to prevent erroneous intervals being reported when adjacent closest intervals are present in the <code>y</code> table. (<a href="https://github.com/rnabioco/valr/issues/348" class="external-link">#348</a>)</p></li>
<li><p>Factor columns that are not used for grouping are returned as factors rather than inappropriately being coerced to integer vectors (<a href="https://github.com/rnabioco/valr/issues/360" class="external-link">#360</a>)</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.0" id="valr-050">valr 0.5.0<a class="anchor" aria-label="anchor" href="#valr-050"></a></h2><p class="text-muted">CRAN release: 2019-01-03</p>
<div class="section level3">
<h3 id="major-changes-0-5-0">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-5-0"></a></h3>
<ul><li>Internal <code>Rcpp</code> functions have been reorganized to remove all dependencies on <code>dplyr</code> C++ functions.</li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-5-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-5-0"></a></h3>
<ul><li><p>Due to internal refactoring of Rcpp functions, only data.frames containing Numeric, Logical, Integer, Character, and List column types are supported. Columns containing Raw, Complex, or other R classes are not supported and will issue an error.</p></li>
<li><p>Factors are now disallowed from grouping variables in multiset operations to avoid sort order discrepancies, and compatibility with factor handling in <code>dplyr</code> v.0.8.0. Factors will now be internally type-converted to character and a warning is issued.</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.2" id="valr-042">valr 0.4.2<a class="anchor" aria-label="anchor" href="#valr-042"></a></h2><p class="text-muted">CRAN release: 2018-11-17</p>
<div class="section level3">
<h3 id="bug-fixes-0-4-2">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-4-2"></a></h3>
<ul><li>Changed the behavior of <code>as.tbl_interval()</code> to call <code><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble()</a></code> only on non-tibble input, which prevents groups from being stripped from <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble()</a></code> input (<a href="https://github.com/rnabioco/valr/issues/338" class="external-link">#338</a>).</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.1" id="valr-041">valr 0.4.1<a class="anchor" aria-label="anchor" href="#valr-041"></a></h2><p class="text-muted">CRAN release: 2018-06-08</p>
<ul><li><p>Added new function, <code><a href="../reference/bed_partition.html">bed_partition()</a></code>, which is similar to <code><a href="../reference/bed_merge.html">bed_merge()</a></code> but collapses intervals to elemental intervals rather than the maximal overlapping region. <code><a href="../reference/bed_partition.html">bed_partition()</a></code> also can compute summaries of data from overlapping intervals. See examples in <code><a href="../reference/bed_partition.html">bed_partition()</a></code> and timings in <code>vignette('benchmarks')</code> <a href="https://github.com/kriemo" class="external-link">@kriemo</a>.</p></li>
<li><p>Several explicit comparisons to the Bioconductor GenomicRanges library are included for users considering using valr. See examples in <code>as.tbl_interval()</code> and timings in <code>vignette('benchmarks')</code>.</p></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.0" id="valr-040">valr 0.4.0<a class="anchor" aria-label="anchor" href="#valr-040"></a></h2><p class="text-muted">CRAN release: 2018-01-25</p>
<div class="section level3">
<h3 id="minor-changes-0-4-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-4-0"></a></h3>
<ul><li><p>All relevant tests from bedtool2 were ported into valr. Bugs identified in corner cases by new tests were fixed (<a href="https://github.com/rnabioco/valr/issues/328" class="external-link">#328</a> <a href="https://github.com/raysinesis" class="external-link">@raysinesis</a>)</p></li>
<li><p><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> now works with grouped inputs (<a href="https://github.com/rnabioco/valr/issues/216" class="external-link">#216</a>)</p></li>
<li><p>Update dplyr header files to v0.7</p></li>
<li><p><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code> and internal <code>intersect_impl</code> were refactored to enable return of non-intersecting intervals.</p></li>
<li><p>The genome argument to <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> was deprecated and will produce a warning if used. Also error handling was added to check and warn if there are intervals smaller than the requested window size in <code>makewindows_impl()</code> (<a href="https://github.com/rnabioco/valr/issues/312" class="external-link">#312</a> <a href="https://github.com/kriemo" class="external-link">@kriemo</a>)</p></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-4-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-4-0"></a></h3>
<ul><li><p>Fixed off by one error in reported distances from <code><a href="../reference/bed_closest.html">bed_closest()</a></code>. Distances reported now are the same as <code>bedtools closest</code> behavior (<a href="https://github.com/rnabioco/valr/issues/311" class="external-link">#311</a>).</p></li>
<li><p><code><a href="../reference/bed_glyph.html">bed_glyph()</a></code> accepts <code>trbl_intervals</code> named other than <code>x</code> and <code>y</code> (<a href="https://github.com/rnabioco/valr/issues/318" class="external-link">#318</a>).</p></li>
<li><p><code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> now returns the number of windows specified by <code>num_win</code> when the input intervals are not evenly divisble into <code>num_win</code>, consistent with <code>bedtools</code> behavior.</p></li>
<li><p>The output of <code>findOverlaps()</code> is now sorted in <code>subtract_impl()</code> to prevent reporting intervals that should have been dropped when calling <code><a href="../reference/bed_subtract.html">bed_subtract()</a></code> (<a href="https://github.com/rnabioco/valr/issues/316" class="external-link">#316</a> <a href="https://github.com/kriemo" class="external-link">@kriemo</a>)</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.1" id="valr-031">valr 0.3.1<a class="anchor" aria-label="anchor" href="#valr-031"></a></h2><p class="text-muted">CRAN release: 2017-07-22</p>
<div class="section level3">
<h3 id="enhancements-0-3-1">Enhancements<a class="anchor" aria-label="anchor" href="#enhancements-0-3-1"></a></h3>
<ul><li><p>A manuscript describing valr has been published in <a href="https://f1000research.com/articles/6-1025/v1" class="external-link">F1000Research</a>.</p></li>
<li><p>New S3 generic <code>as.tbl_interval()</code> converts <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GenomicRanges::GRanges</a></code> objects to <code>tbl_interval</code>.</p></li>
<li><p>New <code><a href="../reference/create_tss.html">create_tss()</a></code> for creating transcription start sites.</p></li>
<li><p>Improve documentation of interval statistics with more complex examples.</p></li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-3-1">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-3-1"></a></h3>
<ul><li>
<code><a href="../reference/bed_sort.html">bed_sort()</a></code> has been de-deprecated to reduce <code>arrange</code> calls in library code.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-3-1">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-3-1"></a></h3>
<ul><li>
<code><a href="../reference/bed_merge.html">bed_merge()</a></code> now reports start/end columns if spec is provided (<a href="https://github.com/rnabioco/valr/issues/288" class="external-link">#288</a>)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.0" id="valr-030">valr 0.3.0<a class="anchor" aria-label="anchor" href="#valr-030"></a></h2><p class="text-muted">CRAN release: 2017-06-15</p>
<div class="section level3">
<h3 id="enhancements-0-3-0">Enhancements<a class="anchor" aria-label="anchor" href="#enhancements-0-3-0"></a></h3>
<ul><li><p>New <code><a href="../reference/create_introns.html">create_introns()</a></code>, <code><a href="../reference/create_utrs5.html">create_utrs5()</a></code> and <code><a href="../reference/create_utrs3.html">create_utrs3()</a></code> functions for generating features from BED12 files.</p></li>
<li><p>Speed-ups in <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> (~50x), <code><a href="../reference/bed_merge.html">bed_merge()</a></code> (~4x), and <code><a href="../reference/bed_flank.html">bed_flank()</a></code> (~4x) (thanks to <a href="https://github.com/kriemo" class="external-link">@kriemo</a> and <a href="https://github.com/sheridar" class="external-link">@sheridar</a>). Thanks to the sponsors of the Biofrontiers Hackathon for the caffeine underlying these improvements.</p></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-3-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-3-0"></a></h3>
<ul><li>intervals from <code><a href="../reference/bed_random.html">bed_random()</a></code> are now sorted properly.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.0" id="valr-020">valr 0.2.0<a class="anchor" aria-label="anchor" href="#valr-020"></a></h2><p class="text-muted">CRAN release: 2017-05-05</p>
<div class="section level3">
<h3 id="major-changes-0-2-0">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-2-0"></a></h3>
<ul><li><p>Package dplyr v0.5.0 headers with valr to remove dplyr LinkingTo dependency.</p></li>
<li><p><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code> now accepts multiple tbls for intersection (<a href="https://github.com/rnabioco/valr/issues/220" class="external-link">#220</a> <a href="https://github.com/kriemo" class="external-link">@kriemo</a>).</p></li>
<li><p>new <code>tbl_interval()</code> and <code>tbl_genome()</code> that wrap tibbles and enforce strict column naming. <code>trbl_interval()</code> and <code>trbl_genome()</code> are constructors that take <code><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tibble::tribble()</a></code> formatting and <code>is.tbl_interval()</code> and <code>is.tbl_genome()</code> are used to check for valid classes.</p></li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-2-0">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-2-0"></a></h3>
<ul><li>intervals returned from <code><a href="../reference/bed_random.html">bed_random()</a></code> are sorted by <code>chrom</code> and <code>start</code> by default.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-2-0">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-2-0"></a></h3>
<ul><li>Merge intervals in <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> and use numeric values for calculation (fixes <a href="https://github.com/rnabioco/valr/issues/204" class="external-link">#204</a>).</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.2" id="valr-012">valr 0.1.2<a class="anchor" aria-label="anchor" href="#valr-012"></a></h2><p class="text-muted">CRAN release: 2017-03-16</p>
<div class="section level3">
<h3 id="major-changes-0-1-2">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-1-2"></a></h3>
<ul><li>Deprecate <code><a href="../reference/bed_sort.html">bed_sort()</a></code> in favor of using <code><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">dplyr::arrange()</a></code> explicitly (fixes <a href="https://github.com/rnabioco/valr/issues/134" class="external-link">#134</a>).</li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-1-2">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-1-2"></a></h3>
<ul><li><p>add <code>src/init.c</code> that calls <code>R_registerRoutines</code> and <code>R_useDynamicSymbols</code> to address NOTE in r-devel</p></li>
<li><p>Deprecate <code>dist</code> parameter in <code><a href="../reference/bed_closest.html">bed_closest()</a></code> in favor of using user supplied functions (<a href="https://github.com/rnabioco/valr/issues/182" class="external-link">#182</a> <a href="https://github.com/kriemo" class="external-link">@kriemo</a>)</p></li>
<li><p>Make <code>.id</code> values sequential across chroms in <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code> output (<a href="https://github.com/rnabioco/valr/issues/171" class="external-link">#171</a>)</p></li>
<li><p>Transfer repository to <a href="http://github.com/rnabioco/valr" class="external-link uri">http://github.com/rnabioco/valr</a>, update links and docs.</p></li>
<li><p>Move shiny app to new repo (<a href="http://github.com/rnabioco/valrdata" class="external-link uri">http://github.com/rnabioco/valrdata</a>).</p></li>
<li><p>Add Kent Riemondy to LICENSE file.</p></li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-1-2">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-1-2"></a></h3>
<ul><li>
<code><a href="../reference/bed_merge.html">bed_merge()</a></code> now merges contained intervals (<a href="https://github.com/rnabioco/valr/issues/177" class="external-link">#177</a>)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.1" id="valr-011">valr 0.1.1<a class="anchor" aria-label="anchor" href="#valr-011"></a></h2><p class="text-muted">CRAN release: 2016-12-01</p>
<div class="section level3">
<h3 id="minor-changes-0-1-1">Minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-1-1"></a></h3>
<ul><li><p>test / vignette guards for Suggested RMySQL</p></li>
<li><p>fixed memory leak in absdist.cpp</p></li>
<li><p>fixed vignette entry names</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.0" id="valr-010">valr 0.1.0<a class="anchor" aria-label="anchor" href="#valr-010"></a></h2><p class="text-muted">CRAN release: 2016-11-21</p>
<div class="section level3">
<h3 id="major-changes-0-1-0">Major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-1-0"></a></h3>
<ul><li>initial release on CRAN</li>
</ul></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer></div>





  </body></html>

