<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>valr examples &bull; valr</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/valr.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="http://github.com/jayhesselberth/valr/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>valr examples</h1>
                        <h4 class="author">Jay Hesselberth <a href="mailto:jay.hesselberth@gmail.com">jay.hesselberth@gmail.com</a></h4>
            
            <h4 class="date">2016-10-31</h4>
          </div>

    
    
<div class="contents">
<div id="demonstrations" class="section level1">
<h1 class="hasAnchor"><html><body><a href="#demonstrations" class="anchor"> </a></body></html>Demonstrations</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(dplyr)</code></pre></div>
<div id="demo-intron" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#demo-intron" class="anchor"> </a></body></html>Creating intron intervals</h2>
<p>Intron intervals can be identified by calculating:</p>
<ol style="list-style-type: decimal"><li>Intervals that are covered by any gene annotation</li>
<li>Then, intervals that are not covered by exons.</li>
</ol><div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))
genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed12</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.refGene.chr22.bed.gz'</span>))

exons &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed12_to_exons.html">bed12_to_exons</a></span>(genes)
exons

introns &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(genes, exons)
introns</code></pre></div>
</div>
<div id="histone-density-and-transcription-start-sites" class="section level2">
<h2 class="hasAnchor"><html><body><a href="#histone-density-and-transcription-start-sites" class="anchor"> </a></body></html>Histone density and transcription start sites</h2>
<p>This demonstration illustrates how to use <code>valr</code> tools to perform a &ldquo;meta-analysis&rdquo; of signals relative to genomic features. Here we to analyze the distribution of histone marks surrounding transcription start sites.</p>
<p>First we load libraries and relevant data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(dplyr)

<span class="co"># `valr_example()` identifies the path of example files</span>
bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
genomefile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(genomefile)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)</code></pre></div>
<p>Then we generate 1 bp intervals to represent transcription start sites (TSSs). We focus on <code>+</code> strand genes, but <code>-</code> genes are easily accomodated by filtering them and using <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> with <code>reverse</code>d window numbers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate 1 bp TSS intervals, `+` strand only</span>
tss &lt;-<span class="st"> </span>genes %&gt;%
<span class="st">  </span><span class="kw">filter</span>(strand ==<span class="st"> '+'</span>) %&gt;%
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">end =</span> start +<span class="st"> </span><span class="dv">1</span>)

<span class="co"># 1000 bp up and downstream</span>
region_size &lt;-<span class="st"> </span><span class="dv">1000</span>
<span class="co"># 50 bp windows</span>
win_size &lt;-<span class="st"> </span><span class="dv">50</span>

<span class="co"># add slop to the TSS, break into windows</span>
x &lt;-<span class="st"> </span>tss %&gt;%
<span class="st">  </span><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(genome, <span class="dt">both =</span> region_size) %&gt;%
<span class="st">  </span><span class="kw"><a href="../reference/bed_makewindows.html">bed_makewindows</a></span>(genome, win_size)

x
<span class="co">#&gt; # A tibble: 13,530 &times; 7</span>
<span class="co">#&gt;    chrom    start      end      name score strand .win_id</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr22 16161065 16161115 LINC00516     3      +       1</span>
<span class="co">#&gt; 2  chr22 16161115 16161165 LINC00516     3      +       2</span>
<span class="co">#&gt; 3  chr22 16161165 16161215 LINC00516     3      +       3</span>
<span class="co">#&gt; 4  chr22 16161215 16161265 LINC00516     3      +       4</span>
<span class="co">#&gt; 5  chr22 16161265 16161315 LINC00516     3      +       5</span>
<span class="co">#&gt; 6  chr22 16161315 16161365 LINC00516     3      +       6</span>
<span class="co">#&gt; 7  chr22 16161365 16161415 LINC00516     3      +       7</span>
<span class="co">#&gt; 8  chr22 16161415 16161465 LINC00516     3      +       8</span>
<span class="co">#&gt; 9  chr22 16161465 16161515 LINC00516     3      +       9</span>
<span class="co">#&gt; 10 chr22 16161515 16161565 LINC00516     3      +      10</span>
<span class="co">#&gt; # ... with 13,520 more rows</span></code></pre></div>
<p>Note that the data are now grouped by <code>.win_id</code>. This grouping is used below with <code><a href="../reference/bed_map.html">bed_map()</a></code> to caluclate summary statistics identified by mapping <code>y</code> signals onto the intervals in <code>x</code>. Note that <code>x</code> intervals with no <code>y</code> overlaps have values of <code>NA</code>, so <code>na.rm = TRUE</code> must be used during computation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># map signals to TSS regions and calculate summary statistics.</span>
res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">win_sum =</span> <span class="kw">sum</span>(value, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(.win_id) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">win_mean =</span> <span class="kw">mean</span>(win_sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),
            <span class="dt">win_sd =</span> <span class="kw">sd</span>(win_sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))</code></pre></div>
<p>Finally, these summary statistics are used to construct a plot that illustrates histone density surrounding TSSs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)

x_labels &lt;-<span class="st"> </span><span class="kw">seq</span>(-region_size, region_size, <span class="dt">by =</span> win_size *<span class="st"> </span><span class="dv">5</span>)
x_breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">41</span>, <span class="dt">by =</span> <span class="dv">5</span>)

sd_limits &lt;-<span class="st"> </span><span class="kw">aes</span>(<span class="dt">ymax =</span> win_mean +<span class="st"> </span>win_sd, <span class="dt">ymin =</span> win_mean -<span class="st"> </span>win_sd)

<span class="kw">ggplot</span>(res, <span class="kw">aes</span>(<span class="dt">x =</span> .win_id, <span class="dt">y =</span> win_mean)) +
<span class="st">  </span><span class="kw">geom_point</span>() +<span class="st"> </span><span class="kw">geom_pointrange</span>(sd_limits) +<span class="st"> </span>
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">labels =</span> x_labels, <span class="dt">breaks =</span> x_breaks) +<span class="st"> </span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">'Position (bp from TSS)'</span>) +<span class="st"> </span><span class="kw">ylab</span>(<span class="st">'Signal'</span>) +<span class="st"> </span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">'Human H3K4me3 signal near transcription starts sites'</span>) +
<span class="st">  </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="img/valr-examples-plot-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#demonstrations">Demonstrations</a><ul class="nav nav-pills nav-stacked"><li><a href="#demo-intron">Creating intron intervals</a></li>
      <li><a href="#histone-density-and-transcription-start-sites">Histone density and transcription start sites</a></li>
      </ul></li>
      </ul></div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jay Hesselberth, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer></div>

  </body></html>
