<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interval statistics • valr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/valr.html">Overview</a>
    </li>
    <li>
      <a href="../articles/interval-stats.html">Interval Statistics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/rnabioco/valr/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Interval statistics</h1>
            
            <h4 class="date">May 30 2017</h4>
          </div>

    
    
<div class="contents">
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p><code>valr</code> can be used to explore relationships between sets of intervals. Here we explore the relationship between transcription start sites and repetitive elements in the human genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(dplyr)
<span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(tidyr)

<span class="co"># load repeats and genes. Data in the valr package is restricted to chr22; the entire</span>
<span class="co"># files can be downloaded from UCSC.</span>
rpts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.rmsk.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>) 
genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.refGene.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">12</span>)

<span class="co"># load chrom sizes</span>
genome &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))

<span class="co"># create 1bp intervals representing transcription start sites</span>
tss &lt;-
<span class="st">  </span><span class="kw">mutate</span>(genes,
         <span class="dt">.start =</span> <span class="kw">ifelse</span>(strand <span class="op">==</span><span class="st"> '+'</span>, start, end),
         <span class="dt">.end =</span> .start <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> .start, <span class="dt">end =</span> .end, name)

tss
<span class="co">#&gt; # A tibble: 1,267 x 4</span>
<span class="co">#&gt;    chrom    start      end         name</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;dbl&gt;        &lt;chr&gt;</span>
<span class="co">#&gt;  1 chr22 16193009 16193010    NR_122113</span>
<span class="co">#&gt;  2 chr22 16157078 16157079    NR_133911</span>
<span class="co">#&gt;  3 chr22 16162065 16162066    NR_073459</span>
<span class="co">#&gt;  4 chr22 16162065 16162066    NR_073460</span>
<span class="co">#&gt;  5 chr22 16231289 16231290    NR_132385</span>
<span class="co">#&gt;  6 chr22 16287937 16287938 NM_001136213</span>
<span class="co">#&gt;  7 chr22 16274608 16274609    NR_046571</span>
<span class="co">#&gt;  8 chr22 16449804 16449805 NM_001005239</span>
<span class="co">#&gt;  9 chr22 17073700 17073701    NM_014406</span>
<span class="co">#&gt; 10 chr22 17082800 17082801    NR_001591</span>
<span class="co">#&gt; # ... with 1,257 more rows</span></code></pre></div>
</div>
<div id="distance-metrics" class="section level2">
<h2 class="hasAnchor">
<a href="#distance-metrics" class="anchor"></a>Distance metrics</h2>
<p>First we define a function that takes <code>x</code> and <code>y</code> intervals and computes distance statistics (using <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>) for specified groups. The value of each statistic is assigned to a <code>.value</code> column.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">distance_stats &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, genome, grp, <span class="dt">type =</span> <span class="ot">NA</span>) {
  <span class="kw">group_by_</span>(x, <span class="dt">.dots =</span> grp) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">do</span>(
       <span class="dt">reldist =</span> <span class="kw"><a href="../reference/bed_reldist.html">bed_reldist</a></span>(., y, <span class="dt">detail =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span>
<span class="st">         </span><span class="kw">select</span>(<span class="dt">.value =</span> .reldist),
       <span class="dt">absdist =</span> <span class="kw"><a href="../reference/bed_absdist.html">bed_absdist</a></span>(., y, genome) <span class="op">%&gt;%</span>
<span class="st">         </span><span class="kw">select</span>(<span class="dt">.value =</span> .absdist)
       ) <span class="op">%&gt;%</span>
<span class="st">    </span>tidyr<span class="op">::</span><span class="kw">gather_</span>(<span class="st">'stat'</span>, <span class="st">'value'</span>, <span class="kw">setdiff</span>(<span class="kw">names</span>(.), <span class="kw">list</span>(grp))) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">type =</span> type)
} </code></pre></div>
<p>We use the <code>distance_stats()</code> function to apply the <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code> function to each group of data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">obs_stats &lt;-<span class="st"> </span><span class="kw">distance_stats</span>(rpts, tss, genome, <span class="st">'name'</span>, <span class="st">'obs'</span>)
obs_stats
<span class="co">#&gt; # A tibble: 2,106 x 4</span>
<span class="co">#&gt;          name    stat             value  type</span>
<span class="co">#&gt;         &lt;chr&gt;   &lt;chr&gt;            &lt;list&gt; &lt;chr&gt;</span>
<span class="co">#&gt;  1       (A)n reldist &lt;tibble [27 x 1]&gt;   obs</span>
<span class="co">#&gt;  2 (AAAAACA)n reldist  &lt;tibble [1 x 1]&gt;   obs</span>
<span class="co">#&gt;  3   (AAAAC)n reldist  &lt;tibble [6 x 1]&gt;   obs</span>
<span class="co">#&gt;  4   (AAAAG)n reldist  &lt;tibble [2 x 1]&gt;   obs</span>
<span class="co">#&gt;  5   (AAAAT)n reldist  &lt;tibble [3 x 1]&gt;   obs</span>
<span class="co">#&gt;  6  (AAAATA)n reldist  &lt;tibble [1 x 1]&gt;   obs</span>
<span class="co">#&gt;  7 (AAAATAA)n reldist  &lt;tibble [1 x 1]&gt;   obs</span>
<span class="co">#&gt;  8    (AAAC)n reldist  &lt;tibble [2 x 1]&gt;   obs</span>
<span class="co">#&gt;  9   (AAACA)n reldist  &lt;tibble [1 x 1]&gt;   obs</span>
<span class="co">#&gt; 10  (AAACAA)n reldist  &lt;tibble [1 x 1]&gt;   obs</span>
<span class="co">#&gt; # ... with 2,096 more rows</span></code></pre></div>
<p>And the same is done for a set of shuffled group of data. <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code> is used to shuffle coordinates of the repeats within each chromosome (i.e., the coordinates change, but the chromosome stays the same.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shfs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(rpts, genome, <span class="dt">within =</span> <span class="ot">TRUE</span>)
shf_stats &lt;-<span class="st"> </span><span class="kw">distance_stats</span>(shfs, tss, genome, <span class="st">'name'</span>, <span class="st">'shuf'</span>)</code></pre></div>
<p>Now we can bind the observed and shuffled data together, and do some tidying to put the data into a format appropriate for a statistical test. This involves:</p>
<ol style="list-style-type: decimal">
<li>
<code>unnest()</code>ing the data frames</li>
<li>creating groups for each repeat (<code>name</code>), stat (<code>reldist</code> or <code>absdist</code>) and type (<code>obs</code> or <code>shf</code>)</li>
<li>adding unique surrogate row numbers for each group</li>
<li>using <code>tidyr::spread()</code> to create two new <code>obs</code> and <code>shuf</code> columns</li>
<li>removing rows with <code>NA</code> values.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res &lt;-<span class="st"> </span><span class="kw">bind_rows</span>(obs_stats, shf_stats) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(value) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">group_by</span>(name, stat, type) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.id =</span> <span class="kw">row_number</span>()) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">spread</span>(type, .value) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">na.omit</span>()

res
<span class="co">#&gt; # A tibble: 16,681 x 5</span>
<span class="co">#&gt; # Groups:   name, stat [1,906]</span>
<span class="co">#&gt;     name    stat   .id     obs     shuf</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;chr&gt; &lt;int&gt;   &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1  (A)n absdist     1 5120782 15543320</span>
<span class="co">#&gt;  2  (A)n absdist     2  226476 11260783</span>
<span class="co">#&gt;  3  (A)n absdist     3   31241  7823902</span>
<span class="co">#&gt;  4  (A)n absdist     4    8931  6280428</span>
<span class="co">#&gt;  5  (A)n absdist     5   41316  5741843</span>
<span class="co">#&gt;  6  (A)n absdist     6   22519  2568034</span>
<span class="co">#&gt;  7  (A)n absdist     7   15173  2434224</span>
<span class="co">#&gt;  8  (A)n absdist     8   21985  2346246</span>
<span class="co">#&gt;  9  (A)n absdist     9   13457  1765667</span>
<span class="co">#&gt; 10  (A)n absdist    10  134007    25500</span>
<span class="co">#&gt; # ... with 16,671 more rows</span></code></pre></div>
<p>Now that the data are formatted, we can use <code>ks.test()</code> to determine whether there are significant differences between the observed and shuffled data for each group. <code>broom::tidy()</code> is used to reformat the results of each test into a <code>data_frame</code>, and the results of each test are <code>gather</code>ed to into a <code>type</code> column for each test type.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(broom)

pvals &lt;-<span class="st"> </span>res <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">do</span>(<span class="dt">twosided =</span> <span class="kw">tidy</span>(<span class="kw">ks.test</span>(.<span class="op">$</span>obs, .<span class="op">$</span>shuf)),
                    <span class="dt">less =</span> <span class="kw">tidy</span>(<span class="kw">ks.test</span>(.<span class="op">$</span>obs, .<span class="op">$</span>shuf, <span class="dt">alternative =</span> <span class="st">'less'</span>)),
                    <span class="dt">greater =</span> <span class="kw">tidy</span>(<span class="kw">ks.test</span>(.<span class="op">$</span>obs, .<span class="op">$</span>shuf, <span class="dt">alternative =</span> <span class="st">'greater'</span>))) <span class="op">%&gt;%</span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">gather</span>(alt, type, <span class="op">-</span>name, <span class="op">-</span>stat) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">unnest</span>(type) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(name<span class="op">:</span>p.value) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(p.value)</code></pre></div>
<p>Histgrams of the different stats help visulaize the distribution of p.values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(pvals, <span class="kw">aes</span>(p.value)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">binwidth =</span> <span class="fl">0.05</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">facet_grid</span>(stat <span class="op">~</span><span class="st"> </span>alt) <span class="op">+</span><span class="st"> </span><span class="kw">theme_bw</span>()</code></pre></div>
<p><img src="img/interval-stats-pvalue_viz-1.png" width="384" style="display: block; margin: auto;"></p>
<p>We can also assess false discovery rates (q.values) using <code>p.adjust()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pvals &lt;-
<span class="st">  </span><span class="kw">group_by</span>(pvals, stat, alt) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">q.value =</span> <span class="kw">p.adjust</span>(p.value)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(q.value)</code></pre></div>
<p>Finally we can visualize these results using <code>stat_ecdf()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res_gather &lt;-<span class="st"> </span>tidyr<span class="op">::</span><span class="kw">gather</span>(res, type, value, <span class="op">-</span>name, <span class="op">-</span>stat, <span class="op">-</span>.id)

signif &lt;-<span class="st"> </span><span class="kw">head</span>(pvals, <span class="dv">5</span>)

res_signif &lt;-
<span class="st">  </span>signif <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">left_join</span>(res_gather, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">'name'</span>, <span class="st">'stat'</span>))

<span class="kw">ggplot</span>(res_signif, <span class="kw">aes</span>(<span class="dt">x =</span> value, <span class="dt">color =</span> type)) <span class="op">+</span>
<span class="st">  </span><span class="kw">stat_ecdf</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">facet_grid</span>(stat <span class="op">~</span><span class="st"> </span>name) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>() <span class="op">+</span><span class="st"> </span><span class="kw">scale_x_log10</span>()</code></pre></div>
<p><img src="img/interval-stats-ecfs-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
<div id="projection-test" class="section level2">
<h2 class="hasAnchor">
<a href="#projection-test" class="anchor"></a>Projection test</h2>
<p><code><a href="../reference/bed_projection.html">bed_projection()</a></code> is a statistical approach to assess the relationship between two intervals based on the binomial distribution. Here, we examine the distribution of repetitive elements within the promoters of coding or non-coding genes.</p>
<p>First we’ll extract 5 kb regions upstream of the transcription start sites to represent the promoter regions for coding and non-coding genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create intervals 5kb upstream of tss representing promoters</span>
promoters &lt;-
<span class="st">  </span><span class="kw"><a href="../reference/bed_flank.html">bed_flank</a></span>(genes, genome, <span class="dt">left =</span> <span class="dv">5000</span>, <span class="dt">strand =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">name =</span> <span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">'NR_'</span>, name), <span class="st">'non-coding'</span>, <span class="st">'coding'</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(chrom<span class="op">:</span>strand)
  
<span class="co"># select coding and non-coding promoters</span>
promoters_coding &lt;-<span class="st"> </span><span class="kw">filter</span>(promoters, name <span class="op">==</span><span class="st"> 'coding'</span>)
promoters_ncoding &lt;-<span class="st"> </span><span class="kw">filter</span>(promoters, name <span class="op">==</span><span class="st"> 'non-coding'</span>)

promoters_coding
<span class="co">#&gt; # A tibble: 973 x 6</span>
<span class="co">#&gt;    chrom    start      end   name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1 chr22 16287937 16292937 coding     0      -</span>
<span class="co">#&gt;  2 chr22 16449804 16454804 coding     0      -</span>
<span class="co">#&gt;  3 chr22 17073700 17078700 coding     0      -</span>
<span class="co">#&gt;  4 chr22 17302589 17307589 coding     0      -</span>
<span class="co">#&gt;  5 chr22 17302589 17307589 coding     0      -</span>
<span class="co">#&gt;  6 chr22 17489112 17494112 coding     0      -</span>
<span class="co">#&gt;  7 chr22 17560848 17565848 coding     0      +</span>
<span class="co">#&gt;  8 chr22 17560848 17565848 coding     0      +</span>
<span class="co">#&gt;  9 chr22 17602213 17607213 coding     0      -</span>
<span class="co">#&gt; 10 chr22 17602257 17607257 coding     0      -</span>
<span class="co">#&gt; # ... with 963 more rows</span>

promoters_ncoding
<span class="co">#&gt; # A tibble: 294 x 6</span>
<span class="co">#&gt;    chrom    start      end       name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;      &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt;  1 chr22 16152078 16157078 non-coding     0      +</span>
<span class="co">#&gt;  2 chr22 16157065 16162065 non-coding     0      +</span>
<span class="co">#&gt;  3 chr22 16157065 16162065 non-coding     0      +</span>
<span class="co">#&gt;  4 chr22 16193009 16198009 non-coding     0      -</span>
<span class="co">#&gt;  5 chr22 16231289 16236289 non-coding     0      -</span>
<span class="co">#&gt;  6 chr22 16269608 16274608 non-coding     0      +</span>
<span class="co">#&gt;  7 chr22 17077800 17082800 non-coding     0      +</span>
<span class="co">#&gt;  8 chr22 17156430 17161430 non-coding     0      -</span>
<span class="co">#&gt;  9 chr22 17229328 17234328 non-coding     0      -</span>
<span class="co">#&gt; 10 chr22 17303363 17308363 non-coding     0      +</span>
<span class="co">#&gt; # ... with 284 more rows</span></code></pre></div>
<p>Next we’ll apply the <code><a href="../reference/bed_projection.html">bed_projection()</a></code> test for each repeat class for both coding and non-coding regions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">
<span class="co"># function to apply bed_projection to groups</span>
projection_stats &lt;-<span class="st"> </span><span class="cf">function</span>(x, y, genome, grp, <span class="dt">type =</span> <span class="ot">NA</span>) {
  <span class="kw">group_by_</span>(x, <span class="dt">.dots =</span> grp) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">do</span>(<span class="dt">n_repeats =</span> <span class="kw">nrow</span>(.),
       <span class="dt">projection =</span> <span class="kw"><a href="../reference/bed_projection.html">bed_projection</a></span>(., y, genome)) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">type =</span> type)
} 

pvals_coding &lt;-<span class="st"> </span><span class="kw">projection_stats</span>(rpts, promoters_coding, genome, <span class="st">'name'</span>, <span class="st">'coding'</span>)
pvals_ncoding &lt;-<span class="st"> </span><span class="kw">projection_stats</span>(rpts, promoters_ncoding, genome, <span class="st">'name'</span>, <span class="st">'non_coding'</span>)

pvals &lt;-
<span class="st">  </span><span class="kw">bind_rows</span>(pvals_ncoding, pvals_coding) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>chrom)

<span class="co"># filter for repeat classes with at least 10 intervals</span>
pvals &lt;-<span class="st"> </span><span class="kw">filter</span>(pvals, n_repeats <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>,
                obs_exp_ratio <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)

<span class="co"># adjust pvalues </span>
pvals &lt;-<span class="st"> </span><span class="kw">mutate</span>(pvals, <span class="dt">q.value =</span> <span class="kw">p.adjust</span>(p.value))

pvals
<span class="co">#&gt; # A tibble: 179 x 7</span>
<span class="co">#&gt;      name       type n_repeats     p.value obs_exp_ratio lower_tail</span>
<span class="co">#&gt;     &lt;chr&gt;      &lt;chr&gt;     &lt;int&gt;       &lt;dbl&gt;         &lt;dbl&gt;      &lt;chr&gt;</span>
<span class="co">#&gt;  1   (A)n non_coding        28 0.003529188     4.7150373      FALSE</span>
<span class="co">#&gt;  2  (AT)n non_coding        48 0.297946444     0.9168128      FALSE</span>
<span class="co">#&gt;  3  (CA)n non_coding        31 0.156135635     1.4195811      FALSE</span>
<span class="co">#&gt;  4  (GT)n non_coding        42 0.247258594     1.0477861      FALSE</span>
<span class="co">#&gt;  5   (T)n non_coding        61 0.404908066     0.7214265      FALSE</span>
<span class="co">#&gt;  6  (TG)n non_coding        40 0.062233679     2.2003507      FALSE</span>
<span class="co">#&gt;  7 A-rich non_coding        54 0.348067174     0.8149447      FALSE</span>
<span class="co">#&gt;  8    Alu non_coding        15 0.044563006     2.9338010      FALSE</span>
<span class="co">#&gt;  9  AluJb non_coding       271 0.022488504     1.7862626      FALSE</span>
<span class="co">#&gt; 10  AluJo non_coding       208 0.021566061     1.9041497      FALSE</span>
<span class="co">#&gt; # ... with 169 more rows, and 1 more variables: q.value &lt;dbl&gt;</span></code></pre></div>
<p>The projection test is a two-tailed statistical test. A significant p-value indicates either enrichment or depletion of query intervals compared to the reference interval sets. A value of <code>lower_tail = TRUE</code> column indicates that the query intervals are depleted, whereas <code>lower_tail = FALSE</code> indicates that the query intervals are enriched.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(knitr)

<span class="co"># find and show top 5 most significant repeats</span>
signif_tests &lt;-
<span class="st">  </span>pvals <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">arrange</span>(q.value) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(type) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">top_n</span>(<span class="op">-</span><span class="dv">5</span>, q.value) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">arrange</span>(type)

knitr<span class="op">::</span><span class="kw">kable</span>(signif_tests)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">name</th>
<th align="left">type</th>
<th align="right">n_repeats</th>
<th align="right">p.value</th>
<th align="right">obs_exp_ratio</th>
<th align="left">lower_tail</th>
<th align="right">q.value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AluSx1</td>
<td align="left">coding</td>
<td align="right">364</td>
<td align="right">0.0000114</td>
<td align="right">2.106148</td>
<td align="left">FALSE</td>
<td align="right">0.0020357</td>
</tr>
<tr class="even">
<td align="left">AluSx3</td>
<td align="left">coding</td>
<td align="right">95</td>
<td align="right">0.0001422</td>
<td align="right">2.914120</td>
<td align="left">FALSE</td>
<td align="right">0.0251780</td>
</tr>
<tr class="odd">
<td align="left">AluJb</td>
<td align="left">coding</td>
<td align="right">271</td>
<td align="right">0.0002220</td>
<td align="right">2.043110</td>
<td align="left">FALSE</td>
<td align="right">0.0384039</td>
</tr>
<tr class="even">
<td align="left">Charlie4z</td>
<td align="left">coding</td>
<td align="right">11</td>
<td align="right">0.0012290</td>
<td align="right">5.807861</td>
<td align="left">FALSE</td>
<td align="right">0.2052397</td>
</tr>
<tr class="odd">
<td align="left">AluSz</td>
<td align="left">coding</td>
<td align="right">259</td>
<td align="right">0.0013182</td>
<td align="right">1.891105</td>
<td align="left">FALSE</td>
<td align="right">0.2188250</td>
</tr>
<tr class="even">
<td align="left">MER3</td>
<td align="left">non_coding</td>
<td align="right">27</td>
<td align="right">0.0000270</td>
<td align="right">8.149447</td>
<td align="left">FALSE</td>
<td align="right">0.0048081</td>
</tr>
<tr class="odd">
<td align="left">L1MC4a</td>
<td align="left">non_coding</td>
<td align="right">23</td>
<td align="right">0.0001448</td>
<td align="right">7.653394</td>
<td align="left">FALSE</td>
<td align="right">0.0254769</td>
</tr>
<tr class="even">
<td align="left">L1ME1</td>
<td align="left">non_coding</td>
<td align="right">38</td>
<td align="right">0.0002034</td>
<td align="right">5.790397</td>
<td align="left">FALSE</td>
<td align="right">0.0355962</td>
</tr>
<tr class="odd">
<td align="left">G-rich</td>
<td align="left">non_coding</td>
<td align="right">54</td>
<td align="right">0.0002172</td>
<td align="right">4.889668</td>
<td align="left">FALSE</td>
<td align="right">0.0377926</td>
</tr>
<tr class="even">
<td align="left">L1MC5</td>
<td align="left">non_coding</td>
<td align="right">29</td>
<td align="right">0.0004561</td>
<td align="right">6.069933</td>
<td align="left">FALSE</td>
<td align="right">0.0784540</td>
</tr>
</tbody>
</table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#distance-metrics">Distance metrics</a></li>
      <li><a href="#projection-test">Projection test</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jay Hesselberth, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
