<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>valr overview • valr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/valr.html">Overview</a>
    </li>
    <li>
      <a href="../articles/interval-stats.html">Interval Statistics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/rnabioco/valr/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>valr overview</h1>
            
            <h4 class="date">2017-05-30</h4>
          </div>

    
    
<div class="contents">
<div id="why-valr" class="section level2">
<h2 class="hasAnchor">
<a href="#why-valr" class="anchor"></a>Why <code>valr</code>?</h2>
<p><strong>Why another tool set for interval manipulations?</strong> There are several <a href="#related_work">other software packages</a> available for genome interval analysis. However, based on our experiences teaching genome analysis, we were motivated to develop a toolset that:</p>
<ul>
<li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the “modern R” syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul>
<p><code>valr</code> can currently be used for analysis of pre-processed data in BED and related formats. We plan to support BAM and VCF files soon via tabix indexes.</p>
<div id="familiar-tools-all-within-r" class="section level3">
<h3 class="hasAnchor">
<a href="#familiar-tools-all-within-r" class="anchor"></a>Familiar tools, all within R</h3>
<p>The functions in <code>valr</code> have similar names to their <code>BEDtools</code> counterparts, and so will be familiar to users coming from the <code>BEDtools</code> suite. Similar to <a href="https://daler.github.io/pybedtools/#why-pybedtools"><code>pybedtools</code></a>, <code>valr</code> has a terse syntax:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(dplyr)

snps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.snps147.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>)
genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>)

<span class="co"># find snps in intergenic regions</span>
intergenic &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(snps, genes)
<span class="co"># distance from intergenic snps to nearest gene</span>
nearby &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(intergenic, genes)

nearby <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="kw">starts_with</span>(<span class="st">'name'</span>), .overlap, .dist) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(<span class="kw">abs</span>(.dist) <span class="op">&lt;</span><span class="st"> </span><span class="dv">1000</span>)
<span class="co">#&gt; # A tibble: 285 x 4</span>
<span class="co">#&gt;         name.x            name.y .overlap .dist</span>
<span class="co">#&gt;          &lt;chr&gt;             &lt;chr&gt;    &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt;  1   rs2261631             P704P        0  -267</span>
<span class="co">#&gt;  2 rs570770556             POTEH        0  -912</span>
<span class="co">#&gt;  3 rs538163832             POTEH        0  -952</span>
<span class="co">#&gt;  4   rs9606135            TPTEP1        0  -421</span>
<span class="co">#&gt;  5  rs11912392 ANKRD62P1-PARP4P3        0   104</span>
<span class="co">#&gt;  6   rs8136454          BC038197        0   355</span>
<span class="co">#&gt;  7   rs5992556              XKR3        0  -455</span>
<span class="co">#&gt;  8 rs114101676              GAB4        0   473</span>
<span class="co">#&gt;  9  rs62236167             CECR7        0   261</span>
<span class="co">#&gt; 10   rs5747023             CECR1        0  -386</span>
<span class="co">#&gt; # ... with 275 more rows</span></code></pre></div>
</div>
<div id="input-data" class="section level3">
<h3 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input data</h3>
<p><code>valr</code> assigns common column names to facilitate comparisons between tbls. All tbls will have <code>chrom</code>, <code>start</code>, and <code>end</code> columns, and some tbls from multi-column formats will have additional pre-determined column names. See the <code><a href="../reference/read_bed.html">read_bed()</a></code> documentation for details.</p>
</div>
<div id="remote-databases" class="section level3">
<h3 class="hasAnchor">
<a href="#remote-databases" class="anchor"></a>Remote databases</h3>
<p>Remote databases can be accessed with <code><a href="../reference/db.html">db_ucsc()</a></code> (to access the UCSC Browser) and <code><a href="../reference/db.html">db_ensembl()</a></code> (to access Ensembl databases).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># access the `refGene` tbl on the `hg38` assembly.</span>
<span class="cf">if</span>(<span class="kw">require</span>(RMySQL)) {
  ucsc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/db.html">db_ucsc</a></span>(<span class="st">'hg38'</span>)
  <span class="kw">tbl</span>(ucsc, <span class="st">'refGene'</span>)
}</code></pre></div>
</div>
<div id="visual-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#visual-documentation" class="anchor"></a>Visual documentation</h3>
<p>The <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code> tool illustrates the results of operations in <code>valr</code>, similar to those found in the <code>BEDtools</code> documentation. This glyph shows the result of intersecting <code>x</code> and <code>y</code> intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw">tribble</span>(
  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end,
  <span class="st">'chr1'</span>, <span class="dv">25</span>,     <span class="dv">50</span>,
  <span class="st">'chr1'</span>, <span class="dv">100</span>,    <span class="dv">125</span>
)

y &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw">tribble</span>(
  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end,
  <span class="st">'chr1'</span>, <span class="dv">30</span>,     <span class="dv">75</span>
)

<span class="kw"><a href="../reference/bed_glyph.html">bed_glyph</a></span>(<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y))</code></pre></div>
<p><img src="img/overview-intersect_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<p>And this glyph illustrates <code><a href="../reference/bed_merge.html">bed_merge()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw">tribble</span>(
  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end,
  <span class="st">'chr1'</span>,      <span class="dv">1</span>,      <span class="dv">50</span>,
  <span class="st">'chr1'</span>,      <span class="dv">10</span>,     <span class="dv">75</span>,
  <span class="st">'chr1'</span>,      <span class="dv">100</span>,    <span class="dv">120</span>
)

<span class="kw"><a href="../reference/bed_glyph.html">bed_glyph</a></span>(<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x))</code></pre></div>
<p><img src="img/overview-merge_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
<div id="reproducible-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#reproducible-reports" class="anchor"></a>Reproducible reports</h3>
<p><code>valr</code> can be used in RMarkdown documents to generate reproducible work-flows for data processing. Because <code>valr</code> is reasonably fast (see the <a href="#benchmarks">benchmarks</a>), we now use it in lieu of other tools for exploratory analysis of genomic data sets in R.</p>
<p>Command-line tools like <code>BEDtools</code> and <code>bedops</code> can be used in reproducible workflows (e.g., with <a href="https://bitbucket.org/snakemake/snakemake/wiki/Home"><code>snakemake</code></a>), but it is cumbersome to move from command-line tools to exploratory analysis and plotting software. <a href="https://pythonhosted.org/pybedtools/"><code>pybedtools</code></a> can be used within <code>ipython notebooks</code> to accomplish a similar goal, but others have pointed out <a href="https://www.r-bloggers.com/why-i-dont-like-jupyter-fka-ipython-notebook/">issues with this approach</a>, including clunky version control. Because RMarkdown files are text files, they are readily kept under version control. Moreover, new features in RStudio (e.g. notebook viewing) enable similar functionality to <code>ipython</code>.</p>
</div>
<div id="grouping-data" class="section level3">
<h3 class="hasAnchor">
<a href="#grouping-data" class="anchor"></a>Grouping data</h3>
<p>The <code>group_by</code> function in dplyr can be used to perform fuctions on subsets of single and multiple <code>data_frame</code>s. Functions in <code>valr</code> leverage grouping to enable a variety of comparisons. For example, intervals can be grouped by <code>strand</code> to perform comparisons among intervals on the same strand.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw">tribble</span>(
  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end, <span class="op">~</span>strand,
  <span class="st">'chr1'</span>, <span class="dv">1</span>,      <span class="dv">100</span>,  <span class="st">'+'</span>,
  <span class="st">'chr1'</span>, <span class="dv">50</span>,     <span class="dv">150</span>,  <span class="st">'+'</span>,
  <span class="st">'chr2'</span>, <span class="dv">100</span>,    <span class="dv">200</span>,  <span class="st">'-'</span>
)

y &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw">tribble</span>(
  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end, <span class="op">~</span>strand,
  <span class="st">'chr1'</span>, <span class="dv">50</span>,     <span class="dv">125</span>,  <span class="st">'+'</span>,
  <span class="st">'chr1'</span>, <span class="dv">50</span>,     <span class="dv">150</span>,  <span class="st">'-'</span>,
  <span class="st">'chr2'</span>, <span class="dv">50</span>,     <span class="dv">150</span>,  <span class="st">'+'</span>
)

<span class="co"># intersect tbls by strand</span>
x &lt;-<span class="st"> </span><span class="kw">group_by</span>(x, strand)
y &lt;-<span class="st"> </span><span class="kw">group_by</span>(y, strand)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 2 x 8</span>
<span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;    &lt;chr&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1       1   100        +      50   125        +       50</span>
<span class="co">#&gt; 2  chr1      50   150        +      50   125        +       75</span></code></pre></div>
<p>Comparisons between intervals on opposite strands are done using the <code><a href="../reference/flip_strands.html">flip_strands()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">group_by</span>(x, strand)

y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flip_strands.html">flip_strands</a></span>(y)
y &lt;-<span class="st"> </span><span class="kw">group_by</span>(y, strand)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 3 x 8</span>
<span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;    &lt;chr&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chr2     100   200        -      50   150        -       50</span>
<span class="co">#&gt; 2  chr1       1   100        +      50   150        +       50</span>
<span class="co">#&gt; 3  chr1      50   150        +      50   150        +      100</span></code></pre></div>
<p>Both single set (e.g. <code><a href="../reference/bed_merge.html">bed_merge()</a></code>) and multi set operations will respect groupings in the input intervals.</p>
</div>
<div id="column-specification" class="section level3">
<h3 class="hasAnchor">
<a href="#column-specification" class="anchor"></a>Column specification</h3>
<p>Columns in <code>BEDtools</code> are referred to by position:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># calculate the mean of column 6 for intervals in `b` that overlap with `a`</span>
<span class="ex">bedtools</span> map -a a.bed -b b.bed -c 6 -o mean</code></pre></div>
<p>In <code>valr</code>, columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate the mean and variance for a `value` column</span>
<span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(a, b, <span class="dt">.mean =</span> <span class="kw">mean</span>(value), <span class="dt">.var =</span> <span class="kw">var</span>(value))

<span class="co"># report concatenated and max values for merged intervals</span>
<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(a, <span class="dt">.concat =</span> <span class="kw"><a href="../reference/bed_map.html">concat</a></span>(value), <span class="dt">.max =</span> <span class="kw">max</span>(value))</code></pre></div>
</div>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<div id="meta-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#meta-analysis" class="anchor"></a>Meta-analysis</h3>
<p>This demonstration illustrates how to use <code>valr</code> tools to perform a “meta-analysis” of signals relative to genomic features. Here we to analyze the distribution of histone marks surrounding transcription start sites.</p>
<p>First we load libraries and relevant data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># `valr_example()` identifies the path of example files</span>
bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
genomefile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
genome &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(genomefile)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)</code></pre></div>
<p>Then we generate 1 bp intervals to represent transcription start sites (TSSs). We focus on <code>+</code> strand genes, but <code>-</code> genes are easily accomodated by filtering them and using <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> with <code>reverse</code>d window numbers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># generate 1 bp TSS intervals, `+` strand only</span>
tss &lt;-<span class="st"> </span>genes <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">filter</span>(strand <span class="op">==</span><span class="st"> '+'</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">end =</span> start <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)

<span class="co"># 1000 bp up and downstream</span>
region_size &lt;-<span class="st"> </span><span class="dv">1000</span>
<span class="co"># 50 bp windows</span>
win_size &lt;-<span class="st"> </span><span class="dv">50</span>

<span class="co"># add slop to the TSS, break into windows and add a group</span>
x &lt;-<span class="st"> </span>tss <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(genome, <span class="dt">both =</span> region_size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw"><a href="../reference/bed_makewindows.html">bed_makewindows</a></span>(genome, win_size)

x
<span class="co">#&gt; # A tibble: 13,530 x 7</span>
<span class="co">#&gt;    chrom    start      end      name score strand .win_id</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;int&gt;</span>
<span class="co">#&gt;  1 chr22 16161065 16161115 LINC00516     3      +       1</span>
<span class="co">#&gt;  2 chr22 16161115 16161165 LINC00516     3      +       2</span>
<span class="co">#&gt;  3 chr22 16161165 16161215 LINC00516     3      +       3</span>
<span class="co">#&gt;  4 chr22 16161215 16161265 LINC00516     3      +       4</span>
<span class="co">#&gt;  5 chr22 16161265 16161315 LINC00516     3      +       5</span>
<span class="co">#&gt;  6 chr22 16161315 16161365 LINC00516     3      +       6</span>
<span class="co">#&gt;  7 chr22 16161365 16161415 LINC00516     3      +       7</span>
<span class="co">#&gt;  8 chr22 16161415 16161465 LINC00516     3      +       8</span>
<span class="co">#&gt;  9 chr22 16161465 16161515 LINC00516     3      +       9</span>
<span class="co">#&gt; 10 chr22 16161515 16161565 LINC00516     3      +      10</span>
<span class="co">#&gt; # ... with 13,520 more rows</span></code></pre></div>
<p>Now we use the <code>.win_id</code> group with <code><a href="../reference/bed_map.html">bed_map()</a></code> to caluclate a sum by mapping <code>y</code> signals onto the intervals in <code>x</code>. These data are regrouped by <code>.win_id</code> and a summary with <code>mean</code> and <code>sd</code> values is calculated.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># map signals to TSS regions and calculate summary statistics.</span>
res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">win_sum =</span> <span class="kw">sum</span>(value, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">group_by</span>(.win_id) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">win_mean =</span> <span class="kw">mean</span>(win_sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),
            <span class="dt">win_sd =</span> <span class="kw">sd</span>(win_sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))

res
<span class="co">#&gt; # A tibble: 41 x 3</span>
<span class="co">#&gt;    .win_id win_mean    win_sd</span>
<span class="co">#&gt;      &lt;int&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1       1 100.8974  85.83423</span>
<span class="co">#&gt;  2       2 110.6829  81.13521</span>
<span class="co">#&gt;  3       3 122.9070  99.09635</span>
<span class="co">#&gt;  4       4 116.2800  96.30098</span>
<span class="co">#&gt;  5       5 116.3500 102.33773</span>
<span class="co">#&gt;  6       6 124.9048  95.08887</span>
<span class="co">#&gt;  7       7 122.9437  94.39792</span>
<span class="co">#&gt;  8       8 127.5946  91.47407</span>
<span class="co">#&gt;  9       9 130.2051  95.71309</span>
<span class="co">#&gt; 10      10 130.1220  88.82809</span>
<span class="co">#&gt; # ... with 31 more rows</span></code></pre></div>
<p>Finally, these summary statistics are used to construct a plot that illustrates histone density surrounding TSSs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)

x_labels &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span>region_size, region_size, <span class="dt">by =</span> win_size <span class="op">*</span><span class="st"> </span><span class="dv">5</span>)
x_breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">41</span>, <span class="dt">by =</span> <span class="dv">5</span>)

sd_limits &lt;-<span class="st"> </span><span class="kw">aes</span>(<span class="dt">ymax =</span> win_mean <span class="op">+</span><span class="st"> </span>win_sd, <span class="dt">ymin =</span> win_mean <span class="op">-</span><span class="st"> </span>win_sd)

<span class="kw">ggplot</span>(res, <span class="kw">aes</span>(<span class="dt">x =</span> .win_id, <span class="dt">y =</span> win_mean)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_pointrange</span>(sd_limits) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">labels =</span> x_labels, <span class="dt">breaks =</span> x_breaks) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">'Position (bp from TSS)'</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(<span class="st">'Signal'</span>) <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">'Human H3K4me3 signal near transcription start sites'</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_classic</span>()</code></pre></div>
<p><img src="img/overview-plot-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="interval-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-statistics" class="anchor"></a>Interval statistics</h3>
<p>Estimates of significance for interval overlaps can be obtained by combining <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code>, <code><a href="../reference/bed_random.html">bed_random()</a></code> and the <code>sample_</code> functions from <code>dplyr</code> with interval statistics in <code>valr</code>.</p>
<p>Here we examine the overlap of repeat classes in the human genome (on <code>chr22</code> only, for simplicity) using <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(purrr)
<span class="kw">library</span>(tidyr)

repeats &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.rmsk.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>) 
genome &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))


shuffle_intervals &lt;-<span class="st"> </span><span class="cf">function</span>(n, .data, genome) {
  <span class="kw">replicate</span>(n, <span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(.data, genome, <span class="dt">seed =</span> <span class="dv">1010486</span>), <span class="dt">simplify =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">bind_rows</span>(<span class="dt">.id =</span> <span class="st">'rep'</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">group_by</span>(rep) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">nest</span>()
}

shuffled &lt;-<span class="st"> </span><span class="kw">shuffle_intervals</span>(<span class="dt">n =</span> <span class="dv">100</span>, repeats, genome) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">jaccard =</span> data <span class="op">%&gt;%</span>
<span class="st">           </span><span class="kw">map</span>(bed_jaccard, repeats) <span class="op">%&gt;%</span>
<span class="st">           </span><span class="kw">map_dbl</span>(<span class="st">"jaccard"</span>))
  
shuffled
<span class="co">#&gt; # A tibble: 100 x 3</span>
<span class="co">#&gt;      rep                  data      jaccard</span>
<span class="co">#&gt;    &lt;chr&gt;                &lt;list&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt;  1     1 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  2     2 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  3     3 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  4     4 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  5     5 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  6     6 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  7     7 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  8     8 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt;  9     9 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt; 10    10 &lt;tibble [10,000 x 6]&gt; 0.0004355887</span>
<span class="co">#&gt; # ... with 90 more rows</span></code></pre></div>
</div>
</div>
<div id="benchmarks" class="section level2">
<h2 class="hasAnchor">
<a href="#benchmarks" class="anchor"></a>Benchmarks</h2>
<p>Certain algorithms in <code>valr</code> were implemented in <code>Rcpp</code> to enable fluid interactive analysis. This graph illustrates the timing of functions in <code>valr</code> for two sets of 1 million random 1 kilobase intervals from the hg19 genome.</p>
<p><img src="img/overview-benchmarks-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div id="api" class="section level2">
<h2 class="hasAnchor">
<a href="#api" class="anchor"></a>API</h2>
<p>Function names are similar to their their <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> counterparts, with some additions.</p>
<div id="reading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-data" class="anchor"></a>Reading data</h3>
<ul>
<li><p>BED and related files are read with <code><a href="../reference/read_bed.html">read_bed()</a></code>, <code><a href="../reference/read_bed.html">read_bed12()</a></code>, <code><a href="../reference/read_bed.html">read_bedgraph()</a></code>, <code><a href="../reference/read_bed.html">read_narrowpeak()</a></code> and <code><a href="../reference/read_bed.html">read_broadpeak()</a></code>.</p></li>
<li><p>Genome files containing chromosome name and size information are loaded with <code><a href="../reference/read_genome.html">read_genome()</a></code>.</p></li>
<li><p>VCF files are loaded with <code><a href="../reference/read_vcf.html">read_vcf()</a></code>.</p></li>
</ul>
</div>
<div id="transforming-single-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#transforming-single-interval-sets" class="anchor"></a>Transforming single interval sets</h3>
<ul>
<li><p>Intervals are ordered with <code>dplyr::arrange()</code>. See <code><a href="../reference/sorting.html">?sorting</a></code> for examples.</p></li>
<li><p>Interval coordinates are adjusted with <code><a href="../reference/bed_slop.html">bed_slop()</a></code> and <code><a href="../reference/bed_shift.html">bed_shift()</a></code>, and new flanking intervals are created with <code><a href="../reference/bed_flank.html">bed_flank()</a></code>.</p></li>
<li><p>Nearby intervals are combined with <code><a href="../reference/bed_merge.html">bed_merge()</a></code> and identified (but not merged) with <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>.</p></li>
<li><p>Intervals not covered by a query are created with <code><a href="../reference/bed_complement.html">bed_complement()</a></code>.</p></li>
</ul>
</div>
<div id="comparing-multiple-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-multiple-interval-sets" class="anchor"></a>Comparing multiple interval sets</h3>
<ul>
<li><p>Find overlaps between two sets of intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>.</p></li>
<li><p>Apply functions to selected columns for overlapping intervals with <code><a href="../reference/bed_map.html">bed_map()</a></code>.</p></li>
<li><p>Remove intervals based on overlaps between two files with <code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>.</p></li>
<li><p>Find overlapping intervals within a window with <code><a href="../reference/bed_window.html">bed_window()</a></code>.</p></li>
<li><p>Find the closest intervals independent of overlaps with <code><a href="../reference/bed_closest.html">bed_closest()</a></code>.</p></li>
</ul>
</div>
<div id="randomizing-intervals" class="section level3">
<h3 class="hasAnchor">
<a href="#randomizing-intervals" class="anchor"></a>Randomizing intervals</h3>
<ul>
<li><p>Generate random intervals from an input genome with <code><a href="../reference/bed_random.html">bed_random()</a></code>.</p></li>
<li><p>Shuffle the coordinates of input intervals with <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code>.</p></li>
<li><p>Random sampling of input intervals is done with the <code>sample_</code> function family in <code>dplyr</code>.</p></li>
</ul>
</div>
<div id="interval-statistics-1" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-statistics-1" class="anchor"></a>Interval statistics</h3>
<ul>
<li><p>Quantify overlaps between two sets of intervals with <code><a href="../reference/bed_fisher.html">bed_fisher()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between two sets of intervals with <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> or <code><a href="../reference/bed_projection.html">bed_projection()</a></code>.</p></li>
</ul>
</div>
<div id="utilities" class="section level3">
<h3 class="hasAnchor">
<a href="#utilities" class="anchor"></a>Utilities</h3>
<ul>
<li><p>Visualize the actions of valr functions with <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code>.</p></li>
<li><p>Constrain intervals to a genome reference with <code><a href="../reference/bound_intervals.html">bound_intervals()</a></code>.</p></li>
<li><p>Subdivide intervals with <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code>.</p></li>
<li><p>Convert BED12 to BED6 format with <code><a href="../reference/bed12_to_exons.html">bed12_to_exons()</a></code>.</p></li>
<li><p>Calculate spacing between intervals with <code><a href="../reference/interval_spacing.html">interval_spacing()</a></code>.</p></li>
</ul>
</div>
</div>
<div id="related_work" class="section level2">
<h2 class="hasAnchor">
<a href="#related_work" class="anchor"></a>Related work</h2>
<ul>
<li><p>Command-line tools <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> and <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://CRAN.R-project.org/package=bedr">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a> and <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#why-valr">Why <code>valr</code>?</a></li>
      <li><a href="#getting-started">Getting started</a></li>
      <li><a href="#benchmarks">Benchmarks</a></li>
      <li><a href="#api">API</a></li>
      <li><a href="#related_work">Related work</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Jay Hesselberth, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
