<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>valr overview • valr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha256-YLGeXaapI0/5IgZopewRJcFXomhRMlYYjugPLSyNjTY=" crossorigin="anonymous">
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/4.3.1/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<!-- Bootstrap 4 JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.slim.min.js" integrity="sha256-pasqAKBDmFT4eHoN2ndd6lN370kFiGUFyTiUHWhU7k8=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha256-CjSoeELFOcH0/uxWu6mC/Vlrc1AARqbm/jiiImDGV3s=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.css">
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="valr overview">
<meta property="og:description" content="">
<meta property="og:image" content="http://valr.hesselberthlab.org/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-54095998-3"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-54095998-3');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    <a class="navbar-brand mb-0 h1" href="../index.html">valr</a>

    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <button type="button" class="badge badge-default d-none d-lg-block" data-toggle="tooltip" data-placement="bottom" title="" data-original-title="Released version">0.5.0.9000</button>

    <div id="navbar" class="collapse navbar-collapse">
      <ul class="navbar-nav ml-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/valr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true">Articles</a>
  <div class="dropdown-menu" aria-labelledby="navbarDropdown">
    <a class="dropdown-item" href="../articles/benchmarks.html">valr Benchmarks</a>
    <a class="dropdown-item" href="../articles/interval-stats.html">Interval statistics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="navbar-nav mr-auto">
<li class="nav-item">
  <a class="nav-link" href="https://github.com/rnabioco/valr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control mr-sm-2" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</form>
      
    </div>
  </div>
</nav><div class="container template-article">


<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header pb-2 mt-4 mb-2 border-bottom toc-ignore">
      <h1 class="display-4" data-toc-skip>valr overview</h1>
            
            <h4 class="date">2019-08-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/valr/blob/master/vignettes/valr.Rmd"><code>vignettes/valr.Rmd</code></a></small>
      <div class="d-none name"><code>valr.Rmd</code></div>

    </div>

    
    
<div id="why-valr" class="section level2">
<h2 class="hasAnchor">
<a href="#why-valr" class="anchor"></a>Why <code>valr</code>?</h2>
<p><strong>Why another tool set for interval manipulations?</strong> There are several <a href="#related_work">other software packages</a> available for genome interval analysis. However, based on our experiences teaching genome analysis, we were motivated to develop a toolset that:</p>
<ul>
<li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the “modern R” syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code><a href="https://dplyr.tidyverse.org/reference/reexports.html">%&gt;%</a></code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul>
<p><code>valr</code> can currently be used for analysis of pre-processed data in BED and related formats. We plan to support BAM and VCF files soon via tabix indexes.</p>
<div id="familiar-tools-natively-in-r" class="section level3">
<h3 class="hasAnchor">
<a href="#familiar-tools-natively-in-r" class="anchor"></a>Familiar tools, natively in R</h3>
<p>The functions in <code>valr</code> have similar names to their <code>BEDtools</code> counterparts, and so will be familiar to users coming from the <code>BEDtools</code> suite. Similar to <a href="https://daler.github.io/pybedtools/#why-pybedtools"><code>pybedtools</code></a>, <code>valr</code> has a terse syntax:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(valr)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a>snps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.snps147.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>)</span>
<span id="cb1-5"><a href="#cb1-5"></a>genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>), <span class="dt">n_fields =</span> <span class="dv">6</span>)</span>
<span id="cb1-6"><a href="#cb1-6"></a></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co"># find snps in intergenic regions</span></span>
<span id="cb1-8"><a href="#cb1-8"></a>intergenic &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(snps, genes)</span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co"># distance from intergenic snps to nearest gene</span></span>
<span id="cb1-10"><a href="#cb1-10"></a>nearby &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(intergenic, genes)</span>
<span id="cb1-11"><a href="#cb1-11"></a></span>
<span id="cb1-12"><a href="#cb1-12"></a>nearby <span class="op">%&gt;%</span></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">starts_with</a></span>(<span class="st">'name'</span>), .overlap, .dist) <span class="op">%&gt;%</span></span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(.dist) <span class="op">&lt;</span><span class="st"> </span><span class="dv">1000</span>)</span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="co">#&gt; # A tibble: 285 x 4</span></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="co">#&gt;    name.x      name.y            .overlap .dist</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="co">#&gt;    &lt;chr&gt;       &lt;chr&gt;                &lt;int&gt; &lt;int&gt;</span></span>
<span id="cb1-18"><a href="#cb1-18"></a><span class="co">#&gt;  1 rs2261631   P704P                    0  -268</span></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co">#&gt;  2 rs570770556 POTEH                    0  -913</span></span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="co">#&gt;  3 rs538163832 POTEH                    0  -953</span></span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="co">#&gt;  4 rs9606135   TPTEP1                   0  -422</span></span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="co">#&gt;  5 rs11912392  ANKRD62P1-PARP4P3        0   105</span></span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="co">#&gt;  6 rs8136454   BC038197                 0   356</span></span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="co">#&gt;  7 rs5992556   XKR3                     0  -456</span></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="co">#&gt;  8 rs114101676 GAB4                     0   474</span></span>
<span id="cb1-26"><a href="#cb1-26"></a><span class="co">#&gt;  9 rs62236167  CECR7                    0   262</span></span>
<span id="cb1-27"><a href="#cb1-27"></a><span class="co">#&gt; 10 rs5747023   CECR1                    0  -387</span></span>
<span id="cb1-28"><a href="#cb1-28"></a><span class="co">#&gt; # … with 275 more rows</span></span></code></pre></div>
</div>
<div id="input-data" class="section level3">
<h3 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input data</h3>
<p><code>valr</code> assigns common column names to facilitate comparisons between tbls. All tbls will have <code>chrom</code>, <code>start</code>, and <code>end</code> columns, and some tbls from multi-column formats will have additional pre-determined column names. See the <code><a href="../reference/read_bed.html">read_bed()</a></code> documentation for details.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>bed_file &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">"3fields.bed.gz"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bed_file) <span class="co"># accepts filepaths or URLs</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; # A tibble: 10 x 3</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#&gt;    chrom  start    end</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co">#&gt;  1 chr1   11873  14409</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt;  2 chr1   14361  19759</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">#&gt;  3 chr1   14406  29370</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#&gt;  4 chr1   34610  36081</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="co">#&gt;  5 chr1   69090  70008</span></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#&gt;  6 chr1  134772 140566</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#&gt;  7 chr1  321083 321115</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#&gt;  8 chr1  321145 321207</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">#&gt;  9 chr1  322036 326938</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">#&gt; 10 chr1  327545 328439</span></span></code></pre></div>
<p><code>valr</code> can also operate on BED-like data.frames already constructed in R, provided that columns named <code>chrom</code>, <code>start</code> and <code>end</code> are present. New tbls can also be contructed using <code><a href="../reference/tbl_interval.html">trbl_interval()</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>bed &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tbl_interval.html">trbl_interval</a></span>(</span>
<span id="cb3-2"><a href="#cb3-2"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start,  <span class="op">~</span>end, </span>
<span id="cb3-3"><a href="#cb3-3"></a>  <span class="st">"chr1"</span>, <span class="dv">1657492</span>, <span class="dv">2657492</span>, </span>
<span id="cb3-4"><a href="#cb3-4"></a>  <span class="st">"chr2"</span>, <span class="dv">2501324</span>, <span class="dv">3094650</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>)</span>
<span id="cb3-6"><a href="#cb3-6"></a></span>
<span id="cb3-7"><a href="#cb3-7"></a>bed</span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; # A tibble: 2 x 3</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt;   chrom   start     end</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt; 1 chr1  1657492 2657492</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#&gt; 2 chr2  2501324 3094650</span></span></code></pre></div>
</div>
<div id="interval-coordinates" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-coordinates" class="anchor"></a>Interval coordinates</h3>
<p><code>valr</code> adheres to the BED <a href="http://genome.ucsc.edu/FAQ/FAQformat#format1">format</a> which specifies that the start position for an interval is zero based and the end position is one-based. The first position in a chromosome is 0. The end position for a chromosome is one position passed the last base, and is not included in the interval. For example:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># a chromosome 100 basepairs in length</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>chrom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tbl_interval.html">trbl_interval</a></span>(</span>
<span id="cb4-3"><a href="#cb4-3"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end, </span>
<span id="cb4-4"><a href="#cb4-4"></a>  <span class="st">"chr1"</span>, <span class="dv">0</span>,      <span class="dv">100</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>)</span>
<span id="cb4-6"><a href="#cb4-6"></a></span>
<span id="cb4-7"><a href="#cb4-7"></a>chrom</span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">#&gt; # A tibble: 1 x 3</span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">#&gt;   chrom start   end</span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">#&gt; 1 chr1      0   100</span></span>
<span id="cb4-12"><a href="#cb4-12"></a></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co"># single basepair intervals</span></span>
<span id="cb4-14"><a href="#cb4-14"></a>bases &lt;-<span class="st"> </span><span class="kw"><a href="../reference/tbl_interval.html">trbl_interval</a></span>(</span>
<span id="cb4-15"><a href="#cb4-15"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end, </span>
<span id="cb4-16"><a href="#cb4-16"></a>  <span class="st">"chr1"</span>, <span class="dv">0</span>,      <span class="dv">1</span>, <span class="co"># first base of chromosome</span></span>
<span id="cb4-17"><a href="#cb4-17"></a>  <span class="st">"chr1"</span>, <span class="dv">1</span>,      <span class="dv">2</span>,  <span class="co"># second base of chromosome</span></span>
<span id="cb4-18"><a href="#cb4-18"></a>  <span class="st">"chr1"</span>, <span class="dv">99</span>,     <span class="dv">100</span> <span class="co"># last base of chromosome</span></span>
<span id="cb4-19"><a href="#cb4-19"></a>)</span>
<span id="cb4-20"><a href="#cb4-20"></a></span>
<span id="cb4-21"><a href="#cb4-21"></a>bases</span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="co">#&gt; # A tibble: 3 x 3</span></span>
<span id="cb4-23"><a href="#cb4-23"></a><span class="co">#&gt;   chrom start   end</span></span>
<span id="cb4-24"><a href="#cb4-24"></a><span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span></span>
<span id="cb4-25"><a href="#cb4-25"></a><span class="co">#&gt; 1 chr1      0     1</span></span>
<span id="cb4-26"><a href="#cb4-26"></a><span class="co">#&gt; 2 chr1      1     2</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="co">#&gt; 3 chr1     99   100</span></span></code></pre></div>
</div>
<div id="remote-databases" class="section level3">
<h3 class="hasAnchor">
<a href="#remote-databases" class="anchor"></a>Remote databases</h3>
<p>Remote databases can be accessed with <code><a href="../reference/db.html">db_ucsc()</a></code> (to access the UCSC Browser) and <code><a href="../reference/db.html">db_ensembl()</a></code> (to access Ensembl databases).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># access the `refGene` tbl on the `hg38` assembly.</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="cf">if</span>(<span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(RMySQL)) {</span>
<span id="cb5-3"><a href="#cb5-3"></a>  ucsc &lt;-<span class="st"> </span><span class="kw"><a href="../reference/db.html">db_ucsc</a></span>(<span class="st">'hg38'</span>)</span>
<span id="cb5-4"><a href="#cb5-4"></a>  <span class="kw"><a href="https://dplyr.tidyverse.org/reference/tbl.html">tbl</a></span>(ucsc, <span class="st">'refGene'</span>)</span>
<span id="cb5-5"><a href="#cb5-5"></a>}</span></code></pre></div>
</div>
<div id="visual-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#visual-documentation" class="anchor"></a>Visual documentation</h3>
<p>The <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code> tool illustrates the results of operations in <code>valr</code>, similar to those found in the <code>BEDtools</code> documentation. This glyph shows the result of intersecting <code>x</code> and <code>y</code> intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>x &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(</span>
<span id="cb6-2"><a href="#cb6-2"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end,</span>
<span id="cb6-3"><a href="#cb6-3"></a>  <span class="st">'chr1'</span>, <span class="dv">25</span>,     <span class="dv">50</span>,</span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="st">'chr1'</span>, <span class="dv">100</span>,    <span class="dv">125</span></span>
<span id="cb6-5"><a href="#cb6-5"></a>)</span>
<span id="cb6-6"><a href="#cb6-6"></a></span>
<span id="cb6-7"><a href="#cb6-7"></a>y &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(</span>
<span id="cb6-8"><a href="#cb6-8"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end,</span>
<span id="cb6-9"><a href="#cb6-9"></a>  <span class="st">'chr1'</span>, <span class="dv">30</span>,     <span class="dv">75</span></span>
<span id="cb6-10"><a href="#cb6-10"></a>)</span>
<span id="cb6-11"><a href="#cb6-11"></a></span>
<span id="cb6-12"><a href="#cb6-12"></a><span class="kw"><a href="../reference/bed_glyph.html">bed_glyph</a></span>(<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y))</span></code></pre></div>
<p><img src="valr_files/figure-html/intersect_glyph-1.png" width="700" style="display: block; margin: auto;"></p>
<p>And this glyph illustrates <code><a href="../reference/bed_merge.html">bed_merge()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>x &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(</span>
<span id="cb7-2"><a href="#cb7-2"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end,</span>
<span id="cb7-3"><a href="#cb7-3"></a>  <span class="st">'chr1'</span>,      <span class="dv">1</span>,      <span class="dv">50</span>,</span>
<span id="cb7-4"><a href="#cb7-4"></a>  <span class="st">'chr1'</span>,      <span class="dv">10</span>,     <span class="dv">75</span>,</span>
<span id="cb7-5"><a href="#cb7-5"></a>  <span class="st">'chr1'</span>,      <span class="dv">100</span>,    <span class="dv">120</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>)</span>
<span id="cb7-7"><a href="#cb7-7"></a></span>
<span id="cb7-8"><a href="#cb7-8"></a><span class="kw"><a href="../reference/bed_glyph.html">bed_glyph</a></span>(<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x))</span></code></pre></div>
<p><img src="valr_files/figure-html/merge_glyph-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div id="reproducible-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#reproducible-reports" class="anchor"></a>Reproducible reports</h3>
<p><code>valr</code> can be used in RMarkdown documents to generate reproducible work-flows for data processing. Because <code>valr</code> is reasonably fast (see the <a href="#benchmarks">benchmarks</a>), we now use it in lieu of other tools for exploratory analysis of genomic data sets in R.</p>
<p>Command-line tools like <code>BEDtools</code> and <code>bedops</code> can be used in reproducible workflows (e.g., with <a href="https://bitbucket.org/snakemake/snakemake/wiki/Home"><code>snakemake</code></a>), but it is cumbersome to move from command-line tools to exploratory analysis and plotting software. <a href="https://pythonhosted.org/pybedtools/"><code>pybedtools</code></a> can be used within <code>ipython notebooks</code> to accomplish a similar goal, but others have pointed out <a href="https://www.r-bloggers.com/why-i-dont-like-jupyter-fka-ipython-notebook/">issues with this approach</a>, including clunky version control. Because RMarkdown files are text files, they are readily kept under version control. Moreover, new features in RStudio (e.g. notebook viewing) enable similar functionality to <code>ipython</code>.</p>
</div>
<div id="grouping-data" class="section level3">
<h3 class="hasAnchor">
<a href="#grouping-data" class="anchor"></a>Grouping data</h3>
<p>The <code>group_by</code> function in dplyr can be used to perform fuctions on subsets of single and multiple <code>data_frame</code>s. Functions in <code>valr</code> leverage grouping to enable a variety of comparisons. For example, intervals can be grouped by <code>strand</code> to perform comparisons among intervals on the same strand.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>x &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(</span>
<span id="cb8-2"><a href="#cb8-2"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end, <span class="op">~</span>strand,</span>
<span id="cb8-3"><a href="#cb8-3"></a>  <span class="st">'chr1'</span>, <span class="dv">1</span>,      <span class="dv">100</span>,  <span class="st">'+'</span>,</span>
<span id="cb8-4"><a href="#cb8-4"></a>  <span class="st">'chr1'</span>, <span class="dv">50</span>,     <span class="dv">150</span>,  <span class="st">'+'</span>,</span>
<span id="cb8-5"><a href="#cb8-5"></a>  <span class="st">'chr2'</span>, <span class="dv">100</span>,    <span class="dv">200</span>,  <span class="st">'-'</span></span>
<span id="cb8-6"><a href="#cb8-6"></a>)</span>
<span id="cb8-7"><a href="#cb8-7"></a></span>
<span id="cb8-8"><a href="#cb8-8"></a>y &lt;-<span class="st"> </span>tibble<span class="op">::</span><span class="kw"><a href="https://tibble.tidyverse.org/reference/tribble.html">tribble</a></span>(</span>
<span id="cb8-9"><a href="#cb8-9"></a>  <span class="op">~</span>chrom, <span class="op">~</span>start, <span class="op">~</span>end, <span class="op">~</span>strand,</span>
<span id="cb8-10"><a href="#cb8-10"></a>  <span class="st">'chr1'</span>, <span class="dv">50</span>,     <span class="dv">125</span>,  <span class="st">'+'</span>,</span>
<span id="cb8-11"><a href="#cb8-11"></a>  <span class="st">'chr1'</span>, <span class="dv">50</span>,     <span class="dv">150</span>,  <span class="st">'-'</span>,</span>
<span id="cb8-12"><a href="#cb8-12"></a>  <span class="st">'chr2'</span>, <span class="dv">50</span>,     <span class="dv">150</span>,  <span class="st">'+'</span></span>
<span id="cb8-13"><a href="#cb8-13"></a>)</span>
<span id="cb8-14"><a href="#cb8-14"></a></span>
<span id="cb8-15"><a href="#cb8-15"></a><span class="co"># intersect tbls by strand</span></span>
<span id="cb8-16"><a href="#cb8-16"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(x, strand)</span>
<span id="cb8-17"><a href="#cb8-17"></a>y &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(y, strand)</span>
<span id="cb8-18"><a href="#cb8-18"></a></span>
<span id="cb8-19"><a href="#cb8-19"></a><span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)</span>
<span id="cb8-20"><a href="#cb8-20"></a><span class="co">#&gt; # A tibble: 2 x 8</span></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span></span>
<span id="cb8-22"><a href="#cb8-22"></a><span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt;</span></span>
<span id="cb8-23"><a href="#cb8-23"></a><span class="co">#&gt; 1 chr1        1   100 +             50   125 +              50</span></span>
<span id="cb8-24"><a href="#cb8-24"></a><span class="co">#&gt; 2 chr1       50   150 +             50   125 +              75</span></span></code></pre></div>
<p>Comparisons between intervals on opposite strands are done using the <code><a href="../reference/flip_strands.html">flip_strands()</a></code> function:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>x &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(x, strand)</span>
<span id="cb9-2"><a href="#cb9-2"></a></span>
<span id="cb9-3"><a href="#cb9-3"></a>y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/flip_strands.html">flip_strands</a></span>(y)</span>
<span id="cb9-4"><a href="#cb9-4"></a>y &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(y, strand)</span>
<span id="cb9-5"><a href="#cb9-5"></a></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)</span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co">#&gt; # A tibble: 3 x 8</span></span>
<span id="cb9-8"><a href="#cb9-8"></a><span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co">#&gt;   &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;       &lt;int&gt;</span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co">#&gt; 1 chr1        1   100 +             50   150 +              50</span></span>
<span id="cb9-11"><a href="#cb9-11"></a><span class="co">#&gt; 2 chr1       50   150 +             50   150 +             100</span></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co">#&gt; 3 chr2      100   200 -             50   150 -              50</span></span></code></pre></div>
<p>Both single set (e.g. <code><a href="../reference/bed_merge.html">bed_merge()</a></code>) and multi set operations will respect groupings in the input intervals.</p>
</div>
<div id="column-specification" class="section level3">
<h3 class="hasAnchor">
<a href="#column-specification" class="anchor"></a>Column specification</h3>
<p>Columns in <code>BEDtools</code> are referred to by position:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># calculate the mean of column 6 for intervals in `b` that overlap with `a`</span></span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="ex">bedtools</span> map -a a.bed -b b.bed -c 6 -o mean</span></code></pre></div>
<p>In <code>valr</code>, columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="co"># calculate the mean and variance for a `value` column</span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(a, b, <span class="dt">.mean =</span> <span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(value), <span class="dt">.var =</span> <span class="kw"><a href="https://rdrr.io/r/stats/cor.html">var</a></span>(value))</span>
<span id="cb11-3"><a href="#cb11-3"></a></span>
<span id="cb11-4"><a href="#cb11-4"></a><span class="co"># report concatenated and max values for merged intervals</span></span>
<span id="cb11-5"><a href="#cb11-5"></a><span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(a, <span class="dt">.concat =</span> <span class="kw"><a href="../reference/bed_map.html">concat</a></span>(value), <span class="dt">.max =</span> <span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(value))</span></code></pre></div>
</div>
</div>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h2>
<div id="meta-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#meta-analysis" class="anchor"></a>Meta-analysis</h3>
<p>This demonstration illustrates how to use <code>valr</code> tools to perform a “meta-analysis” of signals relative to genomic features. Here we to analyze the distribution of histone marks surrounding transcription start sites.</p>
<p>First we load libraries and relevant data.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># `valr_example()` identifies the path of example files</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)</span>
<span id="cb12-3"><a href="#cb12-3"></a>genomefile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>)</span>
<span id="cb12-4"><a href="#cb12-4"></a>bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)</span>
<span id="cb12-5"><a href="#cb12-5"></a></span>
<span id="cb12-6"><a href="#cb12-6"></a>genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)</span>
<span id="cb12-7"><a href="#cb12-7"></a>genome &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(genomefile)</span>
<span id="cb12-8"><a href="#cb12-8"></a>y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)</span></code></pre></div>
<p>Then we generate 1 bp intervals to represent transcription start sites (TSSs). We focus on <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> strand genes, but <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> genes are easily accomodated by filtering them and using <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> with <code>reverse</code>d window numbers.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># generate 1 bp TSS intervals, `+` strand only</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>tss &lt;-<span class="st"> </span>genes <span class="op">%&gt;%</span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(strand <span class="op">==</span><span class="st"> '+'</span>) <span class="op">%&gt;%</span></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">end =</span> start <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb13-5"><a href="#cb13-5"></a></span>
<span id="cb13-6"><a href="#cb13-6"></a><span class="co"># 1000 bp up and downstream</span></span>
<span id="cb13-7"><a href="#cb13-7"></a>region_size &lt;-<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb13-8"><a href="#cb13-8"></a><span class="co"># 50 bp windows</span></span>
<span id="cb13-9"><a href="#cb13-9"></a>win_size &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb13-10"><a href="#cb13-10"></a></span>
<span id="cb13-11"><a href="#cb13-11"></a><span class="co"># add slop to the TSS, break into windows and add a group</span></span>
<span id="cb13-12"><a href="#cb13-12"></a>x &lt;-<span class="st"> </span>tss <span class="op">%&gt;%</span></span>
<span id="cb13-13"><a href="#cb13-13"></a><span class="st">  </span><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(genome, <span class="dt">both =</span> region_size) <span class="op">%&gt;%</span></span>
<span id="cb13-14"><a href="#cb13-14"></a><span class="st">  </span><span class="kw"><a href="../reference/bed_makewindows.html">bed_makewindows</a></span>(win_size)</span>
<span id="cb13-15"><a href="#cb13-15"></a></span>
<span id="cb13-16"><a href="#cb13-16"></a>x</span>
<span id="cb13-17"><a href="#cb13-17"></a><span class="co">#&gt; # A tibble: 13,530 x 7</span></span>
<span id="cb13-18"><a href="#cb13-18"></a><span class="co">#&gt;    chrom    start      end name      score strand .win_id</span></span>
<span id="cb13-19"><a href="#cb13-19"></a><span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;    &lt;int&gt;</span></span>
<span id="cb13-20"><a href="#cb13-20"></a><span class="co">#&gt;  1 chr22 16161065 16161115 LINC00516 3     +            1</span></span>
<span id="cb13-21"><a href="#cb13-21"></a><span class="co">#&gt;  2 chr22 16161115 16161165 LINC00516 3     +            2</span></span>
<span id="cb13-22"><a href="#cb13-22"></a><span class="co">#&gt;  3 chr22 16161165 16161215 LINC00516 3     +            3</span></span>
<span id="cb13-23"><a href="#cb13-23"></a><span class="co">#&gt;  4 chr22 16161215 16161265 LINC00516 3     +            4</span></span>
<span id="cb13-24"><a href="#cb13-24"></a><span class="co">#&gt;  5 chr22 16161265 16161315 LINC00516 3     +            5</span></span>
<span id="cb13-25"><a href="#cb13-25"></a><span class="co">#&gt;  6 chr22 16161315 16161365 LINC00516 3     +            6</span></span>
<span id="cb13-26"><a href="#cb13-26"></a><span class="co">#&gt;  7 chr22 16161365 16161415 LINC00516 3     +            7</span></span>
<span id="cb13-27"><a href="#cb13-27"></a><span class="co">#&gt;  8 chr22 16161415 16161465 LINC00516 3     +            8</span></span>
<span id="cb13-28"><a href="#cb13-28"></a><span class="co">#&gt;  9 chr22 16161465 16161515 LINC00516 3     +            9</span></span>
<span id="cb13-29"><a href="#cb13-29"></a><span class="co">#&gt; 10 chr22 16161515 16161565 LINC00516 3     +           10</span></span>
<span id="cb13-30"><a href="#cb13-30"></a><span class="co">#&gt; # … with 13,520 more rows</span></span></code></pre></div>
<p>Now we use the <code>.win_id</code> group with <code><a href="../reference/bed_map.html">bed_map()</a></code> to caluclate a sum by mapping <code>y</code> signals onto the intervals in <code>x</code>. These data are regrouped by <code>.win_id</code> and a summary with <code>mean</code> and <code>sd</code> values is calculated.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># map signals to TSS regions and calculate summary statistics.</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">win_sum =</span> <span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(value, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)) <span class="op">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(.win_id) <span class="op">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="dt">win_mean =</span> <span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(win_sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>),</span>
<span id="cb14-5"><a href="#cb14-5"></a>            <span class="dt">win_sd =</span> <span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(win_sum, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>))</span>
<span id="cb14-6"><a href="#cb14-6"></a></span>
<span id="cb14-7"><a href="#cb14-7"></a>res</span>
<span id="cb14-8"><a href="#cb14-8"></a><span class="co">#&gt; # A tibble: 41 x 3</span></span>
<span id="cb14-9"><a href="#cb14-9"></a><span class="co">#&gt;    .win_id win_mean win_sd</span></span>
<span id="cb14-10"><a href="#cb14-10"></a><span class="co">#&gt;      &lt;int&gt;    &lt;dbl&gt;  &lt;dbl&gt;</span></span>
<span id="cb14-11"><a href="#cb14-11"></a><span class="co">#&gt;  1       1     101.   85.8</span></span>
<span id="cb14-12"><a href="#cb14-12"></a><span class="co">#&gt;  2       2     111.   81.1</span></span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co">#&gt;  3       3     123.   99.1</span></span>
<span id="cb14-14"><a href="#cb14-14"></a><span class="co">#&gt;  4       4     116.   96.3</span></span>
<span id="cb14-15"><a href="#cb14-15"></a><span class="co">#&gt;  5       5     116.  102. </span></span>
<span id="cb14-16"><a href="#cb14-16"></a><span class="co">#&gt;  6       6     125.   95.1</span></span>
<span id="cb14-17"><a href="#cb14-17"></a><span class="co">#&gt;  7       7     123.   94.4</span></span>
<span id="cb14-18"><a href="#cb14-18"></a><span class="co">#&gt;  8       8     128.   91.5</span></span>
<span id="cb14-19"><a href="#cb14-19"></a><span class="co">#&gt;  9       9     130.   95.7</span></span>
<span id="cb14-20"><a href="#cb14-20"></a><span class="co">#&gt; 10      10     130.   88.8</span></span>
<span id="cb14-21"><a href="#cb14-21"></a><span class="co">#&gt; # … with 31 more rows</span></span></code></pre></div>
<p>Finally, these summary statistics are used to construct a plot that illustrates histone density surrounding TSSs.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</span>
<span id="cb15-2"><a href="#cb15-2"></a></span>
<span id="cb15-3"><a href="#cb15-3"></a>x_labels &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="op">-</span>region_size, region_size, <span class="dt">by =</span> win_size <span class="op">*</span><span class="st"> </span><span class="dv">5</span>)</span>
<span id="cb15-4"><a href="#cb15-4"></a>x_breaks &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">1</span>, <span class="dv">41</span>, <span class="dt">by =</span> <span class="dv">5</span>)</span>
<span id="cb15-5"><a href="#cb15-5"></a></span>
<span id="cb15-6"><a href="#cb15-6"></a>sd_limits &lt;-<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">ymax =</span> win_mean <span class="op">+</span><span class="st"> </span>win_sd, <span class="dt">ymin =</span> win_mean <span class="op">-</span><span class="st"> </span>win_sd)</span>
<span id="cb15-7"><a href="#cb15-7"></a></span>
<span id="cb15-8"><a href="#cb15-8"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(res, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> .win_id, <span class="dt">y =</span> win_mean)) <span class="op">+</span></span>
<span id="cb15-9"><a href="#cb15-9"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html">geom_pointrange</a></span>(sd_limits) <span class="op">+</span><span class="st"> </span></span>
<span id="cb15-10"><a href="#cb15-10"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_continuous</a></span>(<span class="dt">labels =</span> x_labels, <span class="dt">breaks =</span> x_breaks) <span class="op">+</span><span class="st"> </span></span>
<span id="cb15-11"><a href="#cb15-11"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">'Position (bp from TSS)'</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">'Signal'</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb15-12"><a href="#cb15-12"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">'Human H3K4me3 signal near transcription start sites'</span>) <span class="op">+</span></span>
<span id="cb15-13"><a href="#cb15-13"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span>()</span></code></pre></div>
<p><img src="valr_files/figure-html/plot-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div id="api" class="section level2">
<h2 class="hasAnchor">
<a href="#api" class="anchor"></a>API</h2>
<p>Function names are similar to their their <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> counterparts, with some additions.</p>
<div id="data-types" class="section level3">
<h3 class="hasAnchor">
<a href="#data-types" class="anchor"></a>Data types</h3>
<ul>
<li>Create interval sets with <code><a href="../reference/tbl_interval.html">tbl_interval()</a></code> and <code><a href="../reference/tbl_genome.html">tbl_genome()</a></code>, which enforce strict column naming.</li>
</ul>
</div>
<div id="reading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reading-data" class="anchor"></a>Reading data</h3>
<ul>
<li><p>Read BED and related files with <code><a href="../reference/read_bed.html">read_bed()</a></code>, <code><a href="../reference/read_bed.html">read_bed12()</a></code>, <code><a href="../reference/read_bed.html">read_bedgraph()</a></code>, <code><a href="../reference/read_bed.html">read_narrowpeak()</a></code> and <code><a href="../reference/read_bed.html">read_broadpeak()</a></code>.</p></li>
<li><p>Read genome files containing chromosome name and size information with <code><a href="../reference/read_genome.html">read_genome()</a></code>.</p></li>
<li><p>Load VCF files with <code><a href="../reference/read_vcf.html">read_vcf()</a></code>.</p></li>
<li><p>Access remote databases with <code><a href="../reference/db.html">db_ucsc()</a></code> and <code><a href="../reference/db.html">db_ensembl()</a></code>.</p></li>
</ul>
</div>
<div id="transforming-single-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#transforming-single-interval-sets" class="anchor"></a>Transforming single interval sets</h3>
<ul>
<li><p>Adjust interval coordinates with <code><a href="../reference/bed_slop.html">bed_slop()</a></code> and <code><a href="../reference/bed_shift.html">bed_shift()</a></code>, and create new flanking intervals with <code><a href="../reference/bed_flank.html">bed_flank()</a></code>.</p></li>
<li><p>Combine nearby intervals with <code><a href="../reference/bed_merge.html">bed_merge()</a></code> and identify nearby intervals with <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>.</p></li>
<li><p>Generate intervals not covered by a query with <code><a href="../reference/bed_complement.html">bed_complement()</a></code>.</p></li>
<li><p>Order intervals with <code><a href="https://dplyr.tidyverse.org/reference/arrange.html">dplyr::arrange()</a></code>.</p></li>
</ul>
</div>
<div id="comparing-multiple-interval-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-multiple-interval-sets" class="anchor"></a>Comparing multiple interval sets</h3>
<ul>
<li><p>Find overlaps between sets of intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>.</p></li>
<li><p>Apply functions to overlapping sets of intervals with <code><a href="../reference/bed_map.html">bed_map()</a></code>.</p></li>
<li><p>Remove intervals based on overlaps with <code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>.</p></li>
<li><p>Find overlapping intervals within a window with <code><a href="../reference/bed_window.html">bed_window()</a></code>.</p></li>
<li><p>Find closest intervals independent of overlaps with <code><a href="../reference/bed_closest.html">bed_closest()</a></code>.</p></li>
</ul>
</div>
<div id="randomizing-intervals" class="section level3">
<h3 class="hasAnchor">
<a href="#randomizing-intervals" class="anchor"></a>Randomizing intervals</h3>
<ul>
<li><p>Generate random intervals with <code><a href="../reference/bed_random.html">bed_random()</a></code>.</p></li>
<li><p>Shuffle the coordinates of intervals with <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code>.</p></li>
<li><p>Sample input intervals with <code><a href="https://dplyr.tidyverse.org/reference/sample.html">dplyr::sample_n()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/sample.html">dplyr::sample_frac()</a></code>.</p></li>
</ul>
</div>
<div id="interval-statistics" class="section level3">
<h3 class="hasAnchor">
<a href="#interval-statistics" class="anchor"></a>Interval statistics</h3>
<ul>
<li><p>Calculate significance of overlaps between sets of intervals with <code><a href="../reference/bed_fisher.html">bed_fisher()</a></code> and <code><a href="../reference/bed_projection.html">bed_projection()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between sets of intervals with <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>.</p></li>
</ul>
</div>
<div id="utilities" class="section level3">
<h3 class="hasAnchor">
<a href="#utilities" class="anchor"></a>Utilities</h3>
<ul>
<li><p>Create features from BED12 files with <code><a href="../reference/create_introns.html">create_introns()</a></code>, <code><a href="../reference/create_tss.html">create_tss()</a></code>, <code><a href="../reference/create_utrs5.html">create_utrs5()</a></code>, and <code><a href="../reference/create_utrs3.html">create_utrs3()</a></code>.</p></li>
<li><p>Visualize the actions of valr functions with <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code>.</p></li>
<li><p>Constrain intervals to a genome reference with <code><a href="../reference/bound_intervals.html">bound_intervals()</a></code>.</p></li>
<li><p>Subdivide intervals with <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code>.</p></li>
<li><p>Convert BED12 to BED6 format with <code><a href="../reference/bed12_to_exons.html">bed12_to_exons()</a></code>.</p></li>
<li><p>Calculate spacing between intervals with <code><a href="../reference/interval_spacing.html">interval_spacing()</a></code>.</p></li>
</ul>
</div>
</div>
<div id="related-work" class="section level2">
<h2 class="hasAnchor">
<a href="#related-work" class="anchor"></a>Related work</h2>
<ul>
<li><p>Command-line tools <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a> and <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://CRAN.R-project.org/package=bedr">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a> and <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 d-none d-md-block" id="sidebar">
    <div class="sticky-top">
    </div>

        <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://hesselberthlab.org">Jay Hesselberth</a>, Kent Riemondy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    appId: 'TK8WB6UEB2',
    
    apiKey: 'c871e0fb692cad1373a189fc7fe65300',
    indexName: 'valr',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
