<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title><code>valr</code>: Genome interval arithmetic in R. valr</title><!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous"><script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous"><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <div class="container">
      <header><div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/valr-overview.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1><code>valr</code>: Genome interval arithmetic in R</h1>
                        <h4 class="author">Jay Hesselberth <a href="mailto:jay.hesselberth@gmail.com">jay.hesselberth@gmail.com</a></h4>
            
            <h4 class="date">2016-10-11</h4>
          </div>

    
    
<div id="overview" class="section level1">
<h1>Overview</h1>
<p><code>valr</code> provides tools to read and manipulate genome intervals and signals, similar to the standalone <a href="http://bedtools.readthedocs.org/en/latest/"><code>BEDtools</code></a> suite. <code>valr</code> enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a>/C++, and many methods take advantage of the speed and grouping capability provided by <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a>.</p>
<p>We provide several introductions to <code>valr</code>:</p>
<ul><li>This comprehensive <code>vignette</code> covering the core methods.</li>
<li>A tutorial that demonstrates how to use <code>valr</code> in &ldquo;real-world&rdquo; applications.</li>
<li>A <code>shiny</code> application that demonstrates interactive analysis of genome-scale data sets.</li>
</ul><div id="installation" class="section level2">
<h2>Installation</h2>
<p><code>valr</code> can be installed from github, and will be eventually deposited in CRAN.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">'jayhesselberth/valr'</span>)</code></pre></div>
</div>
<div id="comparison-to-other-tools" class="section level2">
<h2>Comparison to other tools</h2>
<p><strong>Why another tool set for interval manipulations?</strong> We already have <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a>, <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>, <a href="https://pythonhosted.org/pybedtools/">pybedtools</a>, <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>, <a href="https://cran.r-project.org/web/packages/bedr/index.html">bedr</a> and <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a>.</p>
<p>We were motivated to develop a toolset that:</p>
<ul><li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the &ldquo;modern R&rdquo; syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul><p>We anticipate <code>valr</code> will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as <code>tidyr</code> makes it easy to convert these to BED12.</p>
<p>Certain algorithms in <code>valr</code> were implemented in <code>Rcpp</code> (including <a href="https://github.com/jayhesselberth/valr/tree/master/src">intersect, merge, subtract, closest</a>) to enable fluid interactive analysis. See the <a href="#benchmarks">benchmarks</a> section for details.</p>
</div>
<div id="non-standard-evaluation" class="section level2">
<h2>Non-standard evaluation</h2>
<p>Several of the methods in <code>valr</code> use <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html">NSE</a> for an expressive syntax. Columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">mean =</span> <span class="kw">mean</span>(value), <span class="dt">var =</span> <span class="kw">var</span>(value))
<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">concat =</span> <span class="kw"><a href="../reference/bed_map.html">concat</a></span>(value), <span class="dt">max =</span> <span class="kw">max</span>(value))</code></pre></div>
</div>
<div id="getting-started" class="section level2">
<h2>Getting started</h2>
<p>Here is an example using <code>valr</code> that creates</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)</code></pre></div>
</div>
</div>
<div id="reading-data" class="section level1 tabset">
<h1>Reading data</h1>
<p><code>valr</code> has several methods to read interval data. These methods:</p>
<ul><li>Take local files or URLs as input.</li>
<li>Return a <code>data.frame</code> in <code>tibble::tbl_df</code> format.</li>
<li>Assign consistent <code>chrom</code>, <code>start</code> and <code>end</code> column names.</li>
<li>Use <a href="https://github.com/hadley/readr"><code>readr</code></a> for speed.</li>
<li>Coerce column types.</li>
</ul><p>The methods include:</p>
<ul><li><code><a href="../reference/read_bed.html">read_bed()</a></code>: read a BED3+ file</li>
<li><code><a href="../reference/read_bed.html">read_bed12()</a></code>: read a BED12 file</li>
<li><code><a href="../reference/read_bed.html">read_bedgraph()</a></code>: read a bedGraph file</li>
<li><code><a href="../reference/read_genome.html">read_genome()</a></code>: read a UCSC &ldquo;chrom size&rdquo; file</li>
<li><code><a href="../reference/read_vcf.html">read_vcf()</a></code>: read the Variant Call Format</li>
<li><code><a href="../reference/read_bed.html">read_narrowpeak()</a></code>: read narrowPeak files</li>
<li><code><a href="../reference/read_bed.html">read_broadpeak()</a></code>: read broadPeak files</li>
</ul><div id="bed-files" class="section level2">
<h2>BED files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'3fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 3</span>
<span class="co">#&gt;    chrom  start    end</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409</span>
<span class="co">#&gt; 2   chr1  14361  19759</span>
<span class="co">#&gt; 3   chr1  14406  29370</span>
<span class="co">#&gt; 4   chr1  34610  36081</span>
<span class="co">#&gt; 5   chr1  69090  70008</span>
<span class="co">#&gt; 6   chr1 134772 140566</span>
<span class="co">#&gt; 7   chr1 321083 321115</span>
<span class="co">#&gt; 8   chr1 321145 321207</span>
<span class="co">#&gt; 9   chr1 322036 326938</span>
<span class="co">#&gt; 10  chr1 327545 328439</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(<span class="dt">n_fields =</span> <span class="dv">6</span>, <span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'6fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 10 &times; 6</span>
<span class="co">#&gt;    chrom  start    end         name score strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;        &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  11873  14409      DDX11L1     3      +</span>
<span class="co">#&gt; 2   chr1  14361  19759       WASH7P    10      -</span>
<span class="co">#&gt; 3   chr1  14406  29370       WASH7P     7      -</span>
<span class="co">#&gt; 4   chr1  34610  36081      FAM138F     3      -</span>
<span class="co">#&gt; 5   chr1  69090  70008        OR4F5     1      +</span>
<span class="co">#&gt; 6   chr1 134772 140566    LOC729737     3      -</span>
<span class="co">#&gt; 7   chr1 321083 321115     DQ597235     1      +</span>
<span class="co">#&gt; 8   chr1 321145 321207     DQ599768     1      +</span>
<span class="co">#&gt; 9   chr1 322036 326938 LOC100133331     3      +</span>
<span class="co">#&gt; 10  chr1 327545 328439    LOC388312     1      +</span>
<span class="kw"><a href="../reference/read_bed.html">read_bed12</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'12fields.bed.gz'</span>))
<span class="co">#&gt; # A tibble: 3 &times; 12</span>
<span class="co">#&gt;   chrom   start     end      name score strand cds_start cds_end item_rgb</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;   &lt;int&gt;    &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1 4797973 4836816  testgene     1      +   4797973 4836816        .</span>
<span class="co">#&gt; 2 chr10 4848118 4880877 diffchrom     1      +   4848118 4880877        .</span>
<span class="co">#&gt; 3 chr20 5073253 5152630 negstrand     1      -   5073253 5152630        .</span>
<span class="co">#&gt; # ... with 3 more variables: exon_count &lt;int&gt;, exon_sizes &lt;chr&gt;,</span>
<span class="co">#&gt; #   exon_starts &lt;chr&gt;</span></code></pre></div>
</div>
<div id="genome-files" class="section level2">
<h2>Genome files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome_human &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_genome.html">read_genome</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hg19.chrom.sizes.gz'</span>))</code></pre></div>
</div>
<div id="bedgraph-files" class="section level2">
<h2>bedGraph files</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.bg.gz'</span>))
<span class="co">#&gt; # A tibble: 4 &times; 4</span>
<span class="co">#&gt;   chrom    start      end value</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr19 49302000 49302300 -1.00</span>
<span class="co">#&gt; 2 chr19 49302300 49302600 -0.75</span>
<span class="co">#&gt; 3 chr19 49302600 49302900 -0.50</span>
<span class="co">#&gt; 4 chr19 49302900 49303200 -0.25</span></code></pre></div>
</div>
<div id="vcf-files" class="section level2">
<h2>VCF files</h2>
<p><code><a href="../reference/read_vcf.html">read_vcf()</a></code> reads VCF files and assigns <code>chrom</code>, <code>start</code> and <code>end</code> columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the <code>REF</code> field in the original file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/read_vcf.html">read_vcf</a></span>(<span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'test.vcf.gz'</span>))
<span class="co">#&gt; # A tibble: 115 &times; 201</span>
<span class="co">#&gt;    CHROM   POS    ID                                          REF</span>
<span class="co">#&gt;    &lt;int&gt; &lt;int&gt; &lt;chr&gt;                                        &lt;chr&gt;</span>
<span class="co">#&gt; 1      1 10172     .                                       CCCTAA</span>
<span class="co">#&gt; 2      1 10390     . CCCCTAACCCCTAACCCTAACCCTAACCCTAACCCTAACCCTAA</span>
<span class="co">#&gt; 3      1 10397     .                                CCCCTAACCCTAA</span>
<span class="co">#&gt; 4      1 10403     .       ACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAC</span>
<span class="co">#&gt; 5      1 10409     .             ACCCTAACCCTAACCCTAACCCTAACCCTAAC</span>
<span class="co">#&gt; 6      1 10415     .                   ACCCTAACCCTAACCCTAACCCTAAC</span>
<span class="co">#&gt; 7      1 10421     .                         ACCCTAACCCTAACCCTAAC</span>
<span class="co">#&gt; 8      1 10428     .                                       CCCTAA</span>
<span class="co">#&gt; 9      1 10440     .                                            C</span>
<span class="co">#&gt; 10     1 10478     .                                            C</span>
<span class="co">#&gt; # ... with 105 more rows, and 197 more variables: ALT &lt;chr&gt;, QUAL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   FILTER &lt;chr&gt;, INFO &lt;chr&gt;, FORMAT &lt;chr&gt;, `101976-101976` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100920-100920` &lt;chr&gt;, `100231-100231` &lt;chr&gt;, `100232-100232` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100919-100919` &lt;chr&gt;, `101977-101977` &lt;chr&gt;, `100630-100630` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100640-100640` &lt;chr&gt;, `100631-100631` &lt;chr&gt;, `101583-101583` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101584-101584` &lt;chr&gt;, `101732-101732` &lt;chr&gt;, `101016-101016` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101708-101708` &lt;chr&gt;, `101730-101730` &lt;chr&gt;, `101582-101582` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101731-101731` &lt;chr&gt;, `101809-101809` &lt;chr&gt;, `101653-101653` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101652-101652` &lt;chr&gt;, `101806-101806` &lt;chr&gt;, `101807-101807` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101808-101808` &lt;chr&gt;, `101810-101810` &lt;chr&gt;, `101811-101811` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101812-101812` &lt;chr&gt;, `101813-101813` &lt;chr&gt;, `101814-101814` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101957-101957` &lt;chr&gt;, `101958-101958` &lt;chr&gt;, `100986-100986` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100987-100987` &lt;chr&gt;, `101897-101897` &lt;chr&gt;, `102071-102071` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102106-102106` &lt;chr&gt;, `101040-101040` &lt;chr&gt;, `101167-101167` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101041-101041` &lt;chr&gt;, `101042-101042` &lt;chr&gt;, `101168-101168` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101898-101898` &lt;chr&gt;, `102070-102070` &lt;chr&gt;, `102111-102111` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100988-100988` &lt;chr&gt;, `101896-101896` &lt;chr&gt;, `102110-102110` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102620-102620` &lt;chr&gt;, `102621-102621` &lt;chr&gt;, `102947-102947` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102948-102948` &lt;chr&gt;, `103089-103089` &lt;chr&gt;, `102693-102693` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102694-102694` &lt;chr&gt;, `102949-102949` &lt;chr&gt;, `102622-102622` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102623-102623` &lt;chr&gt;, `102624-102624` &lt;chr&gt;, `102722-102722` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102712-102712` &lt;chr&gt;, `103339-103339` &lt;chr&gt;, `103124-103124` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `103161-103161` &lt;chr&gt;, `103125-103125` &lt;chr&gt;, `103171-103171` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `103338-103338` &lt;chr&gt;, `103372-103372` &lt;chr&gt;, `103193-103193` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100195-100195` &lt;chr&gt;, `100194-100194` &lt;chr&gt;, `101667-101667` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101137-101137` &lt;chr&gt;, `101291-101291` &lt;chr&gt;, `101292-101292` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101666-101666` &lt;chr&gt;, `32222-32222` &lt;chr&gt;, `32049-32049` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `32050-32050` &lt;chr&gt;, `32411-32411` &lt;chr&gt;, `32221-32221` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100147-100147` &lt;chr&gt;, `100149-100149` &lt;chr&gt;, `100243-100243` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100290-100290` &lt;chr&gt;, `100753-100753` &lt;chr&gt;, `100754-100754` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101377-101377` &lt;chr&gt;, `101426-101426` &lt;chr&gt;, `101435-101435` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101523-101523` &lt;chr&gt;, `101877-101877` &lt;chr&gt;, `100043-100043` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100044-100044` &lt;chr&gt;, `100148-100148` &lt;chr&gt;, `100292-100292` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101381-101381` &lt;chr&gt;, ...</span></code></pre></div>
</div>
</div>
<div id="interval-manipulations" class="section level1">
<h1>Interval manipulations</h1>
<p><code>valr</code> implements several methods for manipulating sets of intervals. Some methods operate on a single set of intervals, while others compare two sets of intervals.</p>
<p>Many methods the same name as the corresponding <code>BEDtool</code>, and some commonly used <code>BEDtools</code> are implemented as <code>dplyr</code> pipes (e.g., see the <a href="#groupby"><code>group_by</code></a> section).</p>
<p>All methods accept one or more sets of <code>x</code> and <code>y</code> intervals, which must either be created using the <a href="#reading%20data"><code>read</code> methods</a>, or have <code>chrom</code>, <code>start</code> and <code>end</code> columns.</p>
<div id="single-set-operations" class="section level2 tabset">
<h2>Single set operations</h2>
<p>These methods operate on a single set of intervals:</p>
<ul><li><code><a href="../reference/bed_sort.html">bed_sort()</a></code>: order intervals</li>
<li><code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>: Cluster (but don&rsquo;t merge) overlapping/nearby intervals.</li>
<li><code><a href="../reference/bed_complement.html">bed_complement()</a></code>: extract intervals <em>not</em> represented by an interval file.</li>
<li><code><a href="../reference/bed_merge.html">bed_merge()</a></code>: combine overlapping and nearby intervals into a single interval.</li>
<li><code><a href="../reference/bed_flank.html">bed_flank()</a></code>: Generate new flanking intervals</li>
<li><code><a href="../reference/bed_slop.html">bed_slop()</a></code>: Expand the size of input intervals</li>
<li><code><a href="../reference/bed_shift.html">bed_shift()</a></code>: Shift the coordinates of an input set, bounded by a genome</li>
</ul><div id="sort" class="section level3">
<h3>Sort</h3>
<p><code>bed_sort</code> orders intervals based on a specification. <code>is_sorted</code> asks whether a tbl is already sorted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_sort.html">bed_sort</a></span>(x)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  4129  5129</span>
<span class="co">#&gt; 2   chr1  4243  5243</span>
<span class="co">#&gt; 3   chr1  7492  8492</span>
<span class="co">#&gt; 4   chr1  8420  9420</span>
<span class="co">#&gt; 5   chr1  8841  9841</span>
<span class="co">#&gt; 6   chr1 11605 12605</span>
<span class="co">#&gt; 7   chr1 11827 12827</span>
<span class="co">#&gt; 8   chr1 17898 18898</span>
<span class="co">#&gt; 9   chr1 18440 19440</span>
<span class="co">#&gt; 10  chr1 22667 23667</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="kw"><a href="../reference/is_sorted.html">is_sorted</a></span>(y)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="cluster" class="section level3">
<h3>Cluster</h3>
<p><code>bed_cluster</code> identifies clustered intervals based on a distance specification and assigns them a unique <code>.id</code>.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-cluster_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_cluster.html">bed_cluster</a></span>(x, <span class="dt">max_dist =</span> <span class="dv">1000</span>)
y
<span class="co">#&gt; # A tibble: 1,000,000 &times; 4</span>
<span class="co">#&gt;    chrom     start       end   .id</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr5 141141997 141142997 45755</span>
<span class="co">#&gt; 2   chr6  32678982  32679982 10714</span>
<span class="co">#&gt; 3   chr1 235945953 235946953 75780</span>
<span class="co">#&gt; 4   chr2  68656891  68657891 22197</span>
<span class="co">#&gt; 5  chr13  67243549  67244549 21953</span>
<span class="co">#&gt; 6   chr6   2122548   2123548   701</span>
<span class="co">#&gt; 7   chr3  40724819  40725819 13134</span>
<span class="co">#&gt; 8   chr8  60275139  60276139 19225</span>
<span class="co">#&gt; 9   chrX 126255368 126256368 40903</span>
<span class="co">#&gt; 10 chr17  18673203  18674203  6196</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="complement" class="section level3">
<h3>Complement</h3>
<p><code>bed_complement</code> identifies intervals in a genome that are not covered by an input.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-complement_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human)
<span class="co">#&gt; # A tibble: 723,572 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     1  2500</span>
<span class="co">#&gt; 2   chr1  3500  8704</span>
<span class="co">#&gt; 3   chr1  9704 13645</span>
<span class="co">#&gt; 4   chr1 14645 14683</span>
<span class="co">#&gt; 5   chr1 15683 21090</span>
<span class="co">#&gt; 6   chr1 22090 32726</span>
<span class="co">#&gt; 7   chr1 33726 39820</span>
<span class="co">#&gt; 8   chr1 40820 41156</span>
<span class="co">#&gt; 9   chr1 42309 45146</span>
<span class="co">#&gt; 10  chr1 46146 47733</span>
<span class="co">#&gt; # ... with 723,562 more rows</span></code></pre></div>
</div>
<div id="merge" class="section level3">
<h3>Merge</h3>
<p><code>bed_merge</code> identifies overlapping intervals and reports new merged ones. <code>is_merged</code> asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-merge_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="fl">1e6</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> n)
<span class="kw"><a href="../reference/is_merged.html">is_merged</a></span>(x)
<span class="co">#&gt; [1] FALSE</span>

<span class="co"># add some signal</span>
x &lt;-<span class="st"> </span>x %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">signal =</span> <span class="kw">runif</span>(n))

<span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x, <span class="dt">maxs =</span> <span class="kw">max</span>(signal))
<span class="co">#&gt; # A tibble: 723,634 &times; 4</span>
<span class="co">#&gt;    chrom start   end signal</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1   chr1  2221  3221      .</span>
<span class="co">#&gt; 2   chr1  6984  7984      .</span>
<span class="co">#&gt; 3   chr1  8560  9560      .</span>
<span class="co">#&gt; 4   chr1 12910 13910      .</span>
<span class="co">#&gt; 5   chr1 15036 16289      .</span>
<span class="co">#&gt; 6   chr1 17434 18434      .</span>
<span class="co">#&gt; 7   chr1 20513 21513      .</span>
<span class="co">#&gt; 8   chr1 26409 27409      .</span>
<span class="co">#&gt; 9   chr1 29597 30597      .</span>
<span class="co">#&gt; 10  chr1 32536 33536      .</span>
<span class="co">#&gt; # ... with 723,624 more rows</span></code></pre></div>
</div>
<div id="flank" class="section level3">
<h3>Flank</h3>
<p><code>bed_flank</code> creates new intervals that flank &ndash; but do not contain &ndash; the input intervals.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-flank_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_flank.html">bed_flank</a></span>(x, genome_human, <span class="dt">both  =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    50   150</span>
<span class="co">#&gt; 2  chr1   125   225</span></code></pre></div>
</div>
<div id="slop" class="section level3">
<h3>Slop</h3>
<p><code>bed_slop</code> pads input intervals based on a specification</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-slop_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_slop.html">bed_slop</a></span>(x, genome_human, <span class="dt">both =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1    10   220</span>
<span class="co">#&gt; 2  chr1   125   335</span></code></pre></div>
</div>
<div id="shift" class="section level3">
<h3>Shift</h3>
<p><code>bed_shift</code> adjusts coordinates toward <code>start</code> or <code>end</code> by a defined <code>size</code>. Intervals created out of bounds are removed, or trimmed.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-shift_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/bed_shift.html">bed_shift</a></span>(x, genome_human, <span class="dt">size =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1   125   150</span>
<span class="co">#&gt; 2  chr1   200   225</span></code></pre></div>
</div>
<div id="interval-spacing" class="section level3">
<h3>Interval spacing</h3>
<p>Interval spacing is easily computed using <code>dplyr</code>. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x)

x %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.spacing =</span> start -<span class="st"> </span><span class="kw">lag</span>(end))
<span class="co">#&gt; Source: local data frame [724,798 x 4]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start   end .spacing</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  5037  6037       NA</span>
<span class="co">#&gt; 2   chr1 15181 16181     9144</span>
<span class="co">#&gt; 3   chr1 16741 17741      560</span>
<span class="co">#&gt; 4   chr1 24102 25102     6361</span>
<span class="co">#&gt; 5   chr1 26330 27330     1228</span>
<span class="co">#&gt; 6   chr1 30208 31208     2878</span>
<span class="co">#&gt; 7   chr1 33914 34914     2706</span>
<span class="co">#&gt; 8   chr1 41156 42156     6242</span>
<span class="co">#&gt; 9   chr1 42365 43365      209</span>
<span class="co">#&gt; 10  chr1 44327 45327      962</span>
<span class="co">#&gt; # ... with 724,788 more rows</span></code></pre></div>
</div>
</div>
<div id="multiple-set-operations" class="section level2 tabset">
<h2>Multiple set operations</h2>
<p>These methods compare two sets of intervals:</p>
<ul><li><code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>: find overlapping intervals</li>
<li><code><a href="../reference/bed_map.html">bed_map()</a></code>: apply a function to selected columns for overlapping intervals</li>
<li><code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>: Remove intervals based on overlaps between two files</li>
<li><code><a href="../reference/bed_window.html">bed_window()</a></code>: Find overlapping intervals within a window</li>
<li><code><a href="../reference/bed_closest.html">bed_closest()</a></code>: find the closest intervals independent of overlaps</li>
</ul><div id="intersection" class="section level3">
<h3>Intersection</h3>
<p><code>bed_intersect</code> is implemented using an <a href="https://github.com/ekg/intervaltree/">interval tree</a> in <code>Rcpp</code>. Column names in the result have <code>.x</code> and <code>.y</code> suffixes, and an <code>.overlap</code> column contains the size of the intersection (values of <code>0</code> indicate book-ended, or touching intervals). See the <a href="#benchmarks">benchmarks</a> section for timing. Though <code>bed_intersect</code> is pretty fast already, we intend to further improve upon this by parallization with <a href="https://rcppcore.github.io/RcppParallel/"><code>RcppParallel</code></a>.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-intersect_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># intersect two sets of 1e6 intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 645,713 &times; 6</span>
<span class="co">#&gt;    chrom start.x end.x start.y end.y .overlap</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt;   &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     622  1622    1379  2379      243</span>
<span class="co">#&gt; 2   chr1    1792  2792    1379  2379      587</span>
<span class="co">#&gt; 3   chr1    1792  2792    2697  3697       95</span>
<span class="co">#&gt; 4   chr1    3811  4811    3246  4246      435</span>
<span class="co">#&gt; 5   chr1    3811  4811    3488  4488      677</span>
<span class="co">#&gt; 6   chr1   23669 24669   23698 24698      971</span>
<span class="co">#&gt; 7   chr1   24538 25538   23698 24698      160</span>
<span class="co">#&gt; 8   chr1   28569 29569   29153 30153      416</span>
<span class="co">#&gt; 9   chr1   37344 38344   37940 38940      404</span>
<span class="co">#&gt; 10  chr1   39461 40461   39726 40726      735</span>
<span class="co">#&gt; # ... with 645,703 more rows</span>

<span class="co"># A records with no overlaps (i.e., `-v`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y, <span class="dt">invert =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; Source: local data frame [523,895 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chrY 59369068 59370068</span>
<span class="co">#&gt; 2   chrY 59367448 59368448</span>
<span class="co">#&gt; 3   chrY 59366697 59367697</span>
<span class="co">#&gt; 4   chrY 59337560 59338560</span>
<span class="co">#&gt; 5   chrY 59333969 59334969</span>
<span class="co">#&gt; 6   chrY 59333547 59334547</span>
<span class="co">#&gt; 7   chrY 59322469 59323469</span>
<span class="co">#&gt; 8   chrY 59315086 59316086</span>
<span class="co">#&gt; 9   chrY 59310957 59311957</span>
<span class="co">#&gt; 10  chrY 59310050 59311050</span>
<span class="co">#&gt; # ... with 523,885 more rows</span></code></pre></div>
<p>One can achieve behaviour similar to BEDtools by combining <code>bed_intersect</code> with <code>dplyr</code> tools.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># `x` records with overlaps (i.e., `-wa`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x)
<span class="co">#&gt; # A tibble: 645,713 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   622  1622</span>
<span class="co">#&gt; 2   chr1  1792  2792</span>
<span class="co">#&gt; 3   chr1  1792  2792</span>
<span class="co">#&gt; 4   chr1  3811  4811</span>
<span class="co">#&gt; 5   chr1  3811  4811</span>
<span class="co">#&gt; 6   chr1 23669 24669</span>
<span class="co">#&gt; 7   chr1 24538 25538</span>
<span class="co">#&gt; 8   chr1 28569 29569</span>
<span class="co">#&gt; 9   chr1 37344 38344</span>
<span class="co">#&gt; 10  chr1 39461 40461</span>
<span class="co">#&gt; # ... with 645,703 more rows</span>

<span class="co"># `y` records with overlaps (i.e., `-wb`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.y, <span class="dt">end =</span> end.y)
<span class="co">#&gt; # A tibble: 645,713 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  1379  2379</span>
<span class="co">#&gt; 2   chr1  1379  2379</span>
<span class="co">#&gt; 3   chr1  2697  3697</span>
<span class="co">#&gt; 4   chr1  3246  4246</span>
<span class="co">#&gt; 5   chr1  3488  4488</span>
<span class="co">#&gt; 6   chr1 23698 24698</span>
<span class="co">#&gt; 7   chr1 23698 24698</span>
<span class="co">#&gt; 8   chr1 29153 30153</span>
<span class="co">#&gt; 9   chr1 37940 38940</span>
<span class="co">#&gt; 10  chr1 39726 40726</span>
<span class="co">#&gt; # ... with 645,703 more rows</span>

<span class="co"># Unique records in `x` (i.e., `-u`)</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x) %&gt;%<span class="st"> </span><span class="kw">unique</span>()
<span class="co">#&gt; # A tibble: 476,035 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   622  1622</span>
<span class="co">#&gt; 2   chr1  1792  2792</span>
<span class="co">#&gt; 3   chr1  3811  4811</span>
<span class="co">#&gt; 4   chr1 23669 24669</span>
<span class="co">#&gt; 5   chr1 24538 25538</span>
<span class="co">#&gt; 6   chr1 28569 29569</span>
<span class="co">#&gt; 7   chr1 37344 38344</span>
<span class="co">#&gt; 8   chr1 39461 40461</span>
<span class="co">#&gt; 9   chr1 44203 45203</span>
<span class="co">#&gt; 10  chr1 44239 45239</span>
<span class="co">#&gt; # ... with 476,025 more rows</span>

<span class="co"># Count `y` intervals that intersect each `x` interval</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%<span class="st"> </span><span class="kw">group_by</span>(chrom, start.x, end.x) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">n</span>())
<span class="co">#&gt; Source: local data frame [476,035 x 4]</span>
<span class="co">#&gt; Groups: chrom, start.x [?]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom start.x end.x count</span>
<span class="co">#&gt;    &lt;chr&gt;   &lt;int&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     622  1622     1</span>
<span class="co">#&gt; 2   chr1    1792  2792     2</span>
<span class="co">#&gt; 3   chr1    3811  4811     2</span>
<span class="co">#&gt; 4   chr1   23669 24669     1</span>
<span class="co">#&gt; 5   chr1   24538 25538     1</span>
<span class="co">#&gt; 6   chr1   28569 29569     1</span>
<span class="co">#&gt; 7   chr1   37344 38344     1</span>
<span class="co">#&gt; 8   chr1   39461 40461     2</span>
<span class="co">#&gt; 9   chr1   44203 45203     1</span>
<span class="co">#&gt; 10  chr1   44239 45239     1</span>
<span class="co">#&gt; # ... with 476,025 more rows</span></code></pre></div>
</div>
<div id="map" class="section level3">
<h3>Map</h3>
<p><code>bed_map</code> maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-map_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bedfile &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'genes.hg19.chr22.bed.gz'</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/valr_example.html">valr_example</a></span>(<span class="st">'hela.h3k4.chip.bg.gz'</span>)

x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bed</a></span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_bed.html">read_bedgraph</a></span>(bgfile)

<span class="kw"><a href="../reference/bed_map.html">bed_map</a></span>(x, y, <span class="dt">means =</span> <span class="kw">mean</span>(value.y), <span class="dt">sds =</span> <span class="kw">sd</span>(value.y))
<span class="co">#&gt; # A tibble: 591 &times; 5</span>
<span class="co">#&gt;    chrom  start.x    end.x    means       sds</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;    &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr22 16150259 16193004 7.914286 7.5707309</span>
<span class="co">#&gt; 2  chr22 16162065 16172265 1.000000        NA</span>
<span class="co">#&gt; 3  chr22 16256331 16287937 1.000000 0.0000000</span>
<span class="co">#&gt; 4  chr22 17071647 17073700 1.000000 0.0000000</span>
<span class="co">#&gt; 5  chr22 17082800 17129720 1.117647 0.3270350</span>
<span class="co">#&gt; 6  chr22 17134598 17156430 1.294118 0.5878675</span>
<span class="co">#&gt; 7  chr22 17227758 17229328 1.000000        NA</span>
<span class="co">#&gt; 8  chr22 17264305 17302584 1.250000 0.5084039</span>
<span class="co">#&gt; 9  chr22 17308363 17310225 1.333333 0.5773503</span>
<span class="co">#&gt; 10 chr22 17385314 17385395 1.000000        NA</span>
<span class="co">#&gt; # ... with 581 more rows</span></code></pre></div>
</div>
<div id="subtract" class="section level3">
<h3>Subtract</h3>
<p><code>bed_substract()</code> removes <code>x</code> intervals that intersect with <code>y</code>.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-subtract_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y)
<span class="co">#&gt; # A tibble: 957,458 &times; 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1  8462  8774</span>
<span class="co">#&gt; 2   chr1 19624 20568</span>
<span class="co">#&gt; 3   chr1 32708 33389</span>
<span class="co">#&gt; 4   chr1 35619 36340</span>
<span class="co">#&gt; 5   chr1 35619 36340</span>
<span class="co">#&gt; 6   chr1 36010 36340</span>
<span class="co">#&gt; 7   chr1 39432 40432</span>
<span class="co">#&gt; 8   chr1 40860 41860</span>
<span class="co">#&gt; 9   chr1 50542 51542</span>
<span class="co">#&gt; 10  chr1 52964 53600</span>
<span class="co">#&gt; # ... with 957,448 more rows</span></code></pre></div>
</div>
<div id="window" class="section level3">
<h3>Window</h3>
<p><code><a href="../reference/bed_window.html">bed_window()</a></code> identifies <code>y</code> intervals that intersect an expanded window of <code>x</code> intervals.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-window_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="co"># a few intersections</span>
<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span>

<span class="co"># can be expanded by casting a wider net</span>
<span class="kw"><a href="../reference/bed_window.html">bed_window</a></span>(x, y, genome, <span class="dt">both =</span> <span class="fl">1e6</span>)
<span class="co">#&gt; # A tibble: 0 &times; 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span></code></pre></div>
</div>
<div id="closest" class="section level3">
<h3>Closest</h3>
<p><code><a href="../reference/bed_closest.html">bed_closest()</a></code> identifies <code>y</code> intervals that are closest to <code>x</code>.</p>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-closest_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y)
<span class="co">#&gt; # A tibble: 98 &times; 7</span>
<span class="co">#&gt;    chrom   start.x     end.x   start.y     end.y .overlap .distance</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;    &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1   5269390   5270390   3858222   3859222        0  -1410168</span>
<span class="co">#&gt; 2   chr1  19390794  19391794   3858222   3859222        0 -15531572</span>
<span class="co">#&gt; 3   chr1  20359370  20360370   3858222   3859222        0 -16500148</span>
<span class="co">#&gt; 4   chr1  24283903  24284903   3858222   3859222        0 -20424681</span>
<span class="co">#&gt; 5   chr1  24832816  24833816   3858222   3859222        0 -20973594</span>
<span class="co">#&gt; 6   chr1  25357822  25358822   3858222   3859222        0 -21498600</span>
<span class="co">#&gt; 7   chr1  59741807  59742807  54783898  54784898        0  -4956909</span>
<span class="co">#&gt; 8   chr1 110045159 110046159 105696769 105697769        0  -4347390</span>
<span class="co">#&gt; 9   chr1 190411099 190412099 169800813 169801813        0 -20609286</span>
<span class="co">#&gt; 10  chr1 209521858 209522858 226372532 226373532        0  16849674</span>
<span class="co">#&gt; # ... with 88 more rows</span></code></pre></div>
</div>
</div>
<div id="randomzing-intervals" class="section level2 tabset">
<h2>Randomzing intervals</h2>
<p><code>valr</code> provides methods for creating new random intervals or permutations of existing intervals:</p>
<ul><li><code>bed_random</code> generates random intervals from an input <code>genome</code>.</li>
<li><code>bed_shuffle</code> shuffles coordinates given a set of input intervals.</li>
<li>Random sampling of input intervals is done with <code>dplyr</code>.</li>
</ul><div id="random" class="section level3">
<h3>Random</h3>
<p><code>bed_random</code> generates random intervals from an input genome. The numbers of intervals from each <code>chrom</code> are proporitional to each chrom <code>size</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human, <span class="dt">n =</span> <span class="fl">1e6</span>, <span class="dt">length =</span> <span class="fl">1e3</span>)
x
<span class="co">#&gt; # A tibble: 1,000,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr12  32683669  32684669</span>
<span class="co">#&gt; 2  chr17  18188351  18189351</span>
<span class="co">#&gt; 3   chr2  86314205  86315205</span>
<span class="co">#&gt; 4   chr1 108132702 108133702</span>
<span class="co">#&gt; 5   chr2 193703758 193704758</span>
<span class="co">#&gt; 6   chrY  40954282  40955282</span>
<span class="co">#&gt; 7   chr5  58487935  58488935</span>
<span class="co">#&gt; 8   chr9  97148691  97149691</span>
<span class="co">#&gt; 9   chr9 120475884 120476884</span>
<span class="co">#&gt; 10  chr4  85575250  85576250</span>
<span class="co">#&gt; # ... with 999,990 more rows</span>

<span class="co"># numbers of sampled intervals are proportional to chrom size</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%<span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom     n</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 80709</span>
<span class="co">#&gt; 2   chr2 78264</span>
<span class="co">#&gt; 3   chr3 64219</span>
<span class="co">#&gt; 4   chr4 61727</span>
<span class="co">#&gt; 5   chr5 58315</span>
<span class="co">#&gt; 6   chr6 55179</span>
<span class="co">#&gt; 7   chr7 51559</span>
<span class="co">#&gt; 8   chrX 50110</span>
<span class="co">#&gt; 9   chr8 47594</span>
<span class="co">#&gt; 10  chr9 45719</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="sample" class="section level3">
<h3>Sample</h3>
<p>Sampling can be done using <code>dplyr</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="co"># sample by number</span>
<span class="kw">sample_n</span>(x, <span class="fl">1e3</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 1,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr9 108077560 108078560</span>
<span class="co">#&gt; 2   chr6   3256985   3257985</span>
<span class="co">#&gt; 3  chr15  91313645  91314645</span>
<span class="co">#&gt; 4   chr7 146648350 146649350</span>
<span class="co">#&gt; 5  chr13  33625192  33626192</span>
<span class="co">#&gt; 6  chr21  17240658  17241658</span>
<span class="co">#&gt; 7  chr20  45627695  45628695</span>
<span class="co">#&gt; 8   chrY   3944743   3945743</span>
<span class="co">#&gt; 9   chr1  84753267  84754267</span>
<span class="co">#&gt; 10  chr1   4516059   4517059</span>
<span class="co">#&gt; # ... with 990 more rows</span>

<span class="co"># or fraction</span>
<span class="kw">sample_frac</span>(x, <span class="fl">0.1</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 100,000 &times; 3</span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr4  98596336  98597336</span>
<span class="co">#&gt; 2   chr4  67664965  67665965</span>
<span class="co">#&gt; 3   chr5  75559151  75560151</span>
<span class="co">#&gt; 4   chr5 145103149 145104149</span>
<span class="co">#&gt; 5  chr15  77796461  77797461</span>
<span class="co">#&gt; 6   chr1  39612264  39613264</span>
<span class="co">#&gt; 7   chr1 192816029 192817029</span>
<span class="co">#&gt; 8  chr11 130054181 130055181</span>
<span class="co">#&gt; 9  chr12  75306482  75307482</span>
<span class="co">#&gt; 10  chr3 110255347 110256347</span>
<span class="co">#&gt; # ... with 99,990 more rows</span>

<span class="co"># or sample intervals within groups</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">sample_n</span>(<span class="dv">1</span>)
<span class="co">#&gt; Source: local data frame [25 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    chrom     start       end</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 131903620 131904620</span>
<span class="co">#&gt; 2  chr10 129196019 129197019</span>
<span class="co">#&gt; 3  chr11 117645449 117646449</span>
<span class="co">#&gt; 4  chr12  13594991  13595991</span>
<span class="co">#&gt; 5  chr13   3513623   3514623</span>
<span class="co">#&gt; 6  chr14  35163126  35164126</span>
<span class="co">#&gt; 7  chr15  62784264  62785264</span>
<span class="co">#&gt; 8  chr16  25412857  25413857</span>
<span class="co">#&gt; 9  chr17  69288892  69289892</span>
<span class="co">#&gt; 10 chr18  45098820  45099820</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
<div id="shuffle" class="section level3">
<h3>Shuffle</h3>
<p><code>bed_shuffle</code> shuffles input intervals. Interval sizes are equal in the input and output.</p>
<pre><code>#&gt; Warning in bind_rows_(x, .id): binding factor and character vector,
#&gt; coercing into character vector</code></pre>
<p><img src="/Users/jayhesselberth/devel/valr/vignettes/README-shuffle_glyph-1.png" width="384" style="display: block; margin: auto;"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human)
y
<span class="co">#&gt; # A tibble: 2 &times; 3</span>
<span class="co">#&gt;    chrom    start      end</span>
<span class="co">#&gt;   &lt;fctr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1   chr2 72642139 72642159</span>
<span class="co">#&gt; 2   chr7 53941540 53941550</span>

<span class="kw">any</span>(x$start ==<span class="st"> </span>y$start)
<span class="co">#&gt; [1] FALSE</span>

<span class="kw">all</span>(x$end -<span class="st"> </span>x$start ==<span class="st"> </span>y$end -<span class="st"> </span>y$start)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
</div>
<div id="other-topics" class="section level2">
<h2>Other topics</h2>
<div id="grouping-related-intervals" class="section level3">
<h3>Grouping related intervals</h3>
<p>The <code>group_by</code> operation from <code>dplyr</code> serves a similar purpose to BEDtools <code>groupby</code>, and can be combined with <code>dplyr::summarize</code> to calculate summary statistics from groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y) %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">overlap.total =</span> <span class="kw">sum</span>(.overlap))
<span class="co">#&gt; # A tibble: 25 &times; 2</span>
<span class="co">#&gt;    chrom overlap.total</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1      26097486</span>
<span class="co">#&gt; 2  chr10      14110924</span>
<span class="co">#&gt; 3  chr11      14155908</span>
<span class="co">#&gt; 4  chr12      13993025</span>
<span class="co">#&gt; 5  chr13      12069158</span>
<span class="co">#&gt; 6  chr14      11245697</span>
<span class="co">#&gt; 7  chr15      10633304</span>
<span class="co">#&gt; 8  chr16       9156965</span>
<span class="co">#&gt; 9  chr17       8366410</span>
<span class="co">#&gt; 10 chr18       8122340</span>
<span class="co">#&gt; # ... with 15 more rows</span></code></pre></div>
</div>
</div>
</div>
<div id="interval-statistics" class="section level1 tabset">
<h1>Interval Statistics</h1>
<p><code>valr</code> provides several methods to assess statistical properties of interval sets including:</p>
<ul><li><code><a href="../reference/bed_fisher.html">bed_fisher()</a></code>: measure overlap significance of two sets of intervals.</li>
<li><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>: quantify absolute distance between query intervals and closest reference intervals</li>
<li><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code>: quantify relative distances between query intervals and closest reference intervals</li>
<li><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>: quantify extent of overlap between two sets of intervals</li>
</ul><p>Several of these methods were described in the <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">Genometricorr</a> software package.</p>
<div id="fishers-test" class="section level2">
<h2>Fisher&rsquo;s test</h2>
<p>The Fisher&rsquo;s test assesses whether two sets of intervals are drawn from the same background genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_fisher.html">bed_fisher</a></span>(x, y, genome_human)
<span class="co">#&gt;   estimate p.value conf.low conf.high                             method</span>
<span class="co">#&gt; 1 2.374292       0 2.359993  2.388549 Fisher's Exact Test for Count Data</span>
<span class="co">#&gt;   alternative</span>
<span class="co">#&gt; 1   two.sided</span></code></pre></div>
</div>
<div id="absolute-distance" class="section level2">
<h2>Absolute distance</h2>
<p><code><a href="../reference/bed_absdist.html">bed_absdist()</a></code> computes the absolute distance between the midpoint of query intervals and the closest midpoints of a set of reference intervals. Absolute distances are scaled by the inter-reference gap for the chromosome as follows. For <code>Q</code> total query points and <code>R</code> reference points on a chromosome, scale the distance for each query point <code>i</code> to the closest reference point by the inter-reference gap for each chromosome.</p>
<p><span class="math display">\[
d_i(x,y) = min_{\substack{k}}(|q_i - r_k|)\frac{R}{Length\ of\ chromsome}
\]</span></p>
<p>By default both absolute and scaled distances are reported as <code>absdist</code> and <code>scaled_absdist</code> respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_absdist.html">bed_absdist</a></span>(x, y, genome_human)
<span class="co">#&gt; # A tibble: 1,000,000 &times; 5</span>
<span class="co">#&gt;    chrom start   end absdist scaled_absdist</span>
<span class="co">#&gt;    &lt;chr&gt; &lt;int&gt; &lt;int&gt;   &lt;dbl&gt;          &lt;dbl&gt;</span>
<span class="co">#&gt; 1   chr1  5286  6286     532     0.17206875</span>
<span class="co">#&gt; 2   chr1  7782  8782     671     0.21702657</span>
<span class="co">#&gt; 3   chr1 12390 13390     317     0.10252969</span>
<span class="co">#&gt; 4   chr1 17373 18373    3692     1.19413128</span>
<span class="co">#&gt; 5   chr1 27669 28669    1332     0.43081876</span>
<span class="co">#&gt; 6   chr1 28814 29814     195     0.06307031</span>
<span class="co">#&gt; 7   chr1 28971 29971      38     0.01229063</span>
<span class="co">#&gt; 8   chr1 34864 35864     581     0.18791719</span>
<span class="co">#&gt; 9   chr1 35062 36062     383     0.12387657</span>
<span class="co">#&gt; 10  chr1 42832 43832    1174     0.37971564</span>
<span class="co">#&gt; # ... with 999,990 more rows</span></code></pre></div>
</div>
<div id="relative-distance" class="section level2">
<h2>Relative distance</h2>
<p><code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between <code>[0,0.5]</code>. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">17</a>). By default a summary with the frequency of each relative distance is reported.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_reldist.html">bed_reldist</a></span>(x, y)
<span class="co">#&gt; # A tibble: 51 &times; 4</span>
<span class="co">#&gt;    reldist counts  total       freq</span>
<span class="co">#&gt;      &lt;dbl&gt;  &lt;int&gt;  &lt;int&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt; 1     0.00  19866 999955 0.01986689</span>
<span class="co">#&gt; 2     0.01  20242 999955 0.02024291</span>
<span class="co">#&gt; 3     0.02  19872 999955 0.01987289</span>
<span class="co">#&gt; 4     0.03  20177 999955 0.02017791</span>
<span class="co">#&gt; 5     0.04  19879 999955 0.01987989</span>
<span class="co">#&gt; 6     0.05  19915 999955 0.01991590</span>
<span class="co">#&gt; 7     0.06  19909 999955 0.01990990</span>
<span class="co">#&gt; 8     0.07  20099 999955 0.02009990</span>
<span class="co">#&gt; 9     0.08  20074 999955 0.02007490</span>
<span class="co">#&gt; 10    0.09  19999 999955 0.01999990</span>
<span class="co">#&gt; # ... with 41 more rows</span></code></pre></div>
</div>
<div id="jaccard-similarity" class="section level2">
<h2>Jaccard similarity</h2>
<p><code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code> quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of <code>[0,1]</code> and is measured as:</p>
<p><span class="math display">\[
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
\]</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw"><a href="../reference/bed_jaccard.html">bed_jaccard</a></span>(x, y)
<span class="co">#&gt; # A tibble: 1 &times; 4</span>
<span class="co">#&gt;       len_i      len_u  jaccard      n</span>
<span class="co">#&gt;       &lt;int&gt;      &lt;int&gt;    &lt;dbl&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1 323448650 2000000000 0.192925 647197</span></code></pre></div>
</div>
</div>
<div id="benchmarks" class="section level1">
<h1>Benchmarks</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># two sets of 1e6 random 1 kb intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)
y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human)

<span class="kw">library</span>(microbenchmark)
<span class="kw">microbenchmark</span>(
  <span class="kw"><a href="../reference/bed_random.html">bed_random</a></span>(genome_human),
  <span class="kw"><a href="../reference/bed_closest.html">bed_closest</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_intersect.html">bed_intersect</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_merge.html">bed_merge</a></span>(x),
  <span class="kw"><a href="../reference/bed_subtract.html">bed_subtract</a></span>(x, y),
  <span class="kw"><a href="../reference/bed_complement.html">bed_complement</a></span>(x, genome_human),
  <span class="kw"><a href="../reference/bed_shuffle.html">bed_shuffle</a></span>(x, genome_human),
  <span class="dt">times =</span> <span class="dv">1</span>,
  <span class="dt">unit =</span> <span class="st">'s'</span>
)
<span class="co">#&gt; Unit: seconds</span>
<span class="co">#&gt;                             expr       min        lq      mean    median</span>
<span class="co">#&gt;         bed_random(genome_human) 0.1323735 0.1323735 0.1323735 0.1323735</span>
<span class="co">#&gt;                bed_closest(x, y) 5.4086166 5.4086166 5.4086166 5.4086166</span>
<span class="co">#&gt;              bed_intersect(x, y) 1.7779920 1.7779920 1.7779920 1.7779920</span>
<span class="co">#&gt;                     bed_merge(x) 2.1477086 2.1477086 2.1477086 2.1477086</span>
<span class="co">#&gt;               bed_subtract(x, y) 3.1903619 3.1903619 3.1903619 3.1903619</span>
<span class="co">#&gt;  bed_complement(x, genome_human) 2.6987088 2.6987088 2.6987088 2.6987088</span>
<span class="co">#&gt;     bed_shuffle(x, genome_human) 1.0127592 1.0127592 1.0127592 1.0127592</span>
<span class="co">#&gt;         uq       max neval</span>
<span class="co">#&gt;  0.1323735 0.1323735     1</span>
<span class="co">#&gt;  5.4086166 5.4086166     1</span>
<span class="co">#&gt;  1.7779920 1.7779920     1</span>
<span class="co">#&gt;  2.1477086 2.1477086     1</span>
<span class="co">#&gt;  3.1903619 3.1903619     1</span>
<span class="co">#&gt;  2.6987088 2.6987088     1</span>
<span class="co">#&gt;  1.0127592 1.0127592     1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked"><li><a href="#overview">Overview</a><ul class="nav nav-pills nav-stacked"><li><a href="#installation">Installation</a></li>
      <li><a href="#comparison-to-other-tools">Comparison to other tools</a></li>
      <li><a href="#non-standard-evaluation">Non-standard evaluation</a></li>
      <li><a href="#getting-started">Getting started</a></li>
      </ul></li>
      <li><a href="#reading-data">Reading data</a><ul class="nav nav-pills nav-stacked"><li><a href="#bed-files">BED files</a></li>
      <li><a href="#genome-files">Genome files</a></li>
      <li><a href="#bedgraph-files">bedGraph files</a></li>
      <li><a href="#vcf-files">VCF files</a></li>
      </ul></li>
      <li><a href="#interval-manipulations">Interval manipulations</a><ul class="nav nav-pills nav-stacked"><li><a href="#single-set-operations">Single set operations</a></li>
      <li><a href="#multiple-set-operations">Multiple set operations</a></li>
      <li><a href="#randomzing-intervals">Randomzing intervals</a></li>
      <li><a href="#other-topics">Other topics</a></li>
      </ul></li>
      <li><a href="#interval-statistics">Interval Statistics</a><ul class="nav nav-pills nav-stacked"><li><a href="#fishers-test">Fisher&rsquo;s test</a></li>
      <li><a href="#absolute-distance">Absolute distance</a></li>
      <li><a href="#relative-distance">Relative distance</a></li>
      <li><a href="#jaccard-similarity">Jaccard similarity</a></li>
      </ul></li>
      <li><a href="#benchmarks">Benchmarks</a></li>
      </ul></div>
      </div>

</div>


      <footer><p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer></div>

  </body></html>
