<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>read genome files (i.e., UCSC &quot;chromSize&quot; files) — read_genome • valr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>

<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container">
      <header>
      
<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">valr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/valr.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jayhesselberth/valr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header>

      <div class="page-header">
  <h1>read genome files (i.e., UCSC &quot;chromSize&quot; files)</h1>
</div>

<div class="row">
  <div class="col-md-9">
    
    <p>Genome files contain chromosome name and size information. These sizes are 
used by downstream functions to identify computed intervals that have ends
outside the defined size.</p>

    <p>Select intervals bounded by a genome.</p>
    

    <pre><span class='fu'>read_genome</span>(<span class='no'>filename</span>)

<span class='fu'>bound_intervals</span>(<span class='no'>x</span>, <span class='no'>genome</span>, <span class='kw'>trim</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <div class="section">
    <h2>Arguments</h2>
    <dl class="dl-horizontal">
      <dt>filename</dt>
      <dd>file containing chrom/contig names and sizes,
one-pair-per-line, tab-delimited</dd>
      <dt>x</dt>
      <dd>a tbl of intervals</dd>
      <dt>genome</dt>
      <dd>a tbl of chrom sizes</dd>
      <dt>trim</dt>
      <dd>adjust coordinates for out-of-bounds intervals</dd>
    </dl>
    </div>
    
    <div class="section">
      <h2>Value</h2>

      <p><code>data_frame</code> with colnames <code>chrom</code> and <code>size</code>, sorted
  by <code>size</code></p>
      <p><code>data_frame</code></p>
    </div>

    <div class="section">
      <h2>See also</h2>

      <p>Other read-funcs: <code><a href='read_bed.html'>read_bed</a></code>,
  <code><a href='read_vcf.html'>read_vcf</a></code></p>
    </div>
    
    <div class="section">
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>genome</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='valr_example.html'>valr_example</a></span>(<span class='st'>'hg19.chrom.sizes.gz'</span>)
<span class='fu'>read_genome</span>(<span class='no'>genome</span>)</div><div class='output co'>#&gt; # A tibble: 25 × 2
#&gt;    chrom      size
#&gt;    &lt;chr&gt;     &lt;int&gt;
#&gt; 1   chr1 249250621
#&gt; 2   chr2 243199373
#&gt; 3   chr3 198022430
#&gt; 4   chr4 191154276
#&gt; 5   chr5 180915260
#&gt; 6   chr6 171115067
#&gt; 7   chr7 159138663
#&gt; 8   chrX 155270560
#&gt; 9   chr8 146364022
#&gt; 10  chr9 141213431
#&gt; # ... with 15 more rows
#&gt; </div><div class='input'>
<span class='co'>## Not run: ------------------------------------</span>
<span class='co'># # can also read from a URL</span>
<span class='co'># read_genome('https://genome.ucsc.edu/goldenpath/help/hg19.chrom.sizes')</span>
<span class='co'>## ---------------------------------------------</span>

<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='kw'>tibble</span><span class='kw'>::</span><span class='fu'>tribble</span>(
 ~<span class='no'>chrom</span>, ~<span class='no'>start</span>, ~<span class='no'>end</span>,
 <span class='st'>"chr1"</span>, -<span class='fl'>100</span>,   <span class='fl'>500</span>,
 <span class='st'>"chr1"</span>, <span class='fl'>100</span>,    <span class='fl'>1e9</span>,
 <span class='st'>"chr1"</span>, <span class='fl'>500</span>,    <span class='fl'>1000</span>
)

<span class='no'>genome</span> <span class='kw'>&lt;-</span> <span class='fu'>read_genome</span>(<span class='fu'><a href='valr_example.html'>valr_example</a></span>(<span class='st'>'hg19.chrom.sizes.gz'</span>))

<span class='fu'>bound_intervals</span>(<span class='no'>x</span>, <span class='no'>genome</span>)</div><div class='output co'>#&gt; # A tibble: 1 × 3
#&gt;   chrom start   end
#&gt;   &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
#&gt; 1  chr1   500  1000
#&gt; </div><div class='input'><span class='fu'>bound_intervals</span>(<span class='no'>x</span>, <span class='no'>genome</span>, <span class='kw'>trim</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; # A tibble: 3 × 3
#&gt;   chrom start       end
#&gt;   &lt;chr&gt; &lt;dbl&gt;     &lt;dbl&gt;
#&gt; 1  chr1     1       500
#&gt; 2  chr1   100 249250621
#&gt; 3  chr1   500      1000
#&gt; </div><div class='input'>
</div></pre>
    </div>
  </div>
  <div class="col-md-3">
  </div>
</div>

      <footer>
      <p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer>
   </div>

  </body>
</html>
