<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="valr">
<title>valr overview • valr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><link href="../deps/Source_Sans_Pro-0.4.2/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.2/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="valr overview">
<meta property="og:description" content="valr">
<meta property="og:image" content="https://rnabioco.github.io/valr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script defer data-domain="rnabioco.github.io" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">valr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.6.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/valr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/interval-stats.html">Interval statistics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/rnabioco/valr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>valr overview</h1>
            
            <h4 data-toc-skip class="date">2022-10-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rnabioco/valr/blob/HEAD/vignettes/valr.Rmd" class="external-link"><code>vignettes/valr.Rmd</code></a></small>
      <div class="d-none name"><code>valr.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="why-valr">Why <code>valr</code>?<a class="anchor" aria-label="anchor" href="#why-valr"></a>
</h2>
<p><strong>Why another tool set for interval manipulations?</strong> There are several <a href="#related_work">other software packages</a> available for genome interval analysis. However, based on our experiences teaching genome analysis, we were motivated to develop a toolset that:</p>
<ul>
<li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the “modern R” syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visualizations with <a href="https://shiny.rstudio.com/" class="external-link"><code>shiny</code></a>.</li>
</ul>
<p><code>valr</code> can currently be used for analysis of pre-processed data in BED and related formats. We plan to support BAM and VCF files soon via tabix indexes.</p>
<div class="section level3">
<h3 id="familiar-tools-natively-in-r">Familiar tools, natively in R<a class="anchor" aria-label="anchor" href="#familiar-tools-natively-in-r"></a>
</h3>
<p>The functions in <code>valr</code> have similar names to their <code>BEDtools</code> counterparts, and so will be familiar to users coming from the <code>BEDtools</code> suite. Similar to <a href="https://daler.github.io/pybedtools/#why-pybedtools" class="external-link"><code>pybedtools</code></a>, <code>valr</code> has a terse syntax:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rnabioco/valr/" class="external-link">valr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'hg19.snps147.chr22.bed.gz'</span><span class="op">)</span>, n_fields <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'genes.hg19.chr22.bed.gz'</span><span class="op">)</span>, n_fields <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find snps in intergenic regions</span></span>
<span><span class="va">intergenic</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bed_subtract.html">bed_subtract</a></span><span class="op">(</span><span class="va">snps</span>, <span class="va">genes</span><span class="op">)</span></span>
<span><span class="co"># distance from intergenic snps to nearest gene</span></span>
<span><span class="va">nearby</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bed_closest.html">bed_closest</a></span><span class="op">(</span><span class="va">intergenic</span>, <span class="va">genes</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nearby</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">'name'</span><span class="op">)</span>, <span class="va">.overlap</span>, <span class="va">.dist</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">.dist</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 284 × 4</span></span></span>
<span><span class="co">#&gt;    name.x      name.y            .overlap .dist</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> rs2261631   P704P                    0  -<span style="color: #BB0000;">268</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> rs570770556 POTEH                    0  -<span style="color: #BB0000;">913</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> rs538163832 POTEH                    0  -<span style="color: #BB0000;">953</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> rs9606135   TPTEP1                   0  -<span style="color: #BB0000;">422</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> rs11912392  ANKRD62P1-PARP4P3        0   105</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> rs8136454   BC038197                 0   356</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> rs5992556   XKR3                     0  -<span style="color: #BB0000;">456</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> rs114101676 GAB4                     0   474</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> rs62236167  CECR7                    0   262</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> rs5747023   CECR1                    0  -<span style="color: #BB0000;">387</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 274 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p><code>valr</code> assigns common column names to facilitate comparisons between tbls. All tbls will have <code>chrom</code>, <code>start</code>, and <code>end</code> columns, and some tbls from multi-column formats will have additional pre-determined column names. See the <code><a href="../reference/read_bed.html">read_bed()</a></code> documentation for details.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">"3fields.bed.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="va">bed_file</span><span class="op">)</span> <span class="co"># accepts filepaths or URLs</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 3</span></span></span>
<span><span class="co">#&gt;    chrom  start    end</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> chr1   <span style="text-decoration: underline;">11</span>873  <span style="text-decoration: underline;">14</span>409</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> chr1   <span style="text-decoration: underline;">14</span>361  <span style="text-decoration: underline;">19</span>759</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> chr1   <span style="text-decoration: underline;">14</span>406  <span style="text-decoration: underline;">29</span>370</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> chr1   <span style="text-decoration: underline;">34</span>610  <span style="text-decoration: underline;">36</span>081</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> chr1   <span style="text-decoration: underline;">69</span>090  <span style="text-decoration: underline;">70</span>008</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> chr1  <span style="text-decoration: underline;">134</span>772 <span style="text-decoration: underline;">140</span>566</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> chr1  <span style="text-decoration: underline;">321</span>083 <span style="text-decoration: underline;">321</span>115</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> chr1  <span style="text-decoration: underline;">321</span>145 <span style="text-decoration: underline;">321</span>207</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> chr1  <span style="text-decoration: underline;">322</span>036 <span style="text-decoration: underline;">326</span>938</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> chr1  <span style="text-decoration: underline;">327</span>545 <span style="text-decoration: underline;">328</span>439</span></span></code></pre></div>
<p><code>valr</code> can also operate on BED-like data.frames already constructed in R, provided that columns named <code>chrom</code>, <code>start</code> and <code>end</code> are present. New tbls can also be constructed as either <code>tibbles</code> or base R <code>data.frames</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>,  <span class="op">~</span><span class="va">end</span>, </span>
<span>  <span class="st">"chr1"</span>, <span class="fl">1657492</span>, <span class="fl">2657492</span>, </span>
<span>  <span class="st">"chr2"</span>, <span class="fl">2501324</span>, <span class="fl">3094650</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">bed</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 3</span></span></span>
<span><span class="co">#&gt;   chrom   start     end</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1  1<span style="text-decoration: underline;">657</span>492 2<span style="text-decoration: underline;">657</span>492</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr2  2<span style="text-decoration: underline;">501</span>324 3<span style="text-decoration: underline;">094</span>650</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="interval-coordinates">Interval coordinates<a class="anchor" aria-label="anchor" href="#interval-coordinates"></a>
</h3>
<p><code>valr</code> adheres to the BED <a href="https://genome.ucsc.edu/FAQ/FAQformat#format1" class="external-link">format</a> which specifies that the start position for an interval is zero based and the end position is one-based. The first position in a chromosome is 0. The end position for a chromosome is one position passed the last base, and is not included in the interval. For example:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a chromosome 100 basepairs in length</span></span>
<span><span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>, </span>
<span>  <span class="st">"chr1"</span>, <span class="fl">0</span>,      <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">chrom</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   chrom start   end</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1      0   100</span></span>
<span></span>
<span><span class="co"># single basepair intervals</span></span>
<span><span class="va">bases</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>, </span>
<span>  <span class="st">"chr1"</span>, <span class="fl">0</span>,      <span class="fl">1</span>, <span class="co"># first base of chromosome</span></span>
<span>  <span class="st">"chr1"</span>, <span class="fl">1</span>,      <span class="fl">2</span>,  <span class="co"># second base of chromosome</span></span>
<span>  <span class="st">"chr1"</span>, <span class="fl">99</span>,     <span class="fl">100</span> <span class="co"># last base of chromosome</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">bases</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 3</span></span></span>
<span><span class="co">#&gt;   chrom start   end</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1      0     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1      1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr1     99   100</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="remote-databases">Remote databases<a class="anchor" aria-label="anchor" href="#remote-databases"></a>
</h3>
<p>Remote databases can be accessed with <code><a href="../reference/db.html">db_ucsc()</a></code> (to access the UCSC Browser) and <code><a href="../reference/db.html">db_ensembl()</a></code> (to access Ensembl databases).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># access the `refGene` tbl on the `hg38` assembly.</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://rmariadb.r-dbi.org" class="external-link">RMariaDB</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">ucsc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/db.html">db_ucsc</a></span><span class="op">(</span><span class="st">'hg38'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/tbl.html" class="external-link">tbl</a></span><span class="op">(</span><span class="va">ucsc</span>, <span class="st">'refGene'</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visual-documentation">Visual documentation<a class="anchor" aria-label="anchor" href="#visual-documentation"></a>
</h3>
<p>The <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code> tool illustrates the results of operations in <code>valr</code>, similar to those found in the <code>BEDtools</code> documentation. This glyph shows the result of intersecting <code>x</code> and <code>y</code> intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">25</span>,     <span class="fl">50</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">100</span>,    <span class="fl">125</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">30</span>,     <span class="fl">75</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bed_glyph.html">bed_glyph</a></span><span class="op">(</span><span class="fu"><a href="../reference/bed_intersect.html">bed_intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="valr_files/figure-html/intersect_glyph-1.png" width="700" style="display: block; margin: auto;"></p>
<p>And this glyph illustrates <code><a href="../reference/bed_merge.html">bed_merge()</a></code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>,</span>
<span>  <span class="st">'chr1'</span>,      <span class="fl">1</span>,      <span class="fl">50</span>,</span>
<span>  <span class="st">'chr1'</span>,      <span class="fl">10</span>,     <span class="fl">75</span>,</span>
<span>  <span class="st">'chr1'</span>,      <span class="fl">100</span>,    <span class="fl">120</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bed_glyph.html">bed_glyph</a></span><span class="op">(</span><span class="fu"><a href="../reference/bed_merge.html">bed_merge</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="valr_files/figure-html/merge_glyph-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="grouping-data">Grouping data<a class="anchor" aria-label="anchor" href="#grouping-data"></a>
</h3>
<p>The <code>group_by</code> function in dplyr can be used to perform functions on subsets of single and multiple <code>data_frame</code>s. Functions in <code>valr</code> leverage grouping to enable a variety of comparisons. For example, intervals can be grouped by <code>strand</code> to perform comparisons among intervals on the same strand.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>, <span class="op">~</span><span class="va">strand</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">1</span>,      <span class="fl">100</span>,  <span class="st">'+'</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">50</span>,     <span class="fl">150</span>,  <span class="st">'+'</span>,</span>
<span>  <span class="st">'chr2'</span>, <span class="fl">100</span>,    <span class="fl">200</span>,  <span class="st">'-'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tribble.html" class="external-link">tribble</a></span><span class="op">(</span></span>
<span>  <span class="op">~</span><span class="va">chrom</span>, <span class="op">~</span><span class="va">start</span>, <span class="op">~</span><span class="va">end</span>, <span class="op">~</span><span class="va">strand</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">50</span>,     <span class="fl">125</span>,  <span class="st">'+'</span>,</span>
<span>  <span class="st">'chr1'</span>, <span class="fl">50</span>,     <span class="fl">150</span>,  <span class="st">'-'</span>,</span>
<span>  <span class="st">'chr2'</span>, <span class="fl">50</span>,     <span class="fl">150</span>,  <span class="st">'+'</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># intersect tbls by strand</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">strand</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">strand</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bed_intersect.html">bed_intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 8</span></span></span>
<span><span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1        1   100 +             50   125 +              50</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1       50   150 +             50   125 +              75</span></span></code></pre></div>
<p>Comparisons between intervals on opposite strands are done using the <code><a href="../reference/flip_strands.html">flip_strands()</a></code> function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">strand</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/flip_strands.html">flip_strands</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">strand</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/bed_intersect.html">bed_intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 8</span></span></span>
<span><span class="co">#&gt;   chrom start.x end.x strand.x start.y end.y strand.y .overlap</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1        1   100 +             50   150 +              50</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1       50   150 +             50   150 +             100</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr2      100   200 -             50   150 -              50</span></span></code></pre></div>
<p>Both single set (e.g. <code><a href="../reference/bed_merge.html">bed_merge()</a></code>) and multi set operations will respect groupings in the input intervals.</p>
</div>
<div class="section level3">
<h3 id="column-specification">Column specification<a class="anchor" aria-label="anchor" href="#column-specification"></a>
</h3>
<p>Columns in <code>BEDtools</code> are referred to by position:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate the mean of column 6 for intervals in `b` that overlap with `a`</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bedtools</span> map <span class="at">-a</span> a.bed <span class="at">-b</span> b.bed <span class="at">-c</span> 6 <span class="at">-o</span> mean</span></code></pre></div>
<p>In <code>valr</code>, columns are referred to by name and can be used in multiple name/value expressions for summaries.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate the mean and variance for a `value` column</span></span>
<span><span class="fu"><a href="../reference/bed_map.html">bed_map</a></span><span class="op">(</span><span class="va">a</span>, <span class="va">b</span>, .mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span>, .var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">var</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># report concatenated and max values for merged intervals</span></span>
<span><span class="fu"><a href="../reference/bed_merge.html">bed_merge</a></span><span class="op">(</span><span class="va">a</span>, .concat <span class="op">=</span> <span class="fu"><a href="../reference/bed_map.html">concat</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span>, .max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level3">
<h3 id="meta-analysis">Meta-analysis<a class="anchor" aria-label="anchor" href="#meta-analysis"></a>
</h3>
<p>This demonstration illustrates how to use <code>valr</code> tools to perform a “meta-analysis” of signals relative to genomic features. Here we to analyze the distribution of histone marks surrounding transcription start sites.</p>
<p>First we load libraries and relevant data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `valr_example()` identifies the path of example files</span></span>
<span><span class="va">bedfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'genes.hg19.chr22.bed.gz'</span><span class="op">)</span></span>
<span><span class="va">genomefile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'hg19.chrom.sizes.gz'</span><span class="op">)</span></span>
<span><span class="va">bgfile</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/valr_example.html">valr_example</a></span><span class="op">(</span><span class="st">'hela.h3k4.chip.bg.gz'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bed</a></span><span class="op">(</span><span class="va">bedfile</span>, n_fields <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">genome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_genome.html">read_genome</a></span><span class="op">(</span><span class="va">genomefile</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_bed.html">read_bedgraph</a></span><span class="op">(</span><span class="va">bgfile</span><span class="op">)</span></span></code></pre></div>
<p>Then we generate 1 bp intervals to represent transcription start sites (TSSs). We focus on <code>+</code> strand genes, but <code>-</code> genes are easily accommodated by filtering them and using <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code> with <code>reversed</code> window numbers.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate 1 bp TSS intervals, `+` strand only</span></span>
<span><span class="va">tss</span> <span class="op">&lt;-</span> <span class="va">genes</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">strand</span> <span class="op">==</span> <span class="st">'+'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>end <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1000 bp up and downstream</span></span>
<span><span class="va">region_size</span> <span class="op">&lt;-</span> <span class="fl">1000</span></span>
<span><span class="co"># 50 bp windows</span></span>
<span><span class="va">win_size</span> <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span></span>
<span><span class="co"># add slop to the TSS, break into windows and add a group</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">tss</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/bed_slop.html">bed_slop</a></span><span class="op">(</span><span class="va">genome</span>, both <span class="op">=</span> <span class="va">region_size</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/bed_makewindows.html">bed_makewindows</a></span><span class="op">(</span><span class="va">win_size</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 13,530 × 7</span></span></span>
<span><span class="co">#&gt;    chrom    start      end name      score strand .win_id</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> chr22 16<span style="text-decoration: underline;">161</span>065 16<span style="text-decoration: underline;">161</span>115 LINC00516 3     +            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> chr22 16<span style="text-decoration: underline;">161</span>115 16<span style="text-decoration: underline;">161</span>165 LINC00516 3     +            2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> chr22 16<span style="text-decoration: underline;">161</span>165 16<span style="text-decoration: underline;">161</span>215 LINC00516 3     +            3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> chr22 16<span style="text-decoration: underline;">161</span>215 16<span style="text-decoration: underline;">161</span>265 LINC00516 3     +            4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> chr22 16<span style="text-decoration: underline;">161</span>265 16<span style="text-decoration: underline;">161</span>315 LINC00516 3     +            5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> chr22 16<span style="text-decoration: underline;">161</span>315 16<span style="text-decoration: underline;">161</span>365 LINC00516 3     +            6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> chr22 16<span style="text-decoration: underline;">161</span>365 16<span style="text-decoration: underline;">161</span>415 LINC00516 3     +            7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> chr22 16<span style="text-decoration: underline;">161</span>415 16<span style="text-decoration: underline;">161</span>465 LINC00516 3     +            8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> chr22 16<span style="text-decoration: underline;">161</span>465 16<span style="text-decoration: underline;">161</span>515 LINC00516 3     +            9</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> chr22 16<span style="text-decoration: underline;">161</span>515 16<span style="text-decoration: underline;">161</span>565 LINC00516 3     +           10</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 13,520 more rows</span></span></span></code></pre></div>
<p>Now we use the <code>.win_id</code> group with <code><a href="../reference/bed_map.html">bed_map()</a></code> to calculate a sum by mapping <code>y</code> signals onto the intervals in <code>x</code>. These data are regrouped by <code>.win_id</code> and a summary with <code>mean</code> and <code>sd</code> values is calculated.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># map signals to TSS regions and calculate summary statistics.</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bed_map.html">bed_map</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, win_sum <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">value</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">.win_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>win_mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">win_sum</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>            win_sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">win_sum</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 41 × 3</span></span></span>
<span><span class="co">#&gt;    .win_id win_mean win_sd</span></span>
<span><span class="co">#&gt;      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>       1     101.   85.8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>       2     111.   81.1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>       3     123.   99.1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>       4     116.   96.3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>       5     116.  102. </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>       6     125.   95.1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>       7     123.   94.4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>       8     128.   91.5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>       9     130.   95.7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>      10     130.   88.8</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 31 more rows</span></span></span></code></pre></div>
<p>Finally, these summary statistics are used to construct a plot that illustrates histone density surrounding TSSs.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">x_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="va">region_size</span>, <span class="va">region_size</span>, by <span class="op">=</span> <span class="va">win_size</span> <span class="op">*</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">x_breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">41</span>, by <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sd_limits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>ymax <span class="op">=</span> <span class="va">win_mean</span> <span class="op">+</span> <span class="va">win_sd</span>, ymin <span class="op">=</span> <span class="va">win_mean</span> <span class="op">-</span> <span class="va">win_sd</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">.win_id</span>, y <span class="op">=</span> <span class="va">win_mean</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_pointrange</a></span><span class="op">(</span><span class="va">sd_limits</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="va">x_labels</span>, breaks <span class="op">=</span> <span class="va">x_breaks</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Position (bp from TSS)'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">'Signal'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Human H3K4me3 signal near transcription start sites'</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="valr_files/figure-html/plot-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="api">API<a class="anchor" aria-label="anchor" href="#api"></a>
</h2>
<p>Function names are similar to their their <a href="https://bedtools.readthedocs.io/en/latest/" class="external-link">BEDtools</a> counterparts, with some additions.</p>
<div class="section level3">
<h3 id="data-types">Data types<a class="anchor" aria-label="anchor" href="#data-types"></a>
</h3>
<ul>
<li>Validate interval sets with <code><a href="../reference/ivl_df.html">check_interval()</a></code> and <code><a href="../reference/ivl_df.html">check_genome()</a></code>, which enforce strict column naming.</li>
</ul>
</div>
<div class="section level3">
<h3 id="reading-data">Reading data<a class="anchor" aria-label="anchor" href="#reading-data"></a>
</h3>
<ul>
<li><p>Read BED and related files with <code><a href="../reference/read_bed.html">read_bed()</a></code>, <code><a href="../reference/read_bed.html">read_bed12()</a></code>, <code><a href="../reference/read_bed.html">read_bedgraph()</a></code>, <code><a href="../reference/read_bed.html">read_narrowpeak()</a></code> and <code><a href="../reference/read_bed.html">read_broadpeak()</a></code>.</p></li>
<li><p>Read genome files containing chromosome name and size information with <code><a href="../reference/read_genome.html">read_genome()</a></code>.</p></li>
<li><p>Load VCF files with <code><a href="../reference/read_vcf.html">read_vcf()</a></code>.</p></li>
<li><p>Access remote databases with <code><a href="../reference/db.html">db_ucsc()</a></code> and <code><a href="../reference/db.html">db_ensembl()</a></code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="transforming-single-interval-sets">Transforming single interval sets<a class="anchor" aria-label="anchor" href="#transforming-single-interval-sets"></a>
</h3>
<ul>
<li><p>Adjust interval coordinates with <code><a href="../reference/bed_slop.html">bed_slop()</a></code> and <code><a href="../reference/bed_shift.html">bed_shift()</a></code>, and create new flanking intervals with <code><a href="../reference/bed_flank.html">bed_flank()</a></code>.</p></li>
<li><p>Combine nearby intervals with <code><a href="../reference/bed_merge.html">bed_merge()</a></code> and identify nearby intervals with <code><a href="../reference/bed_cluster.html">bed_cluster()</a></code>.</p></li>
<li><p>Generate intervals not covered by a query with <code><a href="../reference/bed_complement.html">bed_complement()</a></code>.</p></li>
<li><p>Order intervals with <code><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">dplyr::arrange()</a></code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="comparing-multiple-interval-sets">Comparing multiple interval sets<a class="anchor" aria-label="anchor" href="#comparing-multiple-interval-sets"></a>
</h3>
<ul>
<li><p>Find overlaps between sets of intervals with <code><a href="../reference/bed_intersect.html">bed_intersect()</a></code>.</p></li>
<li><p>Apply functions to overlapping sets of intervals with <code><a href="../reference/bed_map.html">bed_map()</a></code>.</p></li>
<li><p>Remove intervals based on overlaps with <code><a href="../reference/bed_subtract.html">bed_subtract()</a></code>.</p></li>
<li><p>Find overlapping intervals within a window with <code><a href="../reference/bed_window.html">bed_window()</a></code>.</p></li>
<li><p>Find closest intervals independent of overlaps with <code><a href="../reference/bed_closest.html">bed_closest()</a></code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="randomizing-intervals">Randomizing intervals<a class="anchor" aria-label="anchor" href="#randomizing-intervals"></a>
</h3>
<ul>
<li><p>Generate random intervals with <code><a href="../reference/bed_random.html">bed_random()</a></code>.</p></li>
<li><p>Shuffle the coordinates of intervals with <code><a href="../reference/bed_shuffle.html">bed_shuffle()</a></code>.</p></li>
<li><p>Sample input intervals with <code><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">dplyr::sample_n()</a></code> and <code><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">dplyr::sample_frac()</a></code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="interval-statistics">Interval statistics<a class="anchor" aria-label="anchor" href="#interval-statistics"></a>
</h3>
<ul>
<li><p>Calculate significance of overlaps between sets of intervals with <code><a href="../reference/bed_fisher.html">bed_fisher()</a></code> and <code><a href="../reference/bed_projection.html">bed_projection()</a></code>.</p></li>
<li><p>Quantify relative and absolute distances between sets of intervals with <code><a href="../reference/bed_reldist.html">bed_reldist()</a></code> and <code><a href="../reference/bed_absdist.html">bed_absdist()</a></code>.</p></li>
<li><p>Quantify extent of overlap between sets of intervals with <code><a href="../reference/bed_jaccard.html">bed_jaccard()</a></code>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a>
</h3>
<ul>
<li><p>Create features from BED12 files with <code><a href="../reference/create_introns.html">create_introns()</a></code>, <code><a href="../reference/create_tss.html">create_tss()</a></code>, <code><a href="../reference/create_utrs5.html">create_utrs5()</a></code>, and <code><a href="../reference/create_utrs3.html">create_utrs3()</a></code>.</p></li>
<li><p>Visualize the actions of valr functions with <code><a href="../reference/bed_glyph.html">bed_glyph()</a></code>.</p></li>
<li><p>Constrain intervals to a genome reference with <code><a href="../reference/bound_intervals.html">bound_intervals()</a></code>.</p></li>
<li><p>Subdivide intervals with <code><a href="../reference/bed_makewindows.html">bed_makewindows()</a></code>.</p></li>
<li><p>Convert BED12 to BED6 format with <code><a href="../reference/bed12_to_exons.html">bed12_to_exons()</a></code>.</p></li>
<li><p>Calculate spacing between intervals with <code><a href="../reference/interval_spacing.html">interval_spacing()</a></code>.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="related-work">Related work<a class="anchor" aria-label="anchor" href="#related-work"></a>
</h2>
<ul>
<li><p>Command-line tools <a href="https://bedtools.readthedocs.io/en/latest/" class="external-link">BEDtools</a> and <a href="https://bedops.readthedocs.io/en/latest/index.html" class="external-link">bedops</a>.</p></li>
<li><p>The Python library <a href="https://pythonhosted.org/pybedtools/" class="external-link">pybedtools</a> wraps BEDtools.</p></li>
<li><p>The R packages <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html" class="external-link">GenomicRanges</a>, <a href="https://CRAN.R-project.org/package=bedr" class="external-link">bedr</a>, <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html" class="external-link">IRanges</a> and <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529" class="external-link">GenometriCorr</a> provide similar capability with a different philosophy.</p></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>An R package developed by the <a href="https://medschool.cuanschutz.edu/rbi" class="external-link">RNA Bioscience Initiative</a></p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
